<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html
  PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><title>Cytoscape User Manual</title><meta name="generator" content="DocBook XSL Stylesheets V1.70.1"/></head><body><div class="article" lang="en"><div class="titlepage"><div><div><h1 class="title"><a id="d0e1"/>Cytoscape User Manual</h1></div></div><hr/></div><div class="toc"><p><b>Table of Contents</b></p><dl><dt><span class="section"><a href="#Cytoscape 2.6 User Manual">Cytoscape 2.6 User Manual</a></span></dt><dt><span class="section"><a href="#Introduction">Introduction</a></span></dt><dd><dl><dt><span class="section"><a href="#Development">Development</a></span></dt><dt><span class="section"><a href="#License">License</a></span></dt><dt><span class="section"><a href="#What’s New in 2.6">What’s New in 2.6</a></span></dt></dl></dd><dt><span class="section"><a href="#Launching Cytoscape">Launching Cytoscape</a></span></dt><dd><dl><dt><span class="section"><a href="#System requirements">System requirements</a></span></dt><dt><span class="section"><a href="#Getting Started">Getting Started</a></span></dt></dl></dd><dt><span class="section"><a href="#Quick Tour of Cytoscape">Quick Tour of Cytoscape</a></span></dt><dd><dl><dt><span class="section"><a href="#The Menus">The Menus</a></span></dt><dt><span class="section"><a href="#Network Management">Network Management</a></span></dt><dt><span class="section"><a href="#The Network Overview Window">The Network Overview Window</a></span></dt></dl></dd><dt><span class="section"><a href="#Command Line Arguments">Command Line Arguments</a></span></dt><dt><span class="section"><a href="#Cytoscape Preferences">Cytoscape Preferences</a></span></dt><dd><dl><dt><span class="section"><a href="#Managing Properties">Managing Properties</a></span></dt><dt><span class="section"><a href="#Managing Bookmarks">Managing Bookmarks</a></span></dt><dt><span class="section"><a href="#Managing Proxy Servers">Managing Proxy Servers</a></span></dt></dl></dd><dt><span class="section"><a href="#Creating Networks">Creating Networks</a></span></dt><dd><dl><dt><span class="section"><a href="#Import Fixed-Format Network Files">Import Fixed-Format Network Files</a></span></dt><dt><span class="section"><a href="#Import Free-Format Table Files">Import Free-Format Table Files</a></span></dt><dt><span class="section"><a href="#Import Networks from Web Services">Import Networks from Web Services</a></span></dt><dt><span class="section"><a href="#Edit a New Network">Edit a New Network</a></span></dt></dl></dd><dt><span class="section"><a href="#Supported Network File Formats">Supported Network File Formats</a></span></dt><dd><dl><dt><span class="section"><a href="#SIF Format">SIF Format</a></span></dt><dt><span class="section"><a href="#GML Format">GML Format</a></span></dt><dt><span class="section"><a href="#XGMML Format">XGMML Format</a></span></dt><dt><span class="section"><a href="#SBML (Systems Biology Markup Language) Format">SBML (Systems Biology Markup Language) Format</a></span></dt><dt><span class="section"><a href="#BioPAX (Biological PAthways eXchange) Format">BioPAX (Biological PAthways eXchange) Format</a></span></dt><dt><span class="section"><a href="#PSI-MI Format">PSI-MI Format</a></span></dt><dt><span class="section"><a href="#Delimited Text Table and Excel Workbook">Delimited Text Table and Excel Workbook</a></span></dt><dt><span class="section"><a href="#Node Naming Issues in Cytoscape">Node Naming Issues in Cytoscape</a></span></dt></dl></dd><dt><span class="section"><a href="#Node and Edge Attributes">Node and Edge Attributes</a></span></dt><dd><dl><dt><span class="section"><a href="#Cytoscape Attribute File Format">Cytoscape Attribute File Format</a></span></dt><dt><span class="section"><a href="#Import Attribute Table Files">Import Attribute Table Files</a></span></dt></dl></dd><dt><span class="section"><a href="#Loading Gene Expression (Attribute Matrix) Data">Loading Gene Expression (Attribute Matrix) Data</a></span></dt><dd><dl><dt><span class="section"><a href="#Data File Format">Data File Format</a></span></dt><dt><span class="section"><a href="#General Procedure">General Procedure</a></span></dt><dt><span class="section"><a href="#Worked Example">Worked Example</a></span></dt><dt><span class="section"><a href="#Detailed file format (Advanced users)">Detailed file format (Advanced users)</a></span></dt></dl></dd><dt><span class="section"><a href="#Importing Networks and Attributes from External Databases">Importing Networks and Attributes from External Databases</a></span></dt><dd><dl><dt><span class="section"><a href="#Web Service Client Manager">Web Service Client Manager</a></span></dt><dt><span class="section"><a href="#Getting Started">Getting Started</a></span></dt><dt><span class="section"><a href="#Example #1:  Retrieving Protein-Protein Interaction Networks from IntAct">Example #1:  Retrieving Protein-Protein Interaction Networks from IntAct</a></span></dt><dt><span class="section"><a href="#Example #2:  Retrieving Protein-Protein Interaction Networks from NCBI Entrez Gene">Example #2:  Retrieving Protein-Protein Interaction Networks from NCBI Entrez Gene</a></span></dt><dt><span class="section"><a href="#Example #3:  Retrieving Pathways and Networks from Pathway Commons">Example #3:  Retrieving Pathways and Networks from Pathway Commons</a></span></dt><dt><span class="section"><a href="#Future Directions">Future Directions</a></span></dt><dt><span class="section"><a href="#Import Attributes from External Database">Import Attributes from External Database</a></span></dt><dt><span class="section"><a href="#Use Multiple Services in a Workflow">Use Multiple Services in a Workflow</a></span></dt></dl></dd><dt><span class="section"><a href="#Navigation and Layout">Navigation and Layout</a></span></dt><dd><dl><dt><span class="section"><a href="#Basic Network Navigation">Basic Network Navigation</a></span></dt><dt><span class="section"><a href="#Other Mouse Behaviors">Other Mouse Behaviors</a></span></dt><dt><span class="section"><a href="#Automatic Layout Algorithms">Automatic Layout Algorithms</a></span></dt><dt><span class="section"><a href="#Manual Layout">Manual Layout</a></span></dt><dt><span class="section"><a href="#Node Movement and Placement">Node Movement and Placement</a></span></dt></dl></dd><dt><span class="section"><a href="#Visual Styles">Visual Styles</a></span></dt><dd><dl><dt><span class="section"><a href="#What is a Visual Style?">What is a Visual Style?</a></span></dt><dt><span class="section"><a href="#Introduction to the VizMapper User Interface">Introduction to the VizMapper User Interface</a></span></dt><dt><span class="section"><a href="#Introduction to Visual Styles">Introduction to Visual Styles</a></span></dt><dt><span class="section"><a href="#Visual Attributes, Graph Attributes and Visual Mappers">Visual Attributes, Graph Attributes and Visual Mappers</a></span></dt><dt><span class="section"><a href="#Visual Styles Tutorials">Visual Styles Tutorials</a></span></dt><dt><span class="section"><a href="#Advanced Topics">Advanced Topics</a></span></dt><dt><span class="section"><a href="#Managing Visual Styles">Managing Visual Styles</a></span></dt><dt><span class="section"><a href="#Bypassing Visual Styles">Bypassing Visual Styles</a></span></dt></dl></dd><dt><span class="section"><a href="#Finding and Filtering Nodes and Edges">Finding and Filtering Nodes and Edges</a></span></dt><dd><dl><dt><span class="section"><a href="#QuickFind">QuickFind</a></span></dt><dt><span class="section"><a href="#Filters">Filters</a></span></dt><dt><span class="section"><a href="#Using New Filters">Using New Filters</a></span></dt><dt><span class="section"><a href="#Using Old Filters">Using Old Filters</a></span></dt><dt><span class="section"><a href="#The Select Menu">The Select Menu</a></span></dt></dl></dd><dt><span class="section"><a href="#Editing Networks">Editing Networks</a></span></dt><dd><dl><dt><span class="section"><a href="#Add Interactions  (SIF Style)">Add Interactions  (SIF Style)</a></span></dt></dl></dd><dt><span class="section"><a href="#Plugins and the Plugin Manager">Plugins and the Plugin Manager</a></span></dt><dd><dl><dt><span class="section"><a href="#The Plugin Manager">The Plugin Manager</a></span></dt><dt><span class="section"><a href="#Get New Plugins">Get New Plugins</a></span></dt><dt><span class="section"><a href="#Delete Existing Plugins">Delete Existing Plugins</a></span></dt><dt><span class="section"><a href="#Update Existing Plugins">Update Existing Plugins</a></span></dt><dt><span class="section"><a href="#Download Plugins from a Custom Site">Download Plugins from a Custom Site</a></span></dt></dl></dd><dt><span class="section"><a href="#CytoPanels">CytoPanels</a></span></dt><dd><dl><dt><span class="section"><a href="#What are CytoPanels?">What are CytoPanels?</a></span></dt><dt><span class="section"><a href="#Basic Usage">Basic Usage</a></span></dt></dl></dd><dt><span class="section"><a href="#Rendering Engine">Rendering Engine</a></span></dt><dd><dl><dt><span class="section"><a href="#What is Level of Detail (LOD)?">What is Level of Detail (LOD)?</a></span></dt></dl></dd><dt><span class="section"><a href="#Annotation">Annotation</a></span></dt><dd><dl><dt><span class="section"><a href="#Ontology and Annotation File Format">Ontology and Annotation File Format</a></span></dt><dt><span class="section"><a href="#Node Name Mapping">Node Name Mapping</a></span></dt><dt><span class="section"><a href="#Import Ontology and Annotation">Import Ontology and Annotation</a></span></dt><dt><span class="section"><a href="#Custom Annotation Files for Ontologies Other than GO (for Advanced Users)">Custom Annotation Files for Ontologies Other than GO (for Advanced Users)</a></span></dt></dl></dd><dt><span class="section"><a href="#Linkout">Linkout</a></span></dt><dd><dl><dt><span class="section"><a href="#Adding or Removing Links">Adding or Removing Links</a></span></dt><dt><span class="section"><a href="#Use LinkOut from Attribute Browser">Use LinkOut from Attribute Browser</a></span></dt></dl></dd><dt><span class="section"><a href="#Acknowledgements">Acknowledgements</a></span></dt><dt><span class="section"><a href="#Appendix A: Old Annotation Server Format">Appendix A: Old Annotation Server Format</a></span></dt><dd><dl><dt><span class="section"><a href="#Building your own annotation files">Building your own annotation files</a></span></dt><dt><span class="section"><a href="#Load Data into Cytoscape">Load Data into Cytoscape</a></span></dt><dt><span class="section"><a href="#Getting and Reformatting GO Data">Getting and Reformatting GO Data</a></span></dt><dt><span class="section"><a href="#Python script examples">Python script examples</a></span></dt></dl></dd><dt><span class="section"><a href="#Appendix B: GNU Lesser General Public License">Appendix B: GNU Lesser General Public License</a></span></dt><dt><span class="section"><a href="#Appendix C: Increasing memory for Cytoscape">Appendix C: Increasing memory for Cytoscape</a></span></dt><dd><dl><dt><span class="section"><a href="#How to increase memory for Cytoscape">How to increase memory for Cytoscape</a></span></dt><dt><span class="section"><a href="#Changing memory allocations on Windows, Mac, and Linux machines">Changing memory allocations on Windows, Mac, and Linux machines</a></span></dt></dl></dd></dl></div><p> </p><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Cytoscape 2.6 User Manual"/>Cytoscape 2.6 User Manual</h2></div></div></div><p><span class="inlinemediaobject"><img src="images/cyto-logo-smaller.gif" alt="attachment:cyto-logo-smaller.gif"/></span> </p><p>This document is licensed under the <a href="http://creativecommons.org/licenses/by/2.5/" target="_top">Creative Commons license</a>, 2006 </p><p>Authors: The Cytoscape Collaboration </p><p>The Cytoscape project is an ongoing collaboration between: </p><div class="table"><a id="d0e27"/><p class="title"><b>Table 1. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p>University of California at San Diego </p></td><td><p> <span class="inlinemediaobject"><img src="images/ucsd-logo.gif" alt="attachment:ucsd-logo.gif"/></span> </p></td></tr><tr><td><p>Institute for Systems Biology </p></td><td><p> <span class="inlinemediaobject"><img src="images/isb-logo.gif" alt="attachment:isb-logo.gif"/></span> </p></td></tr><tr><td><p>Memorial Sloan-Kettering Cancer Center </p></td><td><p> <span class="inlinemediaobject"><img src="images/msk-logo.gif" alt="attachment:msk-logo.gif"/></span> </p></td></tr><tr><td><p>Institut Pasteur </p></td><td><p> <span class="inlinemediaobject"><img src="images/pasteur-logo.gif" alt="attachment:pasteur-logo.gif"/></span> </p></td></tr><tr><td><p>Agilent Technologies </p></td><td><p> <span class="inlinemediaobject"><img src="images/agilent-logo.gif" alt="attachment:agilent-logo.gif"/></span> </p></td></tr><tr><td><p>University of California at San Francisco </p></td><td><p> <span class="inlinemediaobject"><img src="images/ucsf-logo.gif" alt="attachment:ucsf-logo.gif"/></span> </p></td></tr></tbody></table></div></div><br class="table-break"/><p>Funding for Cytoscape is provided by a federal grant from the U.S. National Institute of General Medical Sciences (NIGMS) of the National Institutes of Health (NIH) under award number GM070743-01. Corporate funding is provided through a contract from Unilever PLC. </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Introduction"/>Introduction</h2></div></div></div><p>Cytoscape is a project dedicated to building open-source network visualization and analysis software. A software "Core" provides basic functionality to layout and query the network and to visually integrate the network with state data. The Core is extensible through a plug-in architecture, allowing rapid development of additional computational analyses and features. </p><p>Cytoscape's roots are in Systems Biology, where it is used for integrating biomolecular interaction networks with high-throughput expression data and other molecular state information. Although applicable to any system of molecular components and interactions, Cytoscape is most powerful when used in conjunction with large databases of protein-protein, protein-DNA, and genetic interactions that are increasingly available for humans and model organisms. Cytoscape allows the visual integration of the network with expression profiles, phenotypes, and other molecular state information, and links the network to databases of functional annotations. </p><p><span class="strong"><strong>The central organizing metaphor of Cytoscape is a network (graph), with genes, proteins, and molecules represented as nodes and interactions represented as links, i.e. edges, between nodes.</strong></span> </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Development"/>Development</h3></div></div></div><p>Cytoscape is a collaborative project between the Institute for Systems Biology (Leroy Hood lab), the University of California San Diego (Trey Ideker lab), Memorial Sloan-Kettering Cancer Center (Chris Sander lab), the Institut Pasteur (Benno Schwikowski lab), Agilent Technologies (Annette Adler lab) and the University of California, San Francisco (Bruce Conklin lab). </p><p>Visit <a href="http://www.cytoscape.org" target="_top">http://www.cytoscape.org</a> for more information. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="License"/>License</h3></div></div></div><p>Cytoscape is protected under the GNU LGPL (Lesser General Public License). The License is included as an appendix to this manual, but can also be found online: <a href="http://www.gnu.org/copyleft/lesser.txt" target="_top">http://www.gnu.org/copyleft/lesser.txt</a>. Cytoscape also includes a number of other open source libraries, which are detailed in the<a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape User Manual/Acknowledgements" target="_top">Cytoscape User Manual/Acknowledgements</a> below. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="What’s New in 2.6"/>What’s New in 2.6</h3></div></div></div><p>Cytoscape version 2.6 contains several new features, plus improvements to the performance and usability of the software. These include: </p><div class="itemizedlist"><ul type="disc"><li>Web Service Client Manager framework to integrate web service clients into Cytoscape. </li><li><p>Web Service client plugins for downloading networks from <a href="http://www.pathwaycommons.org/pc/" target="_top">PathwayCommons</a>, <a href="http://www.ebi.ac.uk/intact/site/index.jsf" target="_top">IntAct</a>, and <a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=gene" target="_top">NCBI Entrez Gene</a>.  </p></li><li><p>Annotation import web service plugin for <a href="http://www.biomart.org/" target="_top">BioMart</a>.  This is mainly for ID translation/synonym mapping.  </p></li><li>Cytoscape themes. </li><li>Dynamic filters. </li><li>Network Manager supports multiple network selection. </li><li>Label Positioning has been improved. </li><li>Session saving occurs in memory. </li><li>XGMML Improvements. </li><li>Network loading improvements. </li><li>Linkout integrated with attribute browser. </li><li>Extra sample Visual Styles using new visual properties introduced in 2.5. </li><li>Many, many bug fixes! </li></ul></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Launching Cytoscape"/>Launching Cytoscape</h2></div></div></div><p>Cytoscape is a Java application verified to run on Linux, Windows, and Mac OS X.  Although not officially supported, other UNIX platforms such as Solaris or FreeBSD may run Cytoscape if Java version 5 or later is available for the platform.   </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="System requirements"/>System requirements</h3></div></div></div><p>The system requirements for Cytoscape depend on the size of the networks the user wants to load, view and manipulate. </p><div class="table"><a id="d0e211"/><p class="title"><b>Table 2. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/></colgroup><tbody><tr><td><p>  </p></td><td><p> <span class="strong"><strong>Small Network Visualization</strong></span> </p></td><td><p> <span class="strong"><strong>Large Network Analysis/Visualization</strong></span> </p></td></tr><tr><td><p> <span class="strong"><strong>Processor</strong></span> </p></td><td><p> 1GHz </p></td><td><p> As fast as possible </p></td></tr><tr><td><p> <span class="strong"><strong>Memory</strong></span> </p></td><td><p> 512MB </p></td><td><p> 2GB+ </p></td></tr><tr><td><p> <span class="strong"><strong>Graphics Card</strong></span> </p></td><td><p> On board Video </p></td><td><p> Highend Graphics Card </p></td></tr><tr><td><p> <span class="strong"><strong>Monitor</strong></span> </p></td><td><p> XGA (1024X768) </p></td><td><p> Wide or Dual Monitor </p></td></tr></tbody></table></div></div><br class="table-break"/><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Getting Started"/>Getting Started</h3></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Install Java"/>Install Java</h4></div></div></div><p>If not already installed on your computer, download and install Java SE 5 or 6.  <span class="emphasis"><em><span class="strong"><strong>Cytoscape 2.5 will no longer run with Java version 1.4.x or lower.  You must install Java SE 5 or 6!!!</strong></span></em></span> </p><p>These can be found at: </p><p><a href="http://java.sun.com/javase/downloads/index_jdk5.jsp" target="_top">Java SE 5</a> <a href="http://java.sun.com/javase/downloads/index.jsp" target="_top">Java SE 6</a> </p><p>In general, Java SE 6 is faster than 5.  If your machine is compatible with the 6 series, please try version 6.   </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Install Cytoscape"/>Install Cytoscape</h4></div></div></div><p>There are a number of options for downloading and installing Cytoscape.  All options can be downloaded from the <a href="http://cytoscape.org" target="_top">http://cytoscape.org</a> website. </p><div class="itemizedlist"><ul type="disc"><li>Automatic installation packages exist for Windows, Mac OS X, and Linux platforms.   </li><li>You can install Cytoscape from a compressed archive distribution. </li><li>You can build Cytoscape from the source code. </li><li><p>You can check out the latest and greatest software from our <a href="http://chianti.ucsd.edu/svn" target="_top">Subversion repository</a>. </p></li></ul></div><p>Cytoscape installations (regardless of platform) containing the following files and directories: </p><div class="table"><a id="d0e335"/><p class="title"><b>Table 3. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>File</strong></span></p></td><td><p><span class="strong"><strong>Description</strong></span></p></td></tr><tr><td><p>cytoscape.jar</p></td><td><p>Main Cytoscape application (Java archive)</p></td></tr><tr><td><p>cytoscape.sh</p></td><td><p>Script to run Cytoscape from command line (Linux, Mac OS X)</p></td></tr><tr><td><p>cytoscape.bat </p></td><td><p>Script to run Cytoscape (Windows)</p></td></tr><tr><td><p>LICENSE.txt/html</p></td><td><p>Cytoscape GNU LGPL License</p></td></tr><tr><td><p>lib/</p></td><td><p>library jar files needed to run Cytoscape.</p></td></tr><tr><td><p>docs/</p></td><td><p>Manuals in different formats.  What you are reading now.</p></td></tr><tr><td><p>licenses/</p></td><td><p>Licence files for the various libraries distributed with Cytoscape.</p></td></tr><tr><td><p>plugins/</p></td><td><p>Directory containing cytoscape plugins, in .jar format.</p></td></tr><tr><td><p>sampleData/</p></td><td> </td></tr><tr><td><p> </p></td><td><p>galFiltered.gml -- Sample molecular interaction network file *</p></td></tr><tr><td><p> </p></td><td><p>galFiltered.sif -- Identical network in Simple Interaction Format *</p></td></tr><tr><td><p> </p></td><td><p>galExpData.pvals -- Sample gene expression matrix file *</p></td></tr><tr><td><p> </p></td><td><p>galFiltered.nodeAttrTable.xls -- Sample node attribute file in Microsoft Excel format</p></td></tr><tr><td><p> </p></td><td><p>galFiltered.cys -- Sample session file created from datasets above plus annotations from several databases *</p></td></tr><tr><td><p> </p></td><td><p>BINDyeast.sif -- Network of all yeast protein-protein interactions in the BIND database as of Dec, 2006 **</p></td></tr><tr><td><p> </p></td><td><p>BINDhuman.sif -- Network of all human protein-protein interactions in the BIND database as of Dec, 2006 **</p></td></tr><tr><td><p> </p></td><td><p>yeastHighQuality.sif -- Sample molecular interaction network file ***</p></td></tr><tr><td><p> </p></td><td><p>interactome_merged.networkTable.gz -- Human interactome network file in tab-delimited format ****</p></td></tr><tr><td><p> </p></td><td><p>sampleStyles.props -- Additional sample Visual Styles </p></td></tr></tbody></table></div></div><br class="table-break"/><p>* From Ideker et al., Science 292:929 (2001) </p><p>** Obtained from data hosted at <a href="http://www.blueprint.org/bind/bind_downloads.html" target="_top">http://www.blueprint.org/bind/bind_downloads.html</a> </p><p>** From von Mering et al., Nature, 417:399 (2002) and Lee et al, Science 298:799 (2002) </p><p>**** Created from Cytoscape tutorial web page.  Original data sets are available at: <a href="http://www.cytoscape.orghttp://cytoscape.org/cgi-bin/moin.cgi/Data_Sets/" target="_top">http://www.cytoscape.orghttp://cytoscape.org/cgi-bin/moin.cgi/Data_Sets/</a>  from "A merged human interactome" by Andrew Garrow, Yeyejide Adeleye and Guy Warner (Unilever, Safety and Environmental Assurance Center). </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Launch the application"/>Launch the application</h4></div></div></div><p>Double-click on the icon created by the installer or by running <code class="code">cytoscape.sh</code> from the command line (Linux or Mac OS X) or double-clicking <code class="code">cytoscape.bat</code> (Windows). Alternatively, you can pass the .jar file to Java directly using the command <code class="code">java -Xmx512M -jar cytoscape.jar -p plugins</code>. The <code class="code">-Xmx512M</code> flag tells java to allocate more memory for Cytoscape and the <code class="code">-p plugins</code> option tells cytoscape to load all of the plugins in the plugins directory.  Loading the plugins is important because many key features like layouts, filters and the attribute browser are included with Cytoscape as plugins in the <code class="code">plugins</code> directory. See the <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Command_Line_Arguments" target="_top">Command Line</a> chapter for more detail. In Windows, it is also possible to directly double-click the .jar file to launch it. However, this does not allow specification of command-line arguments (such as the location of the plugin directory).  </p><p>When you succeed in launching Cytoscape, a window will appear that looks like this (captured on Mac OS 10.4): </p><p><span class="inlinemediaobject"><img src="images/cytoscape_startup_mac.png" alt="attachment:cytoscape_startup_mac.png"/></span> </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Note on Memory Consumption"/>Note on Memory Consumption</h5></div></div></div><p>For users interested in loading large networks, the amount of memory needed by Cytoscape will increase. Memory usage depends on both number of network objects (nodes+edges) and the number of attributes. Here are some <span class="emphasis"><em>rough</em></span> suggestions for memory allocation: </p><p><span class="emphasis"><em>Suggested Memory Size <span class="strong"><strong>Without</strong></span> View</em></span> </p><div class="table"><a id="d0e548"/><p class="title"><b>Table 4. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p>Number of Objects (nodes + edges)</p></td><td><p>Suggested Memory Size</p></td></tr><tr><td><p>0 - 70,000</p></td><td><p> 512M (default)</p></td></tr><tr><td><p>70,000 - 150,000</p></td><td><p>800M</p></td></tr></tbody></table></div></div><br class="table-break"/><p><span class="emphasis"><em>Suggested Memory Size <span class="strong"><strong>With</strong></span> View</em></span> </p><div class="table"><a id="d0e582"/><p class="title"><b>Table 5. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p>Number of Objects (nodes + edges)</p></td><td><p>Suggested Memory Size</p></td></tr><tr><td><p>0 - 20,000</p></td><td><p>512M (default)</p></td></tr><tr><td><p>20,000 - 70,000</p></td><td><p>800M</p></td></tr><tr><td><p>70,000 - 150,000</p></td><td><p>1G</p></td></tr></tbody></table></div></div><br class="table-break"/><p/><div class="section" lang="en"><div class="titlepage"><div><div><h6 class="title"><a id="Overall Memory Size for Cytoscape"/>Overall Memory Size for Cytoscape</h6></div></div></div><p>To increase the maximum memory size for Cytoscape, you can specify it from command line.  For example, if you want to assign 1GB of memory, type: </p><p><span class="emphasis"><em><span class="strong"><strong>java -Xmx1GB -jar cytoscape.jar -p plugins</strong></span></em></span> </p><p>from the command line. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h6 class="title"><a id="Stack Size"/>Stack Size</h6></div></div></div><p>There is one more option related to memory allocation.  Some of the functions in Cytoscape use larger stack space (a temporary memory for some operations, such as Layout).  Since this value is set independently from the <span class="emphasis"><em>Xmx</em></span> value above, sometimes Layout algorithms fails due to the <span class="emphasis"><em>out of memory</em></span> error.  To avoid this, you can set larger heap size by <span class="emphasis"><em><span class="strong"><strong>-Xss</strong></span></em></span> option.  If layout fails for large networks, please try the following: </p><p><span class="emphasis"><em><span class="strong"><strong>java -Xmx1GB -Xss10M -jar cytoscape.jar -p plugins</strong></span></em></span> </p><p>The option <span class="emphasis"><em><span class="strong"><strong>-Xss10M</strong></span></em></span> means set the heap size to 10MB.  In many cases, this solves <span class="emphasis"><em>out of memory</em></span> error for Layouts.  </p><p><span class="inlinemediaobject"><img src="images/one_million_network.png" alt="attachment:one_million_network.png"/></span> </p><p><span class="strong"><strong>Randomly generated scale-free network with 500K nodes and 500k edges:</strong></span>  If memory parameters are set properly, you can visualize huge networks.  In this example, about 5GB of memory is used by Cytoscape.  Stack size is set to 10MB.  To use large memory space (4GB+) you need 64-bit version operating system AND 64-bit version Java SE 5 or 6. </p><p>Note: Some of the web service clients are multi-thread programs and each thread uses the memory size specified by <span class="emphasis"><em><span class="strong"><strong>-Xss</strong></span></em></span> option.  If web service clients fails due to the <span class="emphasis"><em>out of memory</em></span> error, please reduce the stack size and try again. </p><p>For more details, see <a href="http://cytoscape.org/cgi-bin/moin.cgi/How to increase memory for Cytoscape" target="_top">How to increase memory for Cytoscape</a>. </p><p/></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Note on Directory Location"/>Note on Directory Location</h4></div></div></div><p>For the application to work properly, all files should be left in the directory in which they were unpacked. The core Cytoscape application assumes this directory structure when looking for the various libraries needed to run the application.  If you are adventurous, you can get creative with the $CLASSPATH and/or the cytoscape.jar manifest file and run Cytoscape from any location you want. </p></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Quick Tour of Cytoscape"/>Quick Tour of Cytoscape</h2></div></div></div><p>When a network is loaded, Cytoscape will look something like the image below: </p><p><span class="inlinemediaobject"><img src="images/cytoscape_startup_network_26.png" alt="attachment:cytoscape_startup_network_26.png"/></span> </p><p>The main window here has several components: </p><div class="itemizedlist"><ul type="disc"><li>The menu bar at the top (see below for more information about each menu). </li><li>The toolbar, which contains icons for commonly used functions. These functions are also available via the menus. Hover the mouse pointer over an icon and wait momentarily for a description to appear as a tooltip. </li><li>The network management panel (top left panel). This contains an optional network overview pane (shown at the bottom left). </li><li>The main network view window, which displays the network. </li><li>The attribute browser panel (bottom panel), which displays attributes of selected nodes and edges and enables you to modify the values of attributes. </li></ul></div><p>The network management and attribute browser panels are dockable tabbed panels known as CytoPanels.  You can undock any of these panels by clicking on the Float Window control <span class="inlinemediaobject"><img src="images/float_icon.png" alt="attachment:float_icon.png"/></span> in the upper-right corner of the CytoPanel. </p><p>If you select this control, e.g. on the attribute browser panel, you will now have two Cytoscape windows, the main window, and a new window labeled CytoPanel 2, similar to the one shown below.  Popup will be displayed when you put the mouse pointer on a cell. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/attribute_browser_26.png" alt="attachment:attribute_browser_26.png"/></span> </p></li></ul></div><p>Note that CytoPanel 2 now has a Dock Window control. If you select this control, the window will dock onto the main window. </p><p>Cytoscape also has an editor that enables you to build and modify networks interactively by dragging and dropping nodes and edges from a palette onto the main network view window.  The Node shapes and Edge arrows on the palette are defined by the currently used Visual Style.  To edit a network, just select the Editor tab on CytoPanel 1.  An example of an editor, with the palette contained in CytoPanel 1 and defined by the BioMoleculeEditor Visual Style, is shown below. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/editor_25.png" alt="attachment:editor_25.png"/></span> </p></li></ul></div><p><a id="menus"/> </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="The Menus"/>The Menus</h3></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="File"/>File</h4></div></div></div><p>The File menu contains most basic file functionality:  File → Open for opening a Cytoscape session file; File → New for creating a new network, either blank for editing, or from an existing network; File → Save for saving a session file; File → Import for importing data such as networks and attributes; and File → Export for exporting data and images. Also, File → Print allows printing, while File → Quit closes all windows of Cytoscape and exits the program. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_file_26.png" alt="attachment:menu_file_26.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Edit"/>Edit</h4></div></div></div><p>The Edit menu contains Undo and Redo functions which undo and redo edits made in the Attribute Browser, the Network Editor and to layout.  </p><p>There are also options for creating and destroying views (graphical representations of a network) and networks (the raw network data – not yet visualized), as well as an option for deleting selected nodes and edges from the current network. All deleted nodes and edges can be restored to the network via Edit → Undo. Editing preferences for properties and plugins is found under Edit → Preferences → Properties... . </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_edit_26.png" alt="attachment:menu_edit_26.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="View"/>View</h4></div></div></div><p>The View menu allows you to display or hide the network management panel (CytoPanel 1), the attribute browser (CytoPanel 2), the Network Overview (in CytoPanel 1), and the VizMapper. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_view_26.png" alt="attachment:menu_view_26.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Select"/>Select</h4></div></div></div><p>The Select menu contains different options for selecting nodes and edges.  It also contains the Select → Use Filters option, which allows filters to be created for automatic selection of portions of a network whose node or edge attributes meet a filtering criterion. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_select_26.png" alt="attachment:menu_select_26.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Layout"/>Layout</h4></div></div></div><p>The Layout menu has an array of features for visually organizing the network. The features in the top portion of the network (Rotate, Scale, Align and Distribute) are tools for manipulating the network visualization. The bottom section of the menu lists a variety of layout algorithms which automatically lay a network out. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_layout_25.png" alt="attachment:menu_layout_25.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Plugins"/>Plugins</h4></div></div></div><p>The Plugins menu contains options for managing (install/update/delete) your plugins and may have options added by plugins that have been installed, such as the Agilent Literature Search or Merge Networks. Depending on which plugins are loaded, the plugins that you see may be different than what appear here. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_plugins_25.png" alt="attachment:menu_plugins_25.png"/></span> </p></li></ul></div><div class="table"><a id="d0e857"/><p class="title"><b>Table 6. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p>Note: A list of available Cytoscape plugins with descriptions is available online at: <a href="http://cytoscape.org/plugins2.php" target="_top">http://cytoscape.org/plugins2.php</a></p></td></tr></tbody></table></div></div><br class="table-break"/><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Help"/>Help</h4></div></div></div><p>The Help menu allows you to launch the online help viewer and browse the table of contents for this manual. The “About…” option displays information about the running version of Cytoscape. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_help_25.png" alt="attachment:menu_help_25.png"/></span> </p></li></ul></div><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Network Management"/>Network Management</h3></div></div></div><p>Cytoscape 2.3 and newer versions allow multiple networks to be loaded at a time, either with or without a view. A network stores all the nodes and edges that are loaded by the user and a view displays them. You can have many views of the same network. Networks (and their optionally associated views) can be organized hierarchically. </p><p>An example where a number of networks have been loaded and arranged hierarchically is shown below: </p><p><span class="inlinemediaobject"><img src="images/cytoscape_network_hierarchy_25.png" alt="attachment:cytoscape_network_hierarchy_25.png"/></span> </p><p>The network manager (top-right tree view in CytoPanel 1) shows the networks that are loaded. Clicking on a network here will make that view active in the main window, if the view exists (green highlighted networks only). Each network has a name and size (number of nodes and edges), which are shown in the network manager. If a network is loaded from a file, the network name is the name of the file. </p><p>Some networks are very large (thousands of nodes and edges) and can take a long time to display. For this reason, a network in Cytoscape may not contain a ‘view’. Networks that have a view are highlighted in green and networks that don’t have a view are highlighted in red. You can create or destroy a view for a network by right-clicking the network name in the network manager or by choosing the appropriate option in the Edit menu. You can also destroy previously loaded networks this way. In the picture above, seven networks are loaded, six green ones with views and one red one without a view. </p><p>Certain operations in Cytoscape will create new networks. If a new network is created from an old network, for example by selecting a set of nodes in one network and copying these nodes to a new network (via the File → New → Network option), it will be shown as a child of the network that it was derived from. In this way, the relationships between networks that are loaded in Cytoscape can be seen at a glance. Networks in the top part of the tree in the figure above were generated in this manner. </p><p>The available network views are also arranged as multiple overlapping windows in the network view window. You can maximize, minimize, and destroy network views by using the normal window controls for your operating system. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Arrange Network Windows"/>Arrange Network Windows</h4></div></div></div><p>When you work on multiple networks, you can arrange the network view windows from <span class="emphasis"><em><span class="strong"><strong>View → Arrange Network Windows</strong></span></em></span>.  You can re-arrange the network location by these commands. </p><p><span class="inlinemediaobject"><img src="images/arrange_26_1.png" alt="attachment:arrange_26_1.png"/></span>  <span class="inlinemediaobject"><img src="images/arrange_26_3.png" alt="attachment:arrange_26_3.png"/></span> </p><p><span class="inlinemediaobject"><img src="images/arrange_26_2.png" alt="attachment:arrange_26_2.png"/></span>  <span class="inlinemediaobject"><img src="images/arrange_26_4.png" alt="attachment:arrange_26_4.png"/></span> </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="The Network Overview Window"/>The Network Overview Window</h3></div></div></div><p>The network overview window shows an overview (or ‘bird’s eye view’) of the network. It can be used to navigate around a large network view. The blue rectangle indicates the portion of the network currently displayed in the network view window, and it can be dragged with the mouse to view other portions of the network. Zooming in will cause the rectangle to appear smaller and vice versa.  </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/network_overview_25.png" alt="attachment:network_overview_25.png"/></span> </p></li></ul></div></div></div><div class="itemizedlist"><ul type="disc"><li><a id="command_line_args"/></li></ul></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Command Line Arguments"/>Command Line Arguments</h2></div></div></div><p>Cytoscape recognizes a number of optional command line arguments, including run-time specification of network files, attribute files, and session files. This is the output generated when the cytoscape is executed with the "-h" or "--help" flag: </p><p/><pre class="screen">usage: java -Xmx512M -jar cytoscape.jar [OPTIONS]
 -h,--help                      Print this message.
 -v,--version                 Print the version number.
 -s,--session &lt;file&gt;        Load a cytoscape session (.cys) file.
 -N,--network &lt;file&gt;     Load a network file (any format).
 -e,--edge-attrs &lt;file&gt;    Load an edge attributes file (edge attribute format).
 -n,--node-attrs &lt;file&gt;   Load a node attributes file (node attribute format).
 -m,--matrix &lt;file&gt;        Load a node attribute matrix file (table).
 -p,--plugin &lt;file&gt;         Load a plugin jar file, directory of jar files,
                                     plugin class name, or plugin jar URL.
 -P,--props &lt;file&gt;         Load cytoscape properties file (Java properties
                                    format) or individual property: -P name=value.
 -V,--vizmap &lt;file&gt;      Load vizmap properties file (Java properties format).
</pre><p>Any file specified for an option may be specified as either a path or as a URL.  For example you can specify a network as a file (assuming that myNet.sif exists in the current working directory): <code class="code">cytoscape.sh -N myNet.sif</code>. Or you can specify a network as a URL: <code class="code">cytoscape.sh -N http://example.com/myNet.sif</code>. </p><div class="table"><a id="d0e984"/><p class="title"><b>Table 7. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Argument</strong></span> </p></td><td><p><span class="strong"><strong>Description</strong></span> </p></td></tr><tr><td><p>-h,--help </p></td><td><p>This flag generates the help output you see above and exits. </p></td></tr><tr><td><p>-v,--version </p></td><td><p>This flag prints the version number of Cytoscape and exits. </p></td></tr><tr><td><p>-s,--session &lt;file&gt; </p></td><td><p>This option specifies a session file to be loaded.  Since only one session file can be loaded at a given time, this option may only specified once on a given command line.  The option expects a <code class="code">.cys</code> Cytoscape session file.  It is customary, although not necessary, for session file names to contain the .cys extension. </p></td></tr><tr><td><p>-N,--network &lt;file&gt; </p></td><td><p>This option is used to load all types of network files.  SIF, GML, and XGMML files can all be loaded using the -N option.  You can specify as many networks as desired on a single command line. </p></td></tr><tr><td><p>-e,--edge-attrs &lt;file&gt; </p></td><td><p>This option specifies an edge attributes file.  You may specify as many edge attribute files as desired on a single command line. </p></td></tr><tr><td><p>-n,--node-attrs &lt;file&gt; </p></td><td><p>This option specifies a node attributes file.  You may specify as many node attribute files as desired on a single command line. </p></td></tr><tr><td><p>-m,--matrix &lt;file&gt; </p></td><td><p>This option specifies a data matrix file.  In a biological context, the data matrix consists of expression data. All data matrix files are read into node attributes. You may specify as many data matrix files as desired on a single command line. </p></td></tr><tr><td><p>-p,--plugin &lt;file&gt; </p></td><td><p>This option specifies a cytoscape plugin (.jar) file to be loaded by Cytoscape. This option also subsumes the previous "resource plugin option".   You may specify a class name that identifies your plugin and the plugin will be loaded if the plugin is in Cytoscape's CLASSPATH.  For example, assuming that the class MyPlugin can be found in the CLASSPATH, you could specify the plugin like this: <code class="code">cytoscape.sh -p MyPlugin.class</code>.  A final means of specifying plugins is to specify a file name whose contents contain a list of plugin jar files. </p></td></tr><tr><td><p>-P,--props &lt;file&gt; </p></td><td><p>This option specifies Cytoscape properties.  Properties can be specified either as a properties file (in Java's standard properties format), or as individual properties. To specify individual properties, you must specify the property name followed by the property value where the name and value are separated by the '=' sign.  For example to specify the defaultSpeciesName:  <code class="code">cytoscape.sh -P defaultSpeciesName=Human</code>.  If you would like to include spaces in your property, simply enclose the name and value in quotation marks:  <code class="code">cytoscape.sh -P "defaultSpeciesName=Homo Sapiens"</code>.  The property option subsumes previous options -noCanonicalization, -species, and -bioDataServer.  Now it would look like: <code class="code">cytoscape.sh -P defaultSpeciesName=Human -P noCanonicalization=true -P bioDataServer=myServer</code>. </p></td></tr><tr><td><p>-V,--vizmap &lt;file&gt; </p></td><td><p>This option specifies a visual properties file. </p></td></tr></tbody></table></div></div><br class="table-break"/><p>All options described above (including plugins) can be loaded from the GUI once Cytoscape is running. </p></div><div class="itemizedlist"><ul type="disc"><li><a id="properties"/></li></ul></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Cytoscape Preferences"/>Cytoscape Preferences</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Managing Properties"/>Managing Properties</h3></div></div></div><div class="table"><a id="d0e1099"/><p class="title"><b>Table 8. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Important!</strong></span> If you have used previous versions of Cytoscape, you will notice that handling of properties has changed.  The most important change is that properties are no longer saved by default to the current directory or to your home <code class="code">.cytoscape</code> directory.  Properties are stored by default in Cytoscape session files (<code class="code">.cys</code> extension).  The <code class="code">cytoscape.props</code> file still exists in the <code class="code">.cytoscape</code> directory but is only written to when the user explicitly requests that the current settings be made the defaults for all future sessions of Cytoscape.  Unless you have something important in your <code class="code">.cytoscape/cytoscape.props</code> file, your best bet will be to delete the file and use the defaults.</p></td></tr></tbody></table></div></div><br class="table-break"/><p>The Cytoscape Properties editor, accessed via Edit → Preferences → Properties…, is used to specify general and default properties. Properties are now stored in Cytoscape session files, so changes to general properties will be saved as part of the current session, but will only carry over to subsequent sessions if they are set as defaults or exported using the File → Export function. </p><p>Cytoscape properties are configurable via Add, Modify and Delete operations. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/prefs_editor.png" alt="attachment:prefs_editor.png"/></span> </p></li></ul></div><p>Some common properties are described below. </p><div class="table"><a id="d0e1141"/><p class="title"><b>Table 9. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Property name</strong></span></p></td><td><p><span class="strong"><strong>Default value</strong></span></p></td><td><p><span class="strong"><strong>Valid values</strong></span></p></td></tr><tr><td><p>viewThreshold </p></td><td><p>10000 </p></td><td><p>integer &gt; 0 </p></td></tr><tr><td><p>secondaryViewThreshold </p></td><td><p>30000 </p></td><td><p>integer &gt; 0 </p></td></tr><tr><td><p>viewType </p></td><td><p>tabbed </p></td><td><p>tabbed </p></td></tr><tr><td><p>defaultWebBrowser </p></td><td><p> </p></td><td><p>A path to the web browser on your system. This only needs to be specified if Cytoscape can’t find the web browser on your system. </p></td></tr></tbody></table></div></div><br class="table-break"/><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Setting Default Properties"/>Setting Default Properties</h4></div></div></div><p>It is possible to alter the default properties for Cytoscape.  </p><p>Edit the properties via Edit → Preferences → Properties... and check the Make Current Cytoscape Properties Default checkbox. This will save the current properties to the <code class="code">.cytoscape</code> directory, where they will then be applied to all of your Cytoscape sessions from that point on. Otherwise, Cytoscape will automatically save the properties used in a particular session inside its <code class="code">.cys</code> session file, while the default properties will be applied at the beginning of subsequent sessions. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Managing Bookmarks"/>Managing Bookmarks</h3></div></div></div><p>Cytoscape contains a pre-defined list of bookmarks, which point to sample network files located on the Cytoscape web server. Users may add, modify, and delete bookmarks through the Bookmark manager, accessed  by going to Edit → Preferences → Bookmarks… .   </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/BookmarkManager.PNG" alt="attachment:BookmarkManager.PNG"/></span> </p></li></ul></div><p>There are currently two types of bookmarks: network and annotation.  Network bookmarks are URLs pointing to network files available on the Internet.  These are nomal networks that can be loaded into Cytoscape.  The annotation bookmarks are URLs pointing to ontology annotation files.  The annotation bookmarks are only used when importing an ontology. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Managing Proxy Servers"/>Managing Proxy Servers</h3></div></div></div><p>You can define and configure a proxy server for Cytoscape by going to Edit → Preferences → Proxies… . </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/ProxyServerSetting.PNG" alt="attachment:ProxyServerSetting.PNG"/></span> </p></li></ul></div><p>After the proxy server is set, all network traffic related to loading a network from URL will pass through the proxy server.  Other plugins use this capability as well. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Creating Networks"/>Creating Networks</h2></div></div></div><p>There are 4 different ways of creating networks in Cytoscape: </p><div class="orderedlist"><ol type="1"><li>Importing pre-existing, formatted network files. </li><li>Importing pre-existing, unformatted text or Excel files. </li><li>Importing networks from Web Service. </li><li>Creating an empty network and manually adding nodes and edges. </li></ol></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Import Fixed-Format Network Files"/>Import Fixed-Format Network Files</h3></div></div></div><p>Network files can be specified in any of the formats described in the <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Network_Formats" target="_top">Supported Network Formats</a> chapter. Networks are imported into Cytoscape through the "Import Network" window, which can be accessed by going to File → Import → Network (multiple file types). The network file can either be located directly on the local computer, or found on a remote computer (in which case it will be referenced with a URL). </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Load Networks from Local Computer"/>Load Networks from Local Computer</h4></div></div></div><p>By default, Cytoscape loads networks from the local computer. </p><p>The Import Networks dialog shows a default setting of "Data Source Type: Local," meaning that network files from the local computer will be available for importing. Choose the correct file by clicking on the Select button (only file types that Cytoscape recognizes will be shown), and then load the network by clicking on the Import button. Some sample network files of different types have been included in the sampleData folder in Cytoscape. </p><p>Network files in SIF, GML, and XGMML formats may also be loaded directly from the command line using the –N option. </p><p><span class="inlinemediaobject"><img src="images/network_import_dialog1_25.png" alt="attachment:network_import_dialog1_25.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Load Networks from a Remote Computer (URL import)"/>Load Networks from a Remote Computer (URL import)</h4></div></div></div><p>The Import Networks dialog is also capable of importing network files using a URL. To do this, set the Data Source Type to Remote and insert the appropriate URL, either manually or using URL bookmarks. Bookmarked URLs can be accessed by clicking on the arrow to the right of the text field (see the Bookmark Manager in <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Preferences" target="_top">Preferences</a> for more details on bookmarks).  Also, you can drag and drop links from web browser to the URL text box.  Once a URL has been specified, click on the Import button to load the network. </p><p><span class="inlinemediaobject"><img src="images/network_import_dialog2_25.png" alt="attachment:network_import_dialog2_25.png"/></span> </p><p>Importing networks from URL addresses has an important caveat. Because Cytoscape determines file type primarily (not exclusively) by file extension, it can have trouble importing networks with URLs that don't end in a human readable file name. If Cytoscape does not recognize a meaningful file name and extension in the URL, it will attempt to guess the type of file based on MIME type. If the MIME type is not recognizable to any of our import handlers, then the import will fail.   </p><p>Another issue for network import is the presence of firewalls, which can affect which files are accessible to a computer. To work around this problem, Cytoscape supports the use of proxy servers. To configure the proxy server, go to Edit → Preferences→ Proxy Server... .  This is further described in the <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Preferences" target="_top">Preferences</a> chapter. </p><p><a id="import-free-format-tables"/> </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Import Free-Format Table Files"/>Import Free-Format Table Files</h3></div></div></div><p>Introduced in version 2.4, Cytoscape now supports the import of networks from delimited text files and Excel workbooks using Edit → Import → Network from Table (Text/MS Excel)... . An interactive GUI allows users to specify parsing options for specified files. The screen provides a preview that shows how the file will be parsed given the current configuration. As the configuration changes, the preview updates automatically. In addition to specifying how the file will be parsed, the user must also choose the columns that represent the Source nodes, the Target nodes, and an optional edge interaction type. </p><p><span class="inlinemediaobject"><img src="images/network_table_import.png" alt="attachment:network_table_import.png"/></span> </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Supported Files"/>Supported Files</h4></div></div></div><p>The "Import Network from Table" function supports delimited text files and single-sheet Microsoft Excel Workbooks. The following is a sample table file: </p><p/><pre class="screen">source  target  interaction  boolean attribute  string attribute        floating point attribute
YJR022W YNR053C pp      TRUE    abcd12371       1.2344543
YER116C YDL013W pp      TRUE    abcd12372       1.2344543
YNL307C YAL038W pp      FALSE   abcd12373       1.2344543
YNL216W YCR012W pd      TRUE    abcd12374       1.2344543
YNL216W YGR254W pd      TRUE    abcd12375       1.2344543
</pre><p>The network table files should contain at least two columns: source nodes and target nodes. The interaction type is optional in this format. Therefore, a minimal network table looks like the following: </p><p/><pre class="screen">YJR022W YNR053C
YER116C YDL013W
YNL307C YAL038W
YNL216W YCR012W
YNL216W YGR254W
</pre><p>One row in a network table file represents an edge and its edge attributes. This means that a network file is considered a combination of network data and edge attributes. A table may contain columns that aren't meant to be edge attributes. In this case, you can choose not to import those columns by clicking on the column header in the preview window. This function is useful when importing a data table like the following (1): </p><pre class="screen">Unique ID A     Unique ID B     Alternative ID A        Alternative ID B        Aliases A       Aliases B       Interaction detection methods   First author surnames   Pubmed IDs      species A       species B       Interactor types        Source database Interaction ID  Interaction labels      Cross-references        Associated Files        Experiment files        Experiment labels       Different techniques    Different Pubmed articles       Different sources       Weight

7205    5747    TRIP6   PTK2    Q15654  Q05397-1        vv|HPRD Currently not available 14688263|15892868(Marcotte)     Mammalia        Homo sapiens    protein|protein HPRD|Marcotte   0       Thyroid hormone receptor interactor 6-FAK-|PTK2-TRIP6   NA(HPRD)|NA(Marcotte)   HPRD/02859_psimi.xml|other/ORIGINAL_DATA_MARCOTTE.txt   vv(HPRD/02859_psimi.xml)|HPRD(other/ORIGINAL_DATA_MARCOTTE.txt) 17651(ExptRef)|Marcotte 2       2       2       2

4174    7311    MCM5    UBA52   P33992  P62987  neighbouring_reaction   Currently not available 15608231(Reactome)      Homo sapiens    Homo sapiens    protein|protein Reactome        1       P33992-P62988   Reaction:68944&lt;-&gt;Reaction:68946(Reactome)|Reaction:68946&lt;-&gt;Reaction:68944(Reactome)     other/ORIGINAL_DATA_MARCOTTE.txt        neighbouring_reaction(other/REACTOMEhomo_sapiens.interactions.txt)      Reactome        1       1       1       1

7040    7040    TGFB1   TGFB1   P01137  P01137  nmr: nuclear magnetic resonance Currently not available 8679613 Homo sapiens    Homo sapiens    protein|protein BIND    2       TGFB1-TGFB1-    72085(BIND)     BIND/bind_taxid9606.1.psi.xml   nmr: nuclear magnetic resonance(BIND/bind_taxid9606.1.psi.xml)  NotAvailable    1       1       1       1
</pre><p>This data file is a tab-delimited text and contains network data (interactions), edge attributes, and node attributes. To import network and edge attributes from this table, you need to choose Unique ID A as source, Unique ID B as target, and Interactor types as interaction type. Then you need to turn off columns used for node attributes (Alternative ID A, species B, etc.). Other columns can be imported as edge attributes.   </p><p>The network import function cannot import node attributes - only edge attributes. To import node attributes from this table, please see the <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Attributes" target="_top">Attributes</a> section of this manual. </p><p>Note (1): This data is taken from the <span class="emphasis"><em>A merged human interactome</em></span> datasets by Andrew Garrow, Yeyejide Adeleye and Guy Warner (Unilever, Safety and Environmental Assurance Center, 12 October 2006). Actual data files are available at <a href="http://www.cytoscape.orghttp://cytoscape.org/cgi-bin/moin.cgi/Data_Sets/" target="_top">http://www.cytoscape.orghttp://cytoscape.org/cgi-bin/moin.cgi/Data_Sets/</a> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Basic Operations"/>Basic Operations</h4></div></div></div><p>To import network text/Excel tables, please follow these steps: </p><div class="orderedlist"><ol type="1"><li><p>Select File → Import → Network from Table (Text/MS Excel)... </p></li><li>Select a table file by clicking on the Select File button.  </li><li>Define the interaction parameters by specifying which columns of data contain the Source Interaction, Target Interaction, and Interaction Type. Setting the Interaction Type as Default Interaction will result in all interactions being given the value pp; this value can be modified in Advanced Options (below). </li><li>(Optional) Define edge attribute columns, if applicable. Network table files can have edge attribute columns in addition to network data. <div class="itemizedlist"><ul type="disc"><li><p>Enable/Disable Attribute Column - By <span class="emphasis"><em>left</em></span>-clicking on a column header in the preview table, you can enable/disable edge attributes.  If the header is checked and entries are blue, the column will be imported as an edge attribute.  For example, the table below shows that columns 1 through 3 will be used as network data, column 4 will not be imported, and columns 5 and 6 will be imported as edge attributes.   </p><div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/network_table_sample.png" alt="attachment:network_table_sample.png"/></span> </p></li></ul></div></li><li><p>Change Attribute Name and Data Types - If you <span class="emphasis"><em>right</em></span>-click on a column header in the preview table, you can modify the attribute name and data type.  For more detail, see "Modify Attribute Name/Type" below. </p></li></ul></div></li><li>Click the Import button.  </li></ol></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Import List of Nodes Without Edges"/>Import List of Nodes Without Edges</h5></div></div></div><p>Table Import feature supports list of nodes without edges.  If you select source column only, it creates a network without interactions.  This feature is usufl with node expansion function available from some web service clients.  Please read the section <span class="emphasis"><em><a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/ImportingNetworksFromWebServices" target="_top">Importing Networks from External Database</a></em></span> for more detail. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Advanced Options"/>Advanced Options</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/network_import_advanced.png" alt="attachment:network_import_advanced.png"/></span> </p><p>You can select several options by checking the Show Text File Import Options checkbox. </p><div class="itemizedlist"><ul type="disc"><li>Delimiter: You can select multiple delimiters for text tables.  By default, Tab and Space are selected as delimiters. </li><li>Preview Options: When you select a network table file, the first 100 entries will be displayed in the Preview panel.  To display more entries, change the value for this option.  If you want to show all entries in the file, select "Show all entries in the file".  You will need to click the Reload button to update the Preview panel. </li><li>Attribute Names <div class="itemizedlist"><ul type="circle"><li>Transfer first line as attribute names: Selecting this option will cause all edge attribute columns to be named according to the first data entry in that column. </li><li>Start Import Row: Set which row of the table to begin importing data from. For example, if you want to skip the first 3 rows in the file, set 4 for this option. </li><li>Comment Line: Rows starting with this character will not be imported.  This option can be used to skip comment lines in text files. </li></ul></div></li><li>Network Import Options: If the Interaction Type is set to Default Interaction, the value here will be used as the interaction type for all edges. </li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Modify Attribute Name/Type"/>Modify Attribute Name/Type</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/network_table_attr_dialog1.png" alt="attachment:network_table_attr_dialog1.png"/></span> </p><p>Attribute names and data types can be modified here. </p><div class="itemizedlist"><ul type="disc"><li>Modify Attribute Name - just enter a new attribute name and click OK. </li><li>Modify Attribute Data Type - The following attribute data types are supported: <div class="itemizedlist"><ul type="circle"><li>String </li><li>Boolean (True/False) </li><li>Integer </li><li>Floating Point </li><li>List of (one of) String/Boolean/Integer/Floating Point    </li></ul></div></li></ul></div><p>Cytoscape has a basic data type detection function that automatically suggests the attribute data type of a column according to its entries. This can be overridden by selecting the appropriate data type from the radio buttons provided. For lists, a global delimiter must be specified (i.e., all cells in the table must use the same delimiter). </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Import Networks from Web Services"/>Import Networks from Web Services</h3></div></div></div><p>From version 2.6.0, Cytoscape has a new feature called <span class="strong"><strong>Web Service Client Manager</strong></span>.  Users can access verious kinds of databases through this function.  Please read <span class="emphasis"><em><span class="strong"><strong><a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/ImportingNetworksFromWebServices" target="_top">Importing Networks and Attributes from External Database</a></strong></span></em></span> for more detail.  </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Edit a New Network"/>Edit a New Network</h3></div></div></div><p>A new, empty network can also be created and nodes and edges manually added. To create an empty network, go to File → New → Network → Empty Network, and then manually add network components using the Editor in CytoPanel 1 (see the <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Network_Editor" target="_top">Editor</a> chapter for more details). </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Supported Network File Formats"/>Supported Network File Formats</h2></div></div></div><p>Cytoscape can read network/pathway files written in the following formats: </p><div class="itemizedlist"><ul type="disc"><li>Simple interaction file (SIF or .sif format) </li><li>Graph Markup Language (GML or .gml format) </li><li>XGMML (extensible graph markup and modelling language). </li><li>SBML </li><li>BioPAX </li><li>PSI-MI Level 1 and 2.5 </li><li>Delimited text </li><li>Excel Workbook (.xls) </li></ul></div><p>The SIF format specifies nodes and interactions only, while other formats store additional information about network layout and allow network data exchange with a variety of other network programs and data sources. Typically, SIF files are used to import interactions when building a network for the first time, since they are easy to create in a text editor or spreadsheet. Once the interactions have been loaded and network layout has been performed, the network may be saved to GML or XGMML format for interaction with other systems. All file types listed (except Excel) are text files and you can edit and view them in a regular text editor.  </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="SIF Format"/>SIF Format</h3></div></div></div><p>The simple interaction format is convenient for building a graph from a list of interactions. It also makes it easy to combine different interaction sets into a larger network, or add new interactions to an existing data set. The main disadvantage is that this format does not include any layout information, forcing Cytoscape to re-compute a new layout of the network each time it is loaded. </p><p>Lines in the SIF file specify a source node, a relationship type (or edge type), and one or more target nodes: </p><p/><pre class="screen">nodeA &lt;relationship type&gt; nodeB
nodeC &lt;relationship type&gt; nodeA
nodeD &lt;relationship type&gt; nodeE nodeF nodeB
nodeG
...
nodeY &lt;relationship type&gt; nodeZ
</pre><p>A more specific example is: </p><p/><pre class="screen">node1 typeA node2
node2 typeB node3 node4 node5
node0
</pre><p>The first line identifies two nodes, called node1 and node2, and a single relationship between node1 and node2 of type typeA. The second line specifies three new nodes, node3, node4, and node5; here "node2" refers to the same node as in the first line. The second line also specifies three relationships, all of type typeB and with node2 as the source, with node3, node4, and node5 as the targets. This second form is simply shorthand for specifying multiple relationships of the same type with the same source node. The third line indicates how to specify a node that has no relationships with other nodes. This form is not needed for nodes that do have relationships, since the specification of the relationship implicitly identifies the nodes as well.  </p><p>Duplicate entries are ignored. Multiple edges between the same nodes must have different edge types.  For example, the following specifies two edges between the same pair of nodes, one of type xx and one of type yy: </p><p/><pre class="screen">node1 xx node2
node1 xx node2
node1 yy node2
</pre><p>Edges connecting a node to itself (self-edges) are also allowed: </p><p/><pre class="screen">node1 xx node1
</pre><p>Every node and edge in Cytoscape has an identifying name, most commonly used with the node and edge data attribute structures. Node names must be unique, as identically named nodes will be treated as identical nodes. The name of each node will be the name in this file by default (unless another string is mapped to display on the node using the visual mapper). This is discussed in the section on <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Visual_Styles" target="_top">visual styles</a>. The name of each edge will be formed from the name of the source and target nodes plus the interaction type: for example, <code class="code">sourceName (edgeType) targetName</code>. </p><p>The tag &lt;relationship type&gt; can be any string.  Whole words or concatenated words may be used to define types of relationships, e.g. geneFusion, cogInference, pullsDown, activates, degrades, inactivates, inhibits, phosphorylates, upRegulates, etc. </p><p>Some common interaction types used in the Systems Biology community are as follows: </p><pre class="screen">  pp .................. protein – protein interaction
  pd .................. protein -&gt; DNA   
  (e.g. transcription factor binding upstream of a regulating gene.)
</pre><p>Some less common interaction types used are: </p><pre class="screen">  pr .................. protein -&gt; reaction
  rc .................. reaction -&gt; compound
  cr .................. compound -&gt; reaction
  gl .................. genetic lethal relationship
  pm .................. protein-metabolite interaction
  mp .................. metabolite-protein interaction
</pre><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Delimiters"/>Delimiters</h4></div></div></div><p>Whitespace (space or tab) is used to delimit the names in the simple interaction file format. However, in some cases spaces are desired in a node name or edge type. The standard is that, if the file contains any tab characters, then tabs are used to delimit the fields and spaces are considered part of the name. If the file contains no tabs, then any spaces are delimiters that separate names (and names cannot contain spaces). </p><p>If your network unexpectedly contains no edges and node names that look like edge names, it probably means your file contains a stray tab that's fooling the parser. On the other hand, if your network has nodes whose names are half of a full name, then you probably meant to use tabs to separate node names with spaces. </p><p>Networks in simple interactions format are often stored in files with a <code class="code">.sif</code> extension, and Cytoscape recognizes this extension when browsing a directory for files of this type. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="GML Format"/>GML Format</h3></div></div></div><p>In contrast to SIF, GML is a rich graph format language supported by many other network visualization packages. The GML file format specification is available at: </p><p><a href="http://www.infosun.fmi.uni-passau.de/Graphlet/GML/" target="_top">http://www.infosun.fmi.uni-passau.de/Graphlet/GML/</a> </p><p>It is generally not necessary to modify the content of a GML file directly. Once a network is built in SIF format and then laid out, the layout is preserved by saving to and loading from GML. Visual attributes specified in a GML file will result in a new visual style named <code class="code">Filename.style</code> when that GML file is loaded. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="XGMML Format"/>XGMML Format</h3></div></div></div><p>XGMML is the XML evolution of GML and is based on the GML definition.  In addition to network data, XGMML contains node/edge/network attributes.  The XGMML file format specification is available at: </p><p><a href="http://www.cs.rpi.edu/~puninj/XGMML/" target="_top">http://www.cs.rpi.edu/~puninj/XGMML/</a> </p><p>XGMML is now preferred to GML because it offers the flexibility associated with all XML document types. If you're unsure about which to use, choose XGMML. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="SBML (Systems Biology Markup Language) Format"/>SBML (Systems Biology Markup Language) Format</h3></div></div></div><p>The Systems Biology Markup Language (SBML) is an XML format to describe biochemical networks.  SBML file format specification is available at: </p><p><a href="http://sbml.org/documents/" target="_top">http://sbml.org/documents/</a>  </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="BioPAX (Biological PAthways eXchange) Format"/>BioPAX (Biological PAthways eXchange) Format</h3></div></div></div><p>BioPAX is an OWL (Web Ontology Language) document designed to exchange biological pathways data.  The complete set of documents for this format is available at:  </p><p><a href="http://www.biopax.org/index.html" target="_top">http://www.biopax.org/index.html</a> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="PSI-MI Format"/>PSI-MI Format</h3></div></div></div><p>The PSI-MI format is a data exchange format for protein-protein interactions.  It is an XML format used to describe PPI and associated data.  PSI-MI XML format specification is available at: </p><p><a href="http://psidev.sourceforge.net/mi/xml/doc/user/" target="_top">http://psidev.sourceforge.net/mi/xml/doc/user/</a> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Delimited Text Table and Excel Workbook"/>Delimited Text Table and Excel Workbook</h3></div></div></div><p>Cytoscape has native support for Microsoft Excel files (.xls) and delimited text files.  The tables in these files can have network data and edge attributes.  Users can specify columns containg source nodes, target nodes, interaction types, and edge attributes during file import.  Some of the other network analysis tools, such as igraph (<a href="http://cneurocvs.rmki.kfki.hu/igraph/" target="_top">http://cneurocvs.rmki.kfki.hu/igraph/</a>), has feature to export graph as simple text files.  Cytoscape can read these text files and build networks from them.  For more detail, please read the Import Free-Format Tables section section of the <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Creating_Networks" target="_top">Creating Networks</a> chapter.  </p><p><span class="inlinemediaobject"><img src="images/huge_network_igraph.png" alt="attachment:huge_network_igraph.png"/></span> </p><p><span class="strong"><strong>Network generated by igraph's Watts-Strogatz small-world model (50k nodes and 250k esges) visualized by Cytoscape:</strong></span>  You can import networks created by other applications using this Table Import feature. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Node Naming Issues in Cytoscape"/>Node Naming Issues in Cytoscape</h3></div></div></div><p>Typically, genes are represented by nodes, and interactions (or other biological relationships) are represented by edges between nodes. For compactness, a gene also represents its corresponding protein. Nodes may also be used to represent compounds and reactions (or anything else) instead of genes. </p><p>If a network of genes or proteins is to be integrated with Gene Ontology (GO) annotation or gene expression data, the gene names must exactly match the names specified in the other data files. We strongly encourage naming genes and proteins by their systematic ORF name or standard accession number; common names may be displayed on the screen for ease of interpretation, so long as these are available to the program in the annotation directory or in a node attribute file. Cytoscape ships with all yeast ORF-to-common name mappings in a synonym table within the <code class="code">annotation/</code> directory. Other organisms will be supported in the future. </p><p>Why do we recommend using standard gene names? All of the external data formats recognized by Cytoscape provide data associated with particular names of particular objects. For example, a network of protein-protein interactions would list the names of the proteins, and the attribute and expression data would likewise be indexed by the name of the object. </p><p>The problem is in connecting data from different data sources that don't necessarily use the same name for the same object. For example, genes are commonly referred to by different names, including a formal "location on the chromosome" identifier and one or more common names that are used by ordinary researchers when talking about that gene. Additionally, database identifiers from every database where the gene is stored may be used to refer to a gene (e.g. protein accession numbers from Swiss-Prot). If one data source uses the formal name while a different data source used a common name or identifier, then Cytoscape must figure out that these two different names really refer to the same biological entity. </p><p>Cytoscape has two strategies for dealing with this naming issue, one simple and one more complex. The simple strategy is to assume that every data source uses the same set of names for every object. If this is the case, then Cytoscape can easily connect all of the different data sources. </p><p>To handle data sources with different sets of names, as is usually the case when manually integrating gene information from different sources, Cytoscape needs a data server that provides synonym information (see the chapter on <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Annotation" target="_top">Annotation</a>). A synonym table gives a canonical name for each object in a given organism and one or more recognized synonyms for that object. Note that the synonym table itself defines which set of names are the "canonical" names. For example, in budding yeast, the ORF names are commonly used as the canonical names. </p><p>If a synonym server is available, then by default Cytoscape will convert every name that appears in a data file to the associated canonical name. Unrecognized names will not be changed. This conversion of names to a common set allows Cytoscape to connect the genes present in different data sources, even if they have different names – as long as those names are recognized by the synonym server. </p><p>For this to work, Cytoscape must also be provided with the species to which the objects belong, since the data server requires the species in order to uniquely identify the object referred to by a particular name. This is usually done in Cytoscape by specifying the species name on the command line with the –P option (<code class="code">cytoscape.sh -P "defaultSpeciesName=Saccharomyces cerevisiae"</code>) or by editing the properties (under Edit → Preferences → Properties...). </p><p>The automatic canonicalization of names can be turned off using the -P option (<code class="code">cytoscape.sh -P canonicalizeName=false"</code>) or by editing the properties (under Edit → Preferences → Properties...). This canonicalization of names currently does not apply to expression data. Expression data should use the same names as the other data sources or use the canonical names as defined by the synonym table. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Node and Edge Attributes"/>Node and Edge Attributes</h2></div></div></div><p>Interaction networks are useful as stand-alone models. However, they are most powerful for answering scientific questions when integrated with additional information. Cytoscape allows the user to add arbitrary node, edge and network information to Cytoscape as node/edge/network attributes. This could include, for example, annotation data on a gene or confidence values in a protein-protein interaction. These attributes can then be visualized in a user-defined way by setting up a mapping from data attributes to visual attributes (colors, shapes, and so on).  The section on <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Visual_Styles" target="_top"> visual styles</a> discusses this in greater detail. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Cytoscape Attribute File Format"/>Cytoscape Attribute File Format</h3></div></div></div><p>Node and edge attribute files are simply formatted: a node attribute file begins with the name of the attribute on the first line (note that it cannot contain spaces). Each following line contains the name of the node, followed by an equals sign and the value of that attribute. Numbers and text strings are the most common attribute types. All values for a given attribute must have the same type. For example: </p><p/><pre class="screen">FunctionalCategory
YAL001C = metabolism
YAR002W = apoptosis
YBL007C = ribosome
</pre><p>An edge attribute file has much the same structure, except that the name of the edge is the source node name, followed by the interaction type in parentheses, followed by the target node name. Directionality counts, so switching the source and target will refer to a different (or perhaps non-existent) edge. The following is an example edge attributes file: </p><p/><pre class="screen">InteractionStrength
YAL001C (pp) YBR043W = 0.82
YMR022W (pd) YDL112C = 0.441
YDL112C (pd) YMR022W = 0.9013
</pre><p>Since Cytoscape treats edge attributes as directional, the second and third edge attribute values refer to two different edges (source and target are reversed, though the nodes involved are the same). </p><p>Each attribute is stored in a separate file. Node and edge attribute files use the same format. Node attribute file names often use the suffix ".noa", while edge attribute file names use the suffix ".eda". Cytoscape recognizes these suffixes when browsing for attribute files. </p><p>Node and edge attributes may be loaded at the command line using the –n and –e options or via the File → Import menu. </p><p>When expression data is loaded using an expression matrix, it is automatically loaded as node attribute data unless explicitly specified otherwise. </p><p>Node and edge attributes are attached to nodes and edges, and so are independent of networks. Attributes for a given node or edge will be applied to all copies of that node or edge in all loaded network files, regardless of whether the attribute file or network file is imported first. </p><p>Note: In order to import network attributes in Cytoscape 2.4, please go to File → Import → Attribute from Table (text/MS Excel)... or encode them in an XGMML network file (see <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Network_Formats" target="_top">Supported File Formats</a> for more details). </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Detailed file format (Advanced users)"/>Detailed file format (Advanced users)</h4></div></div></div><p>Every attribute file has one header line that gives the name of the attribute, and optionally some additional meta-information about that attribute. The format is as follows: </p><p/><pre class="screen">attributeName (class=formal.class.of.value)
</pre><p>The first field is always the attribute name: it cannot contain spaces. If present, the class field defines the formal (package qualified) name of the class of the attribute values. For example, java.lang.String for Strings, java.lang.Double for floating point values, java.lang.Integer for integer values, etc. If the value is actually a list of values, the class should be the type of the objects in the list.  If no class is specified in the header line, Cytoscape will attempt to guess the type from the first value. If the first value contains numbers in a floating point format, Cytoscape will assume java.lang.Double; if the first value contains only numbers with no decimal point, Cytoscape will assume java.lang.Integer; otherwise Cytoscape will assume java.lang.String. Note that the first value can lead Cytoscape astray: for example, </p><p/><pre class="screen">floatingPointAttribute
firstName = 1
secondName = 2.5
</pre><p>In this case, the first value will make Cytoscape think the values should be integers, when in fact they should be floating point numbers. It's safest to explicitly specify the value type to prevent confusion. A better format would be: </p><p/><pre class="screen">floatingPointAttribute (class=Double)
firstName = 1
secondName = 2.5
</pre><p>or </p><p/><pre class="screen">floatingPointAttribute 
firstName = 1.0
secondName = 2.5
</pre><p>Every line past the first line identifies the name of an object (a node in a node attribute file or an edge in a edge attribute file) along with the String representation of the attribute value. The delimiter is always an equals sign; whitespace (spaces and/or tabs) before and after the equals sign is ignored. This means that your names and values can contain whitespace, but object names cannot contain an equals sign and no guarantees are made concerning leading or trailing whitespace.  Object names must be the Node ID or Edge ID as seen in the left-most column of the attribute browser if the attribute is to map to anything.  These names must be reproduced exactly, including case, or they will not match. </p><p>Edge names are all of the form: </p><p/><pre class="screen">sourceName (edgeType) targetName
</pre><p>Specifically, that is  </p><div class="table"><a id="d0e1769"/><p class="title"><b>Table 10. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p>sourceName space openParen edgeType closeParen space targetName</p></td></tr></tbody></table></div></div><br class="table-break"/><p>Note that tabs are not allowed in edge names.  Tabs can be used to separate the edge name from the "=" delimiter, but not within the edge name itself. Also note that this format is different from the specification of interactions in the SIF file format. To be explicit: a SIF entry for the previous interaction would look like </p><p/><pre class="screen">sourceName edgeType targetName
</pre><p>or </p><div class="table"><a id="d0e1785"/><p class="title"><b>Table 11. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p>sourceName whiteSpace edgeType whiteSpace targetName</p></td></tr></tbody></table></div></div><br class="table-break"/><p>To specify lists of values, use the following syntax: </p><p/><pre class="screen">listAttributeName (class=java.lang.String)
firstObjectName = (firstValue::secondValue::thirdValue)
secondObjectName = (onlyOneValue)
</pre><p>This example shows an attribute whose value is defined as a list of text strings. The first object has three strings, and thus three elements in its list, while the second object has a list with only one element. In the case of a list every attribute value uses list syntax (i.e. parentheses), and each element is of the same class. Again, the class will be inferred if it is not specified in the header line. Lists are not supported by the visual mapper and so can’t be mapped to visual attributes. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Newline Feature"/>Newline Feature</h4></div></div></div><p>Sometimes it is desirable to for attributes to include linebreaks, such as node labels that extend over two lines. You can acomplish by inserting <code class="code">
</code> into the attribute value.  For example: </p><p/><pre class="screen">newlineAttr
YJL157C = This is a long\nline for a label.
</pre><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Import Attribute Table Files"/>Import Attribute Table Files</h3></div></div></div><p>As of Cytoscape 2.4, importing delimited text and MS Excel attribute data tables is now supported.  Using this functionality, users can now easily import data that isn't formatted into Cytoscape node or edge attribute file formats (as described above). </p><p><span class="inlinemediaobject"><img src="images/attribute_table_import_main.png" alt="attachment:attribute_table_import_main.png"/></span> </p><p><span class="emphasis"><em><span class="strong"><strong>Sample Attribute Table 1</strong></span></em></span> </p><div class="table"><a id="d0e1832"/><p class="title"><b>Table 12. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/></colgroup><tbody><tr><td><p>Object Key</p></td><td><p>Alias</p></td><td><p>SGD ID</p></td></tr><tr><td><p>AAC3</p></td><td><p>YBR085W|ANC3</p></td><td><p>S000000289</p></td></tr><tr><td><p>AAT2</p></td><td><p>YLR027C|ASP5</p></td><td><p>S000004017</p></td></tr><tr><td><p>BIK1</p></td><td><p>YCL029C|ARM5|PAC14</p></td><td><p>S000000534</p></td></tr></tbody></table></div></div><br class="table-break"/><p>The attribute table file should contain a primary key column and at least one attribute column.  The maximum number of attribute columns is unlimited.  The <span class="emphasis"><em>Alias</em></span> column is an optional feature, as is using the first row of data as attribute names. Alternatively, you can specify each attribute name from the File → Import → Attribute from Table (text/MS Excel)... user interface. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Basic Operation"/>Basic Operation</h4></div></div></div><p>The user interface of the "Import Attributes from Table" window is similar to that of the "Import Network from Table" window. </p><div class="orderedlist"><ol type="1"><li><p>Select File → Import → Attribute from Table (text/MS Excel)... </p></li><li>Select one of the attribute types from the Attributes radio buttons.  Cytoscape can import node, edge, and network attributes. </li><li>Select a data file.  To load a local file, click on the Select File button and choose a data file.  This can be either a text or Excel (.xls) file.  To load a remote file, type the source URL directly into the text box.  To show a preview of the remote file, click the Reload button on the Show Text File Import Options panel. </li><li>(Optional) If the table is not properly delimited in the preview panel, change the delimiter in the Text File Import Options panel.  The default delimiter is the tab.  This step is not necessary for Excel Workbooks. </li><li>By default, the first column is designated as the primary key.  Change the key column if necessary. <div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/attribute_table_import_primary_key.png" alt="attachment:attribute_table_import_primary_key.png"/></span>   </p></li></ul></div></li><li>Click the Import button. </li></ol></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Advanced Options"/>Advanced Options</h4></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Mapping Key Attributes to the primary key"/>Mapping Key Attributes to the primary key</h5></div></div></div><p>Formerly, Cytoscape only supported mapping between node/edge IDs and the primary keys in attribute files.  With the introduction of Cytoscape 2.4, this limitation has been removed, and now both IDs and attributes with primitive data types (string, boolean, floating point, and integer) can be selected as the Key Attribute using the dropdown list provided. Complex attributes such as lists are not supported. </p><p><span class="inlinemediaobject"><img src="images/attribute_table_import_keyattr.png" alt="attachment:attribute_table_import_keyattr.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Aliases"/>Aliases</h5></div></div></div><p>Cytoscape uses a simple mechanism to manage aliases of objects.   Both nodes and edges can have aliases.  If an attribute is loaded as an alias, it is treated as a special attribute called "alias".  This will be used when mapping attributes.  If the primary key and key attribute for an object do not match, Cytoscape will search for a match between aliases and the key attribute.  To define an alias column in the attribute table, just click on the checkboxes to the left of the column name while importing. </p><p><span class="inlinemediaobject"><img src="images/attribute_table_import_alias.png" alt="attachment:attribute_table_import_alias.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Text File Import Options"/>Text File Import Options</h5></div></div></div><p>For more detail on these options, please see the "Import Free-Format Table Files" section of the user manual in the <a href="http://cytoscape.org/cgi-bin/moin.cgi/Cytoscape_User_Manual/Creating_Networks" target="_top">Creating Networks</a> chapter. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Attribute Browser"/>Attribute Browser</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/attribute_browser26.png" alt="attachment:attribute_browser26.png"/></span> </p><p>When Cytoscape is started, the Attribute Browser appears in the bottom CytoPanel. This browser can be hidden and restored using the F5 key or the View → Show/Hide attribute browser menu option. Like other CytoPanels, the browser can be undocked by pressing the little icon in the browser’s top right corner. </p><p>To swap between displaying node, edge, and network attributes use the tabs on the bottom of the panel labelled "Node Attribute Browser", "Edge Attribute Browser", and "Network Attribute Browser". The attribute browser displays attributes belonging to selected nodes and/or edges and the currently selected network. To populate the browser with rows (as pictured above), simply select nodes and/or edges in a loaded network. By default, only the ID of nodes and edges is shown. To display more than just the ID, click the Select Attributes <span class="inlinemediaobject"><img src="images/attributes_select_icon.png" alt="attachment:attributes_select_icon.png"/></span> button and choose the attributes that are to be displayed (select various attributes by clicking on them, and then click elsewhere on the screen to close the attribute list). Each attribute chosen will result in one column in the attribute browser. Most attribute values can be edited by double-clicking an attribute cell; list values cannot be edited, and neither can the ID. Attribute rows in the browser can be sorted alphabetically by specific attribute by clicking on a column heading. A new attribute can be created using the Create New Attribute <span class="inlinemediaobject"><img src="images/attributes_new_icon.png" alt="attachment:attributes_new_icon.png"/></span> button, and must be one of four types – integer, string, real number (floating point), or boolean. Attributes can be deleted using the Delete Attributes <span class="inlinemediaobject"><img src="images/attributes_delete_icon.png" alt="attachment:attributes_delete_icon.png"/></span> button. <span class="strong"><strong>NOTE: Deleting attributes removes them from Cytoscape, not just the attribute browser!</strong></span> To remove attributes from the browser without deleting them, simply unselect the attribute using the Select Attributes <span class="inlinemediaobject"><img src="images/attributes_select_icon.png" alt="attachment:attributes_select_icon.png"/></span> button. </p><p>The right-click menu on the Attribute Browser has several functions, such as exporting attribute information to spreadsheet applications. For example, use the right-click menu to Select All and then Copy the data, and then paste it into a spreadsheet application.  Each attribute browser panel also has a button for importing new attributes: <span class="inlinemediaobject"><img src="images/attributes_import_icon.png" alt="attachment:attributes_import_icon.png"/></span> . </p><p>The Node Attribute Browser panel has additional buttons for loading Gene Expression attribute matrices ( <span class="inlinemediaobject"><img src="images/attributes_gene_expr_icon.png" alt="attachment:attributes_gene_expr_icon.png"/></span> ) as node attributes. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Attribute Batch Editor"/>Attribute Batch Editor</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/attribute_editor26.png" alt="attachment:attribute_editor26.png"/></span> </p><p>From Cytoscape 2.6, Attribute Browser has new <span class="strong"><strong>Attribute Batch Editor</strong></span>.  This enables you to set and modify attribute values at once.  This function changes values for selected nodes or edges.  For example, if you want to create a new attribute called <span class="emphasis"><em>Modules</em></span> and set module names for each group of selected nodes, you can use <span class="emphasis"><em>Set</em></span> command from this editor.  </p></div></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Loading Gene Expression (Attribute Matrix) Data"/>Loading Gene Expression (Attribute Matrix) Data</h2></div></div></div><p>In addition to normal node and edge attribute data, Cytoscape also supports importing gene expression data. Gene expression data are imported using a different file format than normal attributes; however, the resulting attributes are not treated differently by Cytoscape. Gene expression data (like attribute data) can be loaded at any time, but are (generally) only relevant once a network has been loaded. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Data File Format"/>Data File Format</h3></div></div></div><p>Gene expression ratios or values are specified over one or more experiments using a text file. Ratios result from a comparison of two expression measurements (experiment vs. control). Some expression platforms, such as Affymetrix, directly measure expression values, without a comparison. The file consists of a header and a number of space- or tab-delimited fields, one line per gene, with the following format: </p><p/><pre class="screen">Identifier [CommonName] value1 value2 ... valueN [pval1 pval2 ... pvalN]
</pre><p>Brackets [ ] indicate fields that are optional. </p><p>The first field identifies which Cytoscape node the data refers to.  In the simplest case, this is the gene name - exactly as it appears on the network generated by Cytoscape (case sensitive!).  Alternatively, this can be some node attribute that identifies the node uniquely, such as a probeset identifier for commercial microarrays. </p><p>The next field is an optional common name.  It is not used by Cytoscape, and is provided strictly for the user's convenience. With this common name field, the input format is the same as for commonly-used expression data anaysis packages such as SAM (<a href="http://www-stat.stanford.edu/~tibs/SAM/" target="_top">http://www-stat.stanford.edu/~tibs/SAM/</a>). </p><p>The next set of columns represent expression values, one per experiment.  These can be either absolute expression values or fold change ratios.  Each experiment is identified by its experiment name, given in the first line. </p><p>Optionally, significance measures such as P values may be provided.  These values, generated by many microarray data analysis packages, indicate where the level of gene expression or the fold change appears to be greater than random chance.  If you are using significance measures, then your expression file should contain them in a second set of columns after the expression values.  The column names for the expression significance measures need to match those of the expression values <span class="strong"><strong>exactly</strong></span>. </p><p>For example, here is an excerpt from the file galExpData.pvals in the Cytoscape sampleData directory: </p><p/><pre class="screen">GENE COMMON gal1RG gal4RG gal80R gal1RG gal4RG gal80R
YHR051W COX6 -0.034 0.111 -0.304 3.75720e-01 1.56240e-02 7.91340e-06
YHR124W NDT80 -0.090 0.007 -0.348 2.71460e-01 9.64330e-01 3.44760e-01
YKL181W PRS1 -0.167 -0.233 0.112 6.27120e-03 7.89400e-04 1.44060e-01
YGR072W UPF3 0.245 -0.471 0.787 4.10450e-04 7.51780e-04 1.37130e-05
</pre><p>This indicates that there is data for three experiments: gal1RG, gal4RG, and gal80R.  These names appear two times in the header line: the first time gives the expression values, and the second gives the significance measures.  For instance, the second line tells us that in Experiment gal1RG, the gene YHR051W has an expression value of -0.034 with significance measure 3.75720e-01. </p><p>Some variations on this basic format are recognized; see the formal file format specification below for more information. Expression data files commonly have the file extensions ".mrna" or ".pvals", and these file extensions are recognized by Cytoscape when browsing for data files. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="General Procedure"/>General Procedure</h3></div></div></div><p>Load an expression attribute matrix file using File → Import → Attribute/Expression Matrix... to bring up the import window, or by specifying the filename using the -m option at the command line.  If you use the command line input, you must enter your expression data by node ID.  If you use the dialog box, then you can either load expression data by node ID (the default option), or you can select a node attribute to use in assigning your expression data to your Cytoscape nodes. If you do use a node attribute, then (1) the attribute should already be loaded, and (2) the node attribute value must match the first column in your matrix file. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Worked Example"/>Worked Example</h3></div></div></div><p>For the sample network file <code class="code">sampleData/galFiltered.sif</code>: </p><p><span class="strong"><strong>Option A.</strong></span> </p><p>Load a sample gene expression data set by going to File → Import → Attribute/Expression Matrix... . In the resulting window, in the field labeled "Please select an attribute or expression matrix file...",  use the Select button to enter <code class="code">sampleData/galExpData.pvals</code>.  The identifiers used in this file are the same ones used in the network file <code class="code">sampleData/galFiltered.sif</code>, so you do not need to touch the field labeled "Assign values to nodes using...".  A few lines of this file are shown below: </p><p/><pre class="screen">GENE COMMON gal1RG gal4RG gal80R gal1RG gal4RG gal80R
YHR051W COX6 -0.034 0.111 -0.304 3.75720e-01 1.56240e-02 7.91340e-06
YHR124W NDT80 -0.090 0.007 -0.348 2.71460e-01 9.64330e-01 3.44760e-01
YKL181W PRS1 -0.167 -0.233 0.112 6.27120e-03 7.89400e-04 1.44060e-01
</pre><p><span class="strong"><strong>Option B.</strong></span> </p><p>Step 1. After loading the network, load the node attribute file <code class="code">sampleData/gal.probeset.na</code>, using File → Import → Node attributes... .  This file is shown in part below: </p><p/><pre class="screen">Probeset
YHR051W = probeset2
YHR124W = probeset3
YKL181W = probeset4
</pre><p>Step 2. After loading the node attribute file, select the expression data file <code class="code">sampleData.galExpPvals.probeset.pvals</code>, shown in part below: </p><p/><pre class="screen">GENE COMMON gal1RG gal4RG gal80R gal1RG gal4RG gal80R
probeset2 COX6 -0.034 0.111 -0.304 3.75720e-01 1.56240e-02 7.91340e-06
probeset3 NDT80 -0.090 0.007 -0.348 2.71460e-01 9.64330e-01 3.44760e-01
probeset4 PRS1 -0.167 -0.233 0.112 6.27120e-03 7.89400e-04 1.44060e-01
</pre><p>After selecting this file, in the field labeled "Assign values to nodes using...", select Probeset.  You will see that this loads exactly the same expression data as in Case 1, but provides extra flexibility in case the node name cannot be used as an identifier. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Detailed file format (Advanced users)"/>Detailed file format (Advanced users)</h3></div></div></div><p>In all expression data files, any whitespace (spaces and/or tabs) is considered a delimiter between adjacent fields. Every line of text is either the header line or contains all the measurements for a particular gene. No name conversion is applied to expression data files. </p><p>The names given in the first column of the expression data file should match exactly the names used elsewhere (i.e. in SIF or GML files). </p><p>The first line is a header line with one of the following three header formats: </p><p/><pre class="screen">&lt;text&gt; &lt;text&gt; cond1 cond2 ... cond1 cond2 ... [NumSigConds]

&lt;text&gt; &lt;text&gt; cond1 cond2 ...

&lt;tab&gt;&lt;tab&gt;RATIOS&lt;tab&gt;&lt;tab&gt;...LAMBDAS
</pre><p>The first format specifies that both expression ratios and significance values are included in the file. The first two text tokens (in angled brackets) contain names for each gene, such as the formal and common gene names. The <code class="code">condX</code> token set specifies the names of the experimental conditions; these columns will contain ratio values. This list of condition names must then be duplicated exactly, each spelled the same way and in the same order. Optionally, a final column with the title NumSigConds may be present. If present, this column will contain integer values indicating the number of conditions in which each gene had a statistically significant change according to some threshold. </p><p>The second format is similar to the first except that the duplicate column names are omitted, and there is no NumSigConds field. This format specifies data with ratios but no significance values. </p><p>The third format specifies an MTX header, which is a commonly used format. Two tab characters precede the RATIOS token. This token is followed by a number of tabs equal to the number of conditions, followed by the LAMBDAS token. This format specifies both ratios and significance values. </p><p>Each line after the first is a data line with the following format: </p><p/><pre class="screen">FormalGeneName CommonGeneName ratio1 ratio2 ... [lambda1 lambda2 ...] [numSigConds]
</pre><p>The first two tokens are gene names. The names in the first column are the keys used for node name lookup; these names should be the same as the names used elsewhere in Cytoscape (i.e. in the SIF, GML, or XGMML files). Traditionally in the gene expression microarray community, who defined these file formats, the first token is expected to be the formal name of the gene (in systems where there is a formal naming scheme for genes), while the second is expected to be a synonym for the gene commonly used by biologists, although Cytoscape does not make use of the common name column. The next columns contain floating point values for the ratios, followed by columns with the significance values if specified by the header line. The final column, if specified by the header line, should contain an integer giving the number of significant conditions for that gene.  Missing values are not allowed and will confuse the parser. For example, using two consecutive tabs to indicate a missing value will not work; the parser will regard both tabs as a single delimiter and be unable to parse the line correctly. </p><p>Optionally, the last line of the file may be a special footer line with the following format: </p><p/><pre class="screen">NumSigGenes int1 int2 ...
</pre><p>This line specified the number of genes that were significantly differentially expressed in each condition. The first text token must be spelled exactly as shown; the rest of the line should contain one integer value for each experimental condition. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Importing Networks and Attributes from External Databases"/>Importing Networks and Attributes from External Databases</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Web Service Client Manager"/>Web Service Client Manager</h3></div></div></div><p>Cytoscape 2.6.0 has a new feature called <span class="strong"><strong>Web Service Client Manager</strong></span>.  This is a framework to manage various kinds of web service clients in Cytoscape.  By using web service clients, users can access remote datasources easily. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="What is a Web Service?"/>What is a Web Service?</h4></div></div></div><p>A web service is a standardized, platform-independent mechanism for machines to interact over the network.  These days, many major biological databases publishes their system with web service API: </p><div class="itemizedlist"><ul type="disc"><li><p>List of Biological Web Services:  <a href="http://taverna.sourceforge.net/services" target="_top">http://taverna.sourceforge.net/services</a> </p></li><li><p>Web Services at the EBI:  <a href="http://www.ebi.ac.uk/Tools/webservices/" target="_top">http://www.ebi.ac.uk/Tools/webservices/</a> </p></li></ul></div><p>This enables developers to write a program to access these services.  Cytoscape core developer team have developed several sample web service clients using this framework.  Currently, Cytoscape supports the following web services: </p><div class="itemizedlist"><ul type="disc"><li><p><a href="http://www.ebi.ac.uk/intact/site/index.jsf" target="_top">IntAct</a>:  an open source database of protein interaction data, hosted at EMBL-EBI. </p></li><li><p><a href="http://www.pathwaycommons.org" target="_top">Pathway Commons</a>: an open source portal, providing access to multiple integrated data sets, including:  Reactome, IntAct, HPRD, HumanCyc, MINT, the MSKCC Cancer Cell Map, and the NCI/Nature Pathway Interaction database. </p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=gene" target="_top">NCBI Entrez Gene</a>: a public database of genes, including annotation, sequence and interactions. </p></li><li><p><a href="http://www.biomart.org/" target="_top">Biomart</a>: an open source biological database engine.  Useful for ID/Name mapping. </p></li></ul></div><p>All of these clients are available as Plugins and users can install them through Plugin Manager. </p><p>In the following sections, users learn how to import network from extrenal databases. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Getting Started"/>Getting Started</h3></div></div></div><p>To get started, select: File → Import → Network from web services... </p><p><span class="inlinemediaobject"><img src="images/file_import.png" alt="attachment:file_import.png"/></span> </p><div class="table"><a id="d0e2242"/><p class="title"><b>Table 13. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p> <span class="strong"><strong>Tip:</strong></span>  View the <a href="http://cbio.mskcc.org/~cerami/cytoscape/CytoWebServices.mov" target="_top">animation demo</a> for importing networks from web services. </p></td></tr></tbody></table></div></div><br class="table-break"/><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Example #1:  Retrieving Protein-Protein Interaction Networks from IntAct"/>Example #1:  Retrieving Protein-Protein Interaction Networks from IntAct</h3></div></div></div><div class="itemizedlist"><ul type="disc"><li><p>Select: File → Import → Network from web services... </p></li><li><p>From the pull-down menu, select the Int<code class="code"/>Act Web Service Client. </p></li><li>Enter one or more search terms, such as BRCA1 </li><li>Click the Search button. </li></ul></div><p><span class="inlinemediaobject"><img src="images/intact_import.png" alt="attachment:intact_import.png"/></span> </p><p>After confirming the download of interaction data, the network of BRCA1 will be imported and visualized. </p><p><span class="inlinemediaobject"><img src="images/node_context2.png" alt="attachment:node_context2.png"/></span> </p><p><span class="strong"><strong> Tip:  Expanding the Network:</strong></span>  Several of the Cytoscape web services provide additional options in the node context menu.  To access these options, right-click on a node and select "Use Web Services."  For example, in the screenshot to the right, we have loaded the BRCA1 network from IntAct, and have chosen to merge this node's neighbors into the existing network. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Example #2:  Retrieving Protein-Protein Interaction Networks from NCBI Entrez Gene"/>Example #2:  Retrieving Protein-Protein Interaction Networks from NCBI Entrez Gene</h3></div></div></div><p>An entry of NCBI Entrez Gene has a section called <span class="emphasis"><em>Interactions</em></span>.  NCBI web service client uses this section to build networks. </p><div class="itemizedlist"><ul type="disc"><li><p>Select: <span class="strong"><strong>File → Import → Network from web services...</strong></span> </p></li><li><p>From the pull-down menu, select the <span class="strong"><strong>NCBI Web Service Client</strong></span>. </p></li><li><p>Enter free-keywords.  For example, type <span class="emphasis"><em>human muscular dystrophy</em></span>. </p></li><li>Click the Search button. </li></ul></div><p><span class="inlinemediaobject"><img src="images/entrez_import.png" alt="attachment:entrez_import.png"/></span> </p><p><span class="strong"><strong> Network generated from Entrez Gene data:</strong></span> The network above is generated from interaction data matching the keyword <span class="emphasis"><em>human muscular dystrophy</em></span>.  Edge color represents data source type (BIND, BioGRID, or HPRD). </p><p><span class="strong"><strong>Note: since NCBI client extracts interaction data from a huge dataset, it takes a long time (30 sec. - 5 minutes, depends on machine spec. and network connection) to import large set of interactions. </strong></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Example #3:  Retrieving Pathways and Networks from Pathway Commons"/>Example #3:  Retrieving Pathways and Networks from Pathway Commons</h3></div></div></div><div class="itemizedlist"><ul type="disc"><li><p>Select: File → Import → Network from web services... </p></li><li>From the pull-down menu, select the Pathway Commons Web Service Client. </li></ul></div><p>Then, follow the three-step process outlined below: </p><p><span class="inlinemediaobject"><img src="images/3_steps.png" alt="attachment:3_steps.png"/></span> </p><div class="itemizedlist"><ul type="disc"><li>Step 1:  Enter your search term;  for example:  BRCA1 </li><li>Step 2:  Select the protein or small molecule of interest.  Full details regarding each molecule is shown in the bottom left panel. </li><li>Step 3:  Download a specific pathway or interaction network. </li></ul></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Downloading Pathways and Interaction Networks"/>Downloading Pathways and Interaction Networks</h4></div></div></div><p>In Step 3, you can simply double-click on a pathway of interest, or click on the Interaction Networks tab.  The Interaction Networks tab enables you to filter interactions by data source and/or interaction type.  For example, you can choose to restrict your network to direct physical interactions from HPRD and MINT only: </p><p><span class="inlinemediaobject"><img src="images/intxn_filter.png" alt="attachment:intxn_filter.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Pathway Commons Options"/>Pathway Commons Options</h4></div></div></div><p>You can configure access options from the Options tab.  There are two retrieval options: </p><div class="itemizedlist"><ul type="disc"><li>Simplified Binary Model:  Retrieve a simplified binary network, as inferred from the original BioPAX representation.  In this representation, nodes within a network refer to physical entities only, and edges refer to inferred interactions.  </li><li>Full Model:  Retrieve the full model, as stored in the original BioPAX representation.  In this representation, nodes within a network can refer to physical entities and interactions. </li></ul></div><p>By default, the simplified binary model is selected. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Future Directions"/>Future Directions</h3></div></div></div><p>As additional web service clients become available, they will be made available via the Cytoscape Plugin Manager.  Once installed, these web service clients will be centrally accessible via the same steps defined above: </p><div class="itemizedlist"><ul type="disc"><li><p>File → Import → Network from web services... </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Import Attributes from External Database"/>Import Attributes from External Database</h3></div></div></div><p>Some of the web service clients can import attributes from external databases.  <a href="http://cytoscape.org/cgi-bin/moin.cgi/BioMart" target="_top">BioMart</a> client is an example.  You can install it from Plugin Manager. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Example 1: Import Additional ID Sets and Annotations from BioMart"/>Example 1: Import Additional ID Sets and Annotations from BioMart</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/biomart1.png" alt="attachment:biomart1.png"/></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Load a network.  In this example, we use <span class="emphasis"><em>galFiltered.sif</em></span> in <span class="emphasis"><em>sampleData</em></span> directory. </p></li><li><p>File → Import → Import Attributes from BioMart... </p></li><li><p>Select <span class="strong"><strong>Data Source</strong></span>.  Since <span class="emphasis"><em>galFiltered.sif</em></span> is a yeast network, select yeast dataset. </p></li><li><p>For <span class="strong"><strong>Key Attribute</strong></span> section, select <span class="emphasis"><em>ID</em></span> for <span class="strong"><strong>Attribute</strong></span> and <span class="strong"><strong>Data Type</strong></span> should be <span class="emphasis"><em>Ensemble Gene ID</em></span>.  <span class="strong"><strong>Attribute</strong></span> is the list of available attributes in current Cytoscape session and <span class="strong"><strong>Data Type</strong></span> is the type of ID set of the attribute.  In this case, Cytoscape uses <span class="emphasis"><em>ID</em></span> as the key for mapping.  Because the sample network <span class="emphasis"><em>galFiletred.sif</em></span> uses <span class="emphasis"><em>Ensemble Gene ID</em></span> for its node ID, like <span class="emphasis"><em>YOR072W</em></span>, you need to select <span class="emphasis"><em>Ensemble Gene ID</em></span> for <span class="strong"><strong>Data Type</strong></span>.  So you need to know the type of ID set (<span class="emphasis"><em>Entrez Gene ID</em></span>, <span class="emphasis"><em><a href="http://cytoscape.org/cgi-bin/moin.cgi/UniProt" target="_top">UniProt</a> Unified Acc. Number</em></span>, <span class="emphasis"><em>Ensemble Gene ID</em></span>, etc.) of the attribute selected in the <span class="strong"><strong>Attribute</strong></span> box. </p></li><li><p>Select attributes you want to import.  (Note: You cannot select too many attributes at once because BioMart server has maximum number of selectable annotations.) </p></li><li><p>Press <span class="strong"><strong>Import</strong></span>. </p></li><li><p>Now you can see the newly imported attributes on the Attribute Browser.  You may see some attribute names ends with <span class="emphasis"><em>-TOP</em></span> if there are multiple attribute values for a node.  This is an attribute taken from the first entry of the original list attribute. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/biomart2.png" alt="attachment:biomart2.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Example 2: Import Annotations from NCBI Entrez Gene Database"/>Example 2: Import Annotations from NCBI Entrez Gene Database</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/ncbi1.png" alt="attachment:ncbi1.png"/></span> </p><p>NCBI Entrez Gene database () can be used as network data source and annotation repository.  You can use NCBI web service client to import gene annotations from Entrez Gene. </p><div class="itemizedlist"><ul type="disc"><li><p>File → Import → Import attributes from NCBI Entrez Gene... </p></li><li><p><span class="strong"><strong>Data Source</strong></span> is fixed to <span class="emphasis"><em>NCBI Entrez Gene</em></span>. </p></li><li><p><span class="strong"><strong>Data Type</strong></span> is also fixed to <span class="emphasis"><em>Entrez Gene ID</em></span>.  This means Cytoscape attribute selected in <span class="strong"><strong>Attribute</strong></span> list should be Entrez Gene ID. </p></li><li><p>Select <span class="strong"><strong>Annotation Category</strong></span>.  If you select a category, all of the annotation under the category will be imported (i.e., multiple Cytoscape attributes will be created for each category). </p></li><li><p>Press <span class="strong"><strong>Import</strong></span>.  The following is a sample human network (<span class="emphasis"><em>RUAL.subset.sif</em></span>) annotated by NCBI client. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/ncbi2.png" alt="attachment:ncbi2.png"/></span>  </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Use Multiple Services in a Workflow"/>Use Multiple Services in a Workflow</h3></div></div></div><p>Web services are useful when you combine the result from multiple data sources. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Example: Import and Annotate Networks"/>Example: Import and Annotate Networks</h4></div></div></div><div class="itemizedlist"><ul type="disc"><li><p>Import network from <a href="http://cytoscape.org/cgi-bin/moin.cgi/IntAct" target="_top">IntAct</a> using keyword.  In this example, type <span class="emphasis"><em>p53 AND species:mouse</em></span>. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/workflow1.png" alt="attachment:workflow1.png"/></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Import human orthologs from BioMart. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/workflow2.png" alt="attachment:workflow2.png"/></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Show the othologs as the list of Ensemble Gene ID on the Data Panel.  Copy them and use them as the query for <a href="http://cytoscape.org/cgi-bin/moin.cgi/IntAct" target="_top">IntAct</a>. </p></li><li><p>Import <span class="emphasis"><em>Entrez Gene ID</em></span> from BioMart.  Use <span class="emphasis"><em>ensembl</em></span> attribute for the mapping key. </p></li><li>Import annotations from NCBI.  The resulting networks looks like the following: </li></ul></div><p><span class="inlinemediaobject"><img src="images/workflow_final.png" alt="attachment:workflow_final.png"/></span> </p></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Navigation and Layout"/>Navigation and Layout</h2></div></div></div><p><a id="navigation"/> </p><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Basic Network Navigation"/>Basic Network Navigation</h3></div></div></div><p>Cytoscape uses a <a href="http://en.wikipedia.org/wiki/Zooming_User_Interface" target="_top">Zoomable User Interface</a> for navigating and viewing networks.  ZUIs use two mechanisms for navigation: zooming and panning.  Zooming increases or decreases the magnification of a view based on how much or how little a user wants to see.  Panning allows users to move the focus of a screen to different parts of a view.  </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Zoom"/>Zoom</h4></div></div></div><p>Cytoscape provides two mechanisms for zooming: either using mouse gestures or buttons on the toolbar. Use the zooming buttons located on the toolbar to zoom in and out of the interaction network shown in the current network display. Zoom icons are detailed below: </p><p><span class="inlinemediaobject"><img src="images/image049.jpg" alt="attachment:image049.jpg"/></span> </p><p>From Left to Right: </p><div class="itemizedlist"><ul type="disc"><li>Zoom Out </li><li>Zoom In </li><li>Zoom Selected Region </li><li>Zoom Out to Display all of Current Network </li></ul></div><p>You can also zoom in/out by holding down the right mouse button and moving the mouse to the right (zoom in) or left (zoom out). </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Pan"/>Pan</h4></div></div></div><p>You can pan the network image by holding down the middle mouse button and moving the mouse.  You can also pan the image by holding down the left mouse button over the blue box in the Network Overview panel in the lower left hand of the Cytoscape desktop. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Other Mouse Behaviors"/>Other Mouse Behaviors</h3></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Select"/>Select</h4></div></div></div><p>Click the left mouse button on a node or edge to select that object. You can hold down the Shift key to select more than one node/edge or you can hold down the left mouse button and drag the mouse to select groups of nodes/edges. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Context"/>Context</h4></div></div></div><p>Click the right mouse button on a node/edge to launch a context-sensitive menu with additional information about the node/edge. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Automatic Layout Algorithms"/>Automatic Layout Algorithms</h3></div></div></div><p>The Layout menu has an array of features for organizing the network visually according to one of several algorithms, aligning and rotating groups of nodes, and adjusting the size of the network. Most of these features are available from plugins that are packaged with Cytoscape 2.3 and above. Cytoscape layouts have three different sources, which are reflected in the Layout menu. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="yFiles Layouts"/>yFiles Layouts</h4></div></div></div><p><span class="strong"><strong>yFiles</strong></span> layouts are a set of commercial layouts which are provided courtesy of <a href="http://www.yworks.com" target="_top">yWorks</a>.  Due to license restrictions, the detailed parameters for these layouts are not available (there are no <span class="strong"><strong>yFiles</strong></span> entries in the <span class="strong"><strong>Settings...</strong></span> dialog).  The main layout algorithms provided by yFiles are: </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="yFiles Circular Layout"/>yFiles Circular Layout</h5></div></div></div><p>This algorithm produces layouts that emphasize group and tree structures within a network. It partitions the network by analyzing its connectivity structure, and arranges the partitions as separate circles. The circles themselves are arranged in a radial tree layout fashion. This algorithm is available by selecting <span class="strong"><strong>Layout → yFiles → Circular</strong></span>. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/image053.png" alt="attachment:image053.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="yFiles Hierarchical Layout"/>yFiles Hierarchical Layout</h5></div></div></div><p>The hierarchical layout algorithm is good for representing main direction or "flow" within a network. Nodes are placed in hierarchically arranged layers and the ordering of the nodes within each layer is chosen in such a way that minimizes the number of edge crossings. This algorithm is available by selecting <span class="strong"><strong>Layout → yFiles → Hierarchical</strong></span>. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/image055.png" alt="attachment:image055.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="yFiles Organic Layout"/>yFiles Organic Layout</h5></div></div></div><p>The organic layout algorithm is a kind of spring-embedded algorithm that combines elements of the other algorithms to show the clustered structure of a graph. This algorithm is available by selecting <span class="strong"><strong>Layout → yFiles → Organic</strong></span>. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/image057.png" alt="attachment:image057.png"/></span> </p></li></ul></div><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Cytoscape Layouts"/>Cytoscape Layouts</h4></div></div></div><p><span class="strong"><strong>Cytoscape Layouts</strong></span> are those layouts that have been written or integrated by Cytoscape developers.  These layouts are fully integrated with Cytoscape.  All Cytoscape Layouts have the option to operate on only the selected nodes, and all provide a <span class="strong"><strong>Settings...</strong></span> panel to change the parameters of the algorithm.  Most of the Cytoscape layouts also partition the graph before performing the layout.  In addition, many of these layouts include the option to take either node or edge attributes into account.  Some of these layouts are: </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Grid Layout"/>Grid Layout</h5></div></div></div><p>The grid layout is a simple layout the arranges all of the nodes in a square grid.  This is the default layout and is always available as part of the Cytoscape core.  It is available by selecting <span class="strong"><strong>Layout → Cytoscape Layouts → Grid Layout</strong></span>. A sample screen shot is shown below: </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/grid_layout.png" alt="attachment:grid_layout.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Force Directed Layout"/>Force Directed Layout</h5></div></div></div><p>The force-directed layout is a new layout based on the "force-directed" paradigm.  This layout is based on the algorithm implemented as part of the excellent <a href="http://www.prefuse.org/" target="_top">prefuse toolkit</a> provided by Jeff Heer.  The algorithm is very fast and with the right parameters can provide a very pleasing layout.  The <span class="strong"><strong>Force Directed Layout</strong></span> will also accept a numeric edge attribute to use as a weight for the length of the spring, although this will often require more use of the <span class="strong"><strong>Settings...</strong></span> dialog to achieve the best layout.  This algorithm is available by selecting <span class="strong"><strong>Layout → Cytoscape Layouts → Force-Directed Layout → (unweighted)</strong></span> or the edge attribute you want to use as a weight. A sample screen shot showing a portion of the galFiltered network provided in sample data is provided below: </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/force_layout.png" alt="attachment:force_layout.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Spring-Embedded Layout"/>Spring-Embedded Layout</h5></div></div></div><p>The spring-embedded layout is based on a “force-directed” paradigm as implemented by Kamada and Kawai (1988). Network nodes are treated like physical objects that repel each other, such as electrons. The connections between nodes are treated like metal springs attached to the pair of nodes. These springs repel or attract their end points according to a force function. The layout algorithm sets the positions of the nodes in a way that minimizes the sum of forces in the network. This algorithm can be applied to the entire network or a portion of it by selecting the appropriate options from <span class="strong"><strong>Layout → Cytoscape Layouts → Spring Embedded</strong></span>. A sample screen shot is provided below: </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/image051.png" alt="attachment:image051.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Attribute Circle Layout"/>Attribute Circle Layout</h5></div></div></div><p>The attribute circle layout is a quick, useful layout, particularly for small networks, that will locate all of the nodes in the network around a circle.  The node order is determined by a user-selected node attribute.  The result is that all nodes with the same value for that attribute are located together around the circle.  Using <span class="strong"><strong>Layout → Cytoscape Layouts → Attribute Circle Layout →</strong></span> <span class="emphasis"><em><span class="strong"><strong>attribute</strong></span></em></span> to put all nodes around a circle using <span class="emphasis"><em><span class="strong"><strong>attribute</strong></span></em></span> to position them.  The sample screen shot below shows the a subset of the galFiltered network organized by node Degree. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/attribute_circle_layout.png" alt="attachment:attribute_circle_layout.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Group Attributes Layout"/>Group Attributes Layout</h5></div></div></div><p>The group attributes layout is similar to the attribute circle layout described above except that instead of a single circle with all of the nodes, each set of nodes that share the same value for the attribute are laid out in a separate circle.  The same network shown above (subset of galFiltered) is shown below, using <span class="strong"><strong>Layout → Cytoscape Layouts → Attribute Circle Layout → Degree</strong></span>. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/group_by_attributes_layout.png" alt="attachment:group_by_attributes_layout.png"/></span> </p></li></ul></div><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="JGraph Layouts"/>JGraph Layouts</h4></div></div></div><p>Several other alignment algorithms, including a selection from the JGraph project (<a href="http://jgraph.sourceforge.net" target="_top">http://jgraph.sourceforge.net</a>), are also available under the Layout menu. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Layout Parameters"/>Layout Parameters</h4></div></div></div><p>Many layouts have adjustable parameters that are exposed through the Layouts → Settings... menu option.  This will pop up the following dialog, which allows you to choose which layout algorithm settings to adjust.  The settings presented vary by algorithm and only those algorithms that allow access to their parameters will appear in the drop-down menu at the top of the dialog.  Once you've modified a parameter, clicking the "Execute Layout" button will apply the layout.  This be repeated until a useful layout is achieved.  Settings can be saved using the "Save Settings" button. </p><p><span class="inlinemediaobject"><img src="images/layout_settings_dialog.png" alt="attachment:layout_settings_dialog.png"/></span> </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Manual Layout"/>Manual Layout</h3></div></div></div><p>The simplest method to manually organize a network is to click on a node and drag it.  If you select multiple nodes, all of the selected nodes will be moved together. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Rotate"/>Rotate</h4></div></div></div><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/rotate_dialog.png" alt="attachment:rotate_dialog.png"/></span> </p></li></ul></div><p>Selecting the Layout → Rotate option will open the Rotate window in CytoPanel 5. This function will either rotate the entire network or a selected portion of the network.  The image below shows a network with selected nodes rotated. </p><p>Before </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/rotate_networkA.png" alt="attachment:rotate_networkA.png"/></span> </p></li></ul></div><p>After  </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/rotate_networkB.png" alt="attachment:rotate_networkB.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Scale"/>Scale</h4></div></div></div><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/scale_dialog.png" alt="attachment:scale_dialog.png"/></span> </p></li></ul></div><p>Selecting the Layout → Scale option will open the Scale window in CytoPanel 5. This function will scale the position of the entire network or of the selected portion of the network. Note that only the position of the nodes will be scaled, not the node sizes.  Node size can be adjusted using the VizMapper. The image below shows selected nodes scaled. </p><p>Before </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/scale_networkA.png" alt="attachment:scale_networkA.png"/></span> </p></li></ul></div><p>After  </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/scale_networkB.png" alt="attachment:scale_networkB.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Align/Distribute/Stack"/>Align/Distribute/Stack</h4></div></div></div><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/align_distribute_dialog.png" alt="attachment:align_distribute_dialog.png"/></span> </p></li></ul></div><p>Selecting the Layout → Align/Distribute option will open the Align/Distribute/Stack window in CytoPanel 5. The Align buttons provide different options for either vertically or horizontally aligning selected nodes against a line.  The differences are in what part of the node gets aligned, e.g. the center of the node, the top of the node, the left side of the node.  The Distribute buttons evenly distribute  selected nodes between the two most distant nodes along either the vertical or horizontal axis. The differences are again a function what part of the node is used as a reference point for the distribution.  And the Stack buttons vertically or horizontally stack selected nodes with the full complement of alignment options. The table below provides a decription of what each button does. </p><div class="table"><a id="d0e3048"/><p class="title"><b>Table 14. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Button</strong></span></p></td><td><p><span class="strong"><strong>Before</strong></span></p></td><td><p><span class="strong"><strong>After</strong></span></p></td><td><p><span class="strong"><strong>Description of Align Options</strong></span> </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_ALIGN_TOP.gif" alt="attachment:V_ALIGN_TOP.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_begin.png" alt="attachment:small.align_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_vertical_top.png" alt="attachment:small.align_vertical_top.png"/></span></p></td><td><p>Vertical Align Top - The tops of the selected nodes are aligned with the top-most node. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_ALIGN_CENTER.gif" alt="attachment:V_ALIGN_CENTER.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_begin.png" alt="attachment:small.align_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_vertical_center.png" alt="attachment:small.align_vertical_center.png"/></span></p></td><td><p>Vertical Align Center - The centers of the selected nodes are aligned along a line defined by the midpoint between the top and bottom-most nodes. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_ALIGN_BOTTOM.gif" alt="attachment:V_ALIGN_BOTTOM.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_begin.png" alt="attachment:small.align_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_vertical_bottom.png" alt="attachment:small.align_vertical_bottom.png"/></span></p></td><td><p>Vertical Align Bottom - The bottoms of the selected nodes are aligned with the bottom-most node. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_ALIGN_LEFT.gif" alt="attachment:H_ALIGN_LEFT.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_begin.png" alt="attachment:small.align_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_horizontal_left.png" alt="attachment:small.align_horizontal_left.png"/></span></p></td><td><p>Horizontal Align Left - The left hand sides of the selected nodes are aligned with the left-most node. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_ALIGN_CENTER.gif" alt="attachment:H_ALIGN_CENTER.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_begin.png" alt="attachment:small.align_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_horizontal_center.png" alt="attachment:small.align_horizontal_center.png"/></span></p></td><td><p>Horizontal Align Center - The centers of the selected nodes are aligned along a line defined by the midpoint between the left and right-most nodes. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_ALIGN_RIGHT.gif" alt="attachment:H_ALIGN_RIGHT.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_begin.png" alt="attachment:small.align_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.align_horizontal_right.png" alt="attachment:small.align_horizontal_right.png"/></span></p></td><td><p>Horizontal Align Right  - The right hand sides of the selected nodes are aligned with the right-most node. </p></td></tr></tbody></table></div></div><br class="table-break"/><div class="table"><a id="d0e3254"/><p class="title"><b>Table 15. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Button</strong></span></p></td><td><p><span class="strong"><strong>Before</strong></span></p></td><td><p><span class="strong"><strong>After</strong></span></p></td><td><p><span class="strong"><strong>Description of Distribute Options</strong></span> </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_DIST_TOP.gif" alt="attachment:V_DIST_TOP.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_begin_vertical.png" alt="attachment:small.distribute_begin_vertical.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_vertical_top.png" alt="attachment:small.distribute_vertical_top.png"/></span></p></td><td><p>Vertical Distribute Top - The tops of the selected nodes are distributed evenly between the top-most and bottom-most nodes, which should stay stationary. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_DIST_CENTER.gif" alt="attachment:V_DIST_CENTER.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_begin_vertical.png" alt="attachment:small.distribute_begin_vertical.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_vertical_center.png" alt="attachment:small.distribute_vertical_center.png"/></span></p></td><td><p>Vertical Distribute Center - The centers of the selected nodes are distributed evenly between the top-most and bottom-most nodes, which should stay stationary. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_DIST_BOTTOM.gif" alt="attachment:V_DIST_BOTTOM.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_begin_vertical.png" alt="attachment:small.distribute_begin_vertical.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_vertical_bottom.png" alt="attachment:small.distribute_vertical_bottom.png"/></span></p></td><td><p>Vertical Distribute Bottom - The bottoms of the selected nodes are distributed evenly between the top-most and bottom-most nodes, which should stay stationary. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_DIST_LEFT.gif" alt="attachment:H_DIST_LEFT.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_begin_horizontal.png" alt="attachment:small.distribute_begin_horizontal.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_horizontal_left.png" alt="attachment:small.distribute_horizontal_left.png"/></span></p></td><td><p>Horizontal Distribute Left  - The left hand sides of the selected nodes are distributed evenly between the left-most and right-most nodes, which should stay stationary. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_DIST_CENTER.gif" alt="attachment:H_DIST_CENTER.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_begin_horizontal.png" alt="attachment:small.distribute_begin_horizontal.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_horizontal_center.png" alt="attachment:small.distribute_horizontal_center.png"/></span></p></td><td><p>Horizontal Distribute Center - The centers of the selected nodes are distributed evenly between the left-most and right-most nodes, which should stay stationary. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_DIST_RIGHT.gif" alt="attachment:H_DIST_RIGHT.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_begin_horizontal.png" alt="attachment:small.distribute_begin_horizontal.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.distribute_horizontal_right.png" alt="attachment:small.distribute_horizontal_right.png"/></span></p></td><td><p>Horizontal Distribute Right - The right hand sides of the selected nodes are distributed evenly between the left-most and right-most nodes, which should stay stationary. </p></td></tr></tbody></table></div></div><br class="table-break"/><div class="table"><a id="d0e3460"/><p class="title"><b>Table 16. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Button</strong></span></p></td><td><p><span class="strong"><strong>Before</strong></span></p></td><td><p><span class="strong"><strong>After</strong></span></p></td><td><p><span class="strong"><strong>Description of Stack Options</strong></span> </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_STACK_LEFT.gif" alt="attachment:V_STACK_LEFT.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_vertical_begin.png" alt="attachment:small.stack_vertical_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_vertical_left.png" alt="attachment:small.stack_vertical_left.png"/></span></p></td><td><p>Vertical Stack Left - Vertically stacked below top-most node with the left-hand sides of the selected nodes aligned. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_STACK_CENTER.gif" alt="attachment:V_STACK_CENTER.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_vertical_begin.png" alt="attachment:small.stack_vertical_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_vertical_center.png" alt="attachment:small.stack_vertical_center.png"/></span></p></td><td><p>Vertical Stack Center - Vertically stacked below top-most node with the centers of selected nodes aligned. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/V_STACK_RIGHT.gif" alt="attachment:V_STACK_RIGHT.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_vertical_begin.png" alt="attachment:small.stack_vertical_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_vertical_right.png" alt="attachment:small.stack_vertical_right.png"/></span></p></td><td><p>Vertical Stack Right  - Vertically stacked below top-most node with the right-hand sides of the selected nodes aligned. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_STACK_TOP.gif" alt="attachment:H_STACK_TOP.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_horizontal_begin.png" alt="attachment:small.stack_horizontal_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_horizontal_top.png" alt="attachment:small.stack_horizontal_top.png"/></span></p></td><td><p>Horizontal Stack Top - Horizontally stacked to the right of the left-most node with the tops of the selected nodes aligned. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_STACK_CENTER.gif" alt="attachment:H_STACK_CENTER.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_horizontal_begin.png" alt="attachment:small.stack_horizontal_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_horizontal_center.png" alt="attachment:small.stack_horizontal_center.png"/></span></p></td><td><p>Horizontal Stack Center - Horizontally stacked to the right of the left-most node with the centers of selected nodes aligned. </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/H_STACK_BOTTOM.gif" alt="attachment:H_STACK_BOTTOM.gif"/></span> </p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_horizontal_begin.png" alt="attachment:small.stack_horizontal_begin.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/small.stack_horizontal_bottom.png" alt="attachment:small.stack_horizontal_bottom.png"/></span></p></td><td><p>Horizontal Stack Bottom - Horizontal Stack Center - Horizontally stacked to the right of the left-most node with the bottoms of the selected nodes aligned. </p></td></tr></tbody></table></div></div><br class="table-break"/><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Edge Handles"/>Edge Handles</h4></div></div></div><p>This is a little-known feature! If you select an edge and then Ctrl-left-click on the edge, an edge "handle" will appear.  This handle can be used to change the shape of the line. To remove a handle, simply Ctrl-left-click on the handle again. </p><div class="table"><a id="d0e3672"/><p class="title"><b>Table 17. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p>Straight Edge </p></td><td><p>Smooth Edge </p></td></tr><tr><td><p><span class="inlinemediaobject"><img src="images/edge_handles1.png" alt="attachment:edge_handles1.png"/></span></p></td><td><p><span class="inlinemediaobject"><img src="images/edge_handles2.png" alt="attachment:edge_handles2.png"/></span></p></td></tr></tbody></table></div></div><br class="table-break"/><p>Two other options under Select → Edges provide further control: "Smooth Selected Edges" turns an edge consisting of line segments into a smoothed bezier curve, and "Straighten Selected Edges" turns a curved edge back into line segments. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/menu_select_edges.png" alt="attachment:menu_select_edges.png"/></span> </p></li></ul></div><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Node Movement and Placement"/>Node Movement and Placement</h3></div></div></div><p>In addition to the ability to click on a node and drag it to a new position, Cytoscape now has the ability to move nodes using the arrow keys on the keyboard.  By selecting one or more nodes using the mouse and clicking one of the arrow keys (←, →, ↑, ↓) the selected nodes will move one pixel in the chosen direction.  If an arrow key is pressed while holding the Shift key down, the selected nodes will 10 pixels in the chosen direction.  </p><p>Mouse movement has also been enhanced.  If the Shift key is held down while dragging a node, the node will only move horizontally, vertically, or along a 45 degree diagonal. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Visual Styles"/>Visual Styles</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="What is a Visual Style?"/>What is a Visual Style?</h3></div></div></div><p>One of Cytoscape's strengths in network visualization is the ability to allow users to encode any attribute of their data (name, type, degree, weight, expression data, etc.) as a visual property (such as color, size, transparency, or font type). A set of these encoded or mapped attributes is called a <span class="emphasis"><em><span class="strong"><strong>Visual Style</strong></span></em></span> and can be created or edited using the Cytoscape <span class="emphasis"><em><span class="strong"><strong>VizMapper</strong></span></em></span>. With the VizMapper, the visual appearance of your network is easily customized. For example, you can: </p><div class="itemizedlist"><ul type="disc"><li>Specify a default color and shape for all nodes. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/DefaultColorAndShape.png" alt="attachment:DefaultColorAndShape.png"/></span> </p></li></ul></div></li><li>Use specific line types to indicate different types of interactions. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/LineTypes.png" alt="attachment:LineTypes.png"/></span> </p></li></ul></div></li><li>Encode specific physical entities as different node shapes. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/NodeShapeMapping.png" alt="attachment:NodeShapeMapping.png"/></span> </p></li></ul></div></li><li>Set node sizes based on the degree of connectivity of the nodes.  You can visually see the hub of a network... <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/DegreeSize.png" alt="attachment:DegreeSize.png"/></span> </p></li></ul></div></li><li>...or, set the font size of the node labels instead. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/DegreeLabelSize.png" alt="attachment:DegreeLabelSize.png"/></span> </p></li></ul></div></li><li>Set node widths and heights based on label size. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/LabelWidthAndHeight.png" alt="attachment:LabelWidthAndHeight.png"/></span> </p></li></ul></div></li><li>Visualize gene expression data along a color gradient. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/ColorGradient.png" alt="attachment:ColorGradient.png"/></span> </p></li></ul></div></li><li>Control edge transparency (opacity) using edge weights. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/OpacityForEdges.png" alt="attachment:OpacityForEdges.png"/></span> </p></li></ul></div></li><li>Control multiple edge visual properties using edge score. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/MultipleEdgeMapping.png" alt="attachment:MultipleEdgeMapping.png"/></span> </p></li></ul></div></li><li>Browse extremely-dense networks by controlling the opacity of nodes. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/OpacityForNodesAndEdges.png" alt="attachment:OpacityForNodesAndEdges.png"/></span> </p></li></ul></div></li><li>Show module locations in a large network. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/ModuleLocations.png" alt="attachment:ModuleLocations.png"/></span> </p></li></ul></div></li><li>Overlay a subnetwork on huge interactome using opacity and color control. <div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/Overlay.png" alt="attachment:Overlay.png"/></span> </p></li></ul></div></li></ul></div><p>Cytoscape 2.6.0 and later has additional sample Visual Styles.  You can try those samples to understand how Visual Styles change appearence of a network.  The following is a list of network views based on sample styles applied to <span class="emphasis"><em>galFiltered.sif</em></span> network : </p><p><span class="inlinemediaobject"><img src="images/default_style.png" alt="attachment:default_style.png"/></span>  <span class="inlinemediaobject"><img src="images/metro_style.png" alt="attachment:metro_style.png"/></span> <span class="inlinemediaobject"><img src="images/solid_style.png" alt="attachment:solid_style.png"/></span>  <span class="inlinemediaobject"><img src="images/ripple_style.png" alt="attachment:ripple_style.png"/></span> <span class="inlinemediaobject"><img src="images/skeleton_style.png" alt="attachment:skeleton_style.png"/></span>  <span class="inlinemediaobject"><img src="images/universe_style.png" alt="attachment:universe_style.png"/></span> </p><p>The VizMapper can be accessed by going to View → Open VizMapper or by clicking on the VizMapper icon <span class="inlinemediaobject"><img src="images/VizMapIcon.png" alt="attachment:VizMapIcon.png"/></span>. Also, starting in Cytoscape 2.5, direct access to the VizMapper is provided via a tab on the Control Panel at the left-hand side of the screen (formerly known as CytoPanel 1). </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Introduction to the VizMapper User Interface"/>Introduction to the VizMapper User Interface</h3></div></div></div><p>As of Cytoscape 2.5, the VizMapper has undergone a complete interface redesign.  There are three types of components in the new VizMapper:  </p><div class="orderedlist"><ol type="1"><li>Main Panel <div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/VizMapperMainPanel.png" alt="attachment:VizMapperMainPanel.png"/></span> </p></li><li>This panel allows you to create/delete/view/switch between different visual styles using the Current Visual Style options. The Visual Mapping Browser at the bottom displays the mapping details for a given visual style and is used to edit these details as well. </li></ul></div></li><li>Default Appearance Editor <div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/DefaultEditorPanel.png" alt="attachment:DefaultEditorPanel.png"/></span> </p></li><li>Clicking on the section labelled "Defaults" on the Main Panel will bring up this editor, which allows users to visually edit the default appearance of nodes and edges for the selected visual style. </li></ul></div></li><li>Continuous Editors <div class="itemizedlist"><ul type="disc"><li>These are editors for continuous mapping, which is a mapping from numerical value to visual attributes.  They are accessed through the Visual Mapping Browser on the Main Panel. Using these windows, users can edit continuous mapping more intuitively. </li><li>Color Gradient Editor </li><li>Continuous-to-Discrete Editor </li><li>Continuous-to-Continuous Editor </li></ul></div></li></ol></div><p>These editors will be discussed in further detail below. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Introduction to Visual Styles"/>Introduction to Visual Styles</h3></div></div></div><p>The Cytoscape distribution includes several predefined visual styles to get you started. To demonstrate these styles, try out the following example: </p><p><span class="strong"><strong>Step 1. Load some sample data</strong></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Load a sample session file: From the main menu, select <span class="emphasis"><em><span class="strong"><strong>File → Open</strong></span></em></span>, and select the file <span class="emphasis"><em>sampleData/galFiltered.cys</em></span>. </p></li><li><p>The session file includes a network, some annotations, and sample visual styles.  By default, <span class="strong"><strong>galFiltered Style</strong></span> is selected.  Gene expression values for each node will be colored along a color gradient between red and green (where red represents a low expression ratio and green represents a high expression ratio, using thresholds set for the gal4RGexp experiment bundled with Cytoscape in the <span class="emphasis"><em>sampleData/galExpData.pvals</em></span> file).  Also, node size is mapped onto the degree (number of edges connected to the node) and you can see the hubs of the network as larger nodes.  See the sample screenshot below: </p><p><span class="inlinemediaobject"><img src="images/galFilteredSessionDefault.png" alt="attachment:galFilteredSessionDefault.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 2. Switch between different Visual Styles</strong></span> </p><p>You can change visual styles by making a selection from the Current Visual Style dropdown list (found at the top of the VizMapper Main Panel). </p><p>For example, if you select <span class="strong"><strong>Sample1</strong></span>, a new visual style will be applied to your network, and you will see a white background and round blue nodes. Additionally, if you zoom in closer, you can see that protein-DNA interactions (specified with the label "pd") are drawn with dashed red edges, whereas protein-protein interactions (specified with the label "pp") are drawn with a solid light blue edge (see sample screenshot below). </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/VizMapperSample1Style26.png" alt="attachment:VizMapperSample1Style26.png"/></span> </p></li></ul></div><p>Finally, if you select <span class="strong"><strong>Solid</strong></span>,  you can see the graphics below: </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/VizMapperSolidStyle.png" alt="attachment:VizMapperSolidStyle.png"/></span> </p></li></ul></div><p>This Visual Style does not have mappings except node/edge labels, but you can modify the network graphics by editing <span class="emphasis"><em>Default View</em></span>. </p><p>Additional sample styles are available as <span class="emphasis"><em>sampleStyles.props</em></span> file in the <span class="emphasis"><em><a href="http://cytoscape.org/cgi-bin/moin.cgi/SampleData" target="_top">SampleData</a></em></span> directory.  You can import the sample file from <span class="emphasis"><em><span class="strong"><strong>File → Import → Vizmap Property File</strong></span></em></span>. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Visual Attributes, Graph Attributes and Visual Mappers"/>Visual Attributes, Graph Attributes and Visual Mappers</h3></div></div></div><p>The Cytoscape VizMapper uses three core concepts: </p><div class="itemizedlist"><ul type="disc"><li><p>A <span class="emphasis"><em>visual attribute</em></span> is any visual setting that can be applied to your network. For example, you can change all nodes from circles to squares by changing the node shape visual attribute. </p></li><li><p>A <span class="emphasis"><em>network attribute</em></span> is any data attribute associated with a node or an edge. For example, each edge in a network may be associated with a label, such as “pd” (protein-DNA interactions), or “pp” (protein-protein interactions). </p></li><li><p>A <span class="emphasis"><em>visual mapper</em></span> maps network attributes to visual attributes. For example, a visual mapper can map all protein-DNA interactions to the color blue, and all protein-protein interactions to the color red. </p></li></ul></div><p>Cytoscape allows a wide variety of visual attributes to be controlled. These are summarized in the tables below. </p><div class="table"><a id="d0e4108"/><p class="title"><b>Table 18. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Visual Attributes Associated with Nodes</strong></span> </p></td></tr><tr><td><p>Node Color </p></td></tr><tr><td><p>Node Opacity </p></td></tr><tr><td><p>Node Border Color </p></td></tr><tr><td><p>Node Border Opacity </p></td></tr><tr><td><p>Node Border Line Style. Solid and dashed lines are supported. </p></td></tr><tr><td><p>Node Border Line Width </p></td></tr><tr><td><p>Node Shape. The following options are available: </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/NewVizMapperNodeShape.png" alt="attachment:NewVizMapperNodeShape.png"/></span> </p></td></tr><tr><td><p>Node Size: the width and height of each node. </p></td></tr><tr><td><p>Node Label: the text label for each node. </p></td></tr><tr><td><p>Node Label Color </p></td></tr><tr><td><p>Node Label Opacity </p></td></tr><tr><td><p>Node Label Position: the position of the label relative to the node. </p></td></tr><tr><td><p>Node Font: node label font and size. </p></td></tr></tbody></table></div></div><br class="table-break"/><div class="table"><a id="d0e4182"/><p class="title"><b>Table 19. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Visual Attributes Associated with Edges</strong></span> </p></td></tr><tr><td><p>Edge Color </p></td></tr><tr><td><p>Edge Opacity </p></td></tr><tr><td><p>Edge Line Style. Solid or dashed lines are supported.</p></td></tr><tr><td><p>Edge Line Width </p></td></tr><tr><td><p>Edge Source and Target Arrow Shape: The following options are available: </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/NewVizMapperArrowType.png" alt="attachment:NewVizMapperArrowType.png"/></span> </p></td></tr><tr><td><p>Edge Source and Target Arrow Color </p></td></tr><tr><td><p>Edge Source and Target Arrow Opacity </p></td></tr><tr><td><p>Edge Label: the text label for each edge. </p></td></tr><tr><td><p>Edge Label Color </p></td></tr><tr><td><p>Edge Label Opacity </p></td></tr><tr><td><p>Edge Font: edge label font and size. </p></td></tr></tbody></table></div></div><br class="table-break"/><div class="table"><a id="d0e4248"/><p class="title"><b>Table 20. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Global Visual Properties</strong></span> </p></td></tr><tr><td><p>Background Color </p></td></tr><tr><td><p>Selected Node Color </p></td></tr><tr><td><p>Selected Edge Color </p></td></tr></tbody></table></div></div><br class="table-break"/><p>For each visual attribute, you can specify a default value or define a dynamic visual mapping. Cytoscape currently supports three different types of visual mappers: </p><div class="orderedlist"><ol type="1"><li><p><span class="strong"><strong>Passthrough Mapper</strong></span> </p><div class="itemizedlist"><ul type="disc"><li>The values of network attributes are passed directly through to visual attributes. A passthrough mapper is only used to specify node/edge labels. For example, a passthrough mapper can label all nodes with their common gene names. </li></ul></div></li><li><p><span class="strong"><strong>Discrete Mapper</strong></span> </p><div class="itemizedlist"><ul type="disc"><li>Discrete network attributes are mapped to discrete visual attributes. For example, a discrete mapper can map all protein-protein interactions to the color blue. </li></ul></div></li><li><p><span class="strong"><strong>Continuous Mapper</strong></span> </p><div class="itemizedlist"><ul type="disc"><li>Continuous graph attributes are mapped to visual attributes. Depending on the visual attribute, there are three kinds of continuous mappers: <div class="orderedlist"><ol type="i"><li><p><span class="strong"><strong>Continuous-to-Continuous Mapper</strong></span>: for example, you can map a continuous numerical value to a node size.  </p></li><li><p><span class="strong"><strong>Color Gradient Mapper</strong></span>: This is a special case of continuous-to-continuous mapping.  Continuous numerical values are mapped to a color gradient. </p></li><li><p><span class="strong"><strong>Continuous-to-Discrete Mapper</strong></span>: for example, all values below 0 are mapped to square nodes, and all values above 0 are mapped to circular nodes.  </p></li></ol></div></li><li>However, note that there is no way to smoothly morph between circular nodes and square nodes. </li></ul></div></li></ol></div><p>The table below shows visual mapper support for each visual property. </p><p><span class="strong"><strong>Legend</strong></span> </p><div class="table"><a id="d0e4322"/><p class="title"><b>Table 21. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Symbol</strong></span> </p></td><td><p><span class="strong"><strong>Description</strong></span> </p></td></tr><tr><td><p> - </p></td><td><p>Mapping is not supported for the specified visual property. </p></td></tr><tr><td><p> X </p></td><td><p>Mapping is fully supported for the specified visual property. </p></td></tr><tr><td><p> o </p></td><td><p>Mapping is partially supported for the specified visual property. Support for “continuous to continuous” mapping is not supported. </p></td></tr></tbody></table></div></div><br class="table-break"/><p><span class="strong"><strong>Node Visual Mappings</strong></span> </p><div class="table"><a id="d0e4364"/><p class="title"><b>Table 22. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Node Visual Properties</strong></span> </p></td><td><p><span class="strong"><strong>Passthrough Mapper</strong></span> </p></td><td><p><span class="strong"><strong>Discrete Mapper</strong></span> </p></td><td><p><span class="strong"><strong>Continuous Mapper</strong></span> </p></td><td class="auto-generated"> </td></tr><tr><td><p>Color</p></td><td><p> Node Color </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td></tr><tr><td><p> Node Opacity </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Border Color </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Border Opacity </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Label Color </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Label Opacity </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Numeric </p></td><td><p> Node Size </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td></tr><tr><td><p> Node Font Size </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Line Width </p></td><td><p> - </p></td><td><p> X </p></td><td><p> X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Other</p></td><td><p> Node Border Type </p></td><td><p>  - </p></td><td><p> X </p></td><td><p> o </p></td></tr><tr><td><p> Node Shape </p></td><td><p> - </p></td><td><p> X </p></td><td><p> o </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Label </p></td><td><p> X </p></td><td><p> X </p></td><td><p> o </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Tooltip </p></td><td><p> X </p></td><td><p> X </p></td><td><p> o </p></td><td class="auto-generated"> </td></tr><tr><td><p> Node Font Family </p></td><td><p>  - </p></td><td><p> X </p></td><td><p> o </p></td><td class="auto-generated"> </td></tr></tbody></table></div></div><br class="table-break"/><p><span class="strong"><strong>Edge Visual Mappings</strong></span> </p><div class="table"><a id="d0e4589"/><p class="title"><b>Table 23. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/><col/><col/></colgroup><tbody><tr><td><p><span class="strong"><strong>Edge Properties</strong></span> </p></td><td><p><span class="strong"><strong>Passthrough Mapper</strong></span> </p></td><td><p><span class="strong"><strong>Discrete Mapper</strong></span> </p></td><td><p><span class="strong"><strong>Continuous Mapper</strong></span> </p></td><td class="auto-generated"> </td></tr><tr><td><p>Color</p></td><td><p>Edge Color </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td></tr><tr><td><p>Edge Opacity </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Edge Target Arrow Color </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Edge Source Arrow Color </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Edge Target Arrow Opacity </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Edge Source Arrow Opacity </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Edge Label Color </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Edge Label Opacity </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Numeric</p></td><td><p> Edge Line Width </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td></tr><tr><td><p>Edge Font Size </p></td><td><p>- </p></td><td><p>X </p></td><td><p>X </p></td><td class="auto-generated"> </td></tr><tr><td><p>Other </p></td><td><p> Edge Line Type </p></td><td><p> - </p></td><td><p>X </p></td><td><p>o </p></td></tr><tr><td><p> Edge Source Arrow Shape </p></td><td><p> - </p></td><td><p> X </p></td><td><p> o </p></td><td class="auto-generated"> </td></tr><tr><td><p> Edge Target Arrow Shape </p></td><td><p> - </p></td><td><p>X </p></td><td><p>o </p></td><td class="auto-generated"> </td></tr><tr><td><p> Edge Label </p></td><td><p> X </p></td><td><p> X </p></td><td><p>o </p></td><td class="auto-generated"> </td></tr><tr><td><p> Edge Tooltip </p></td><td><p> X </p></td><td><p> X </p></td><td><p>o </p></td><td class="auto-generated"> </td></tr><tr><td><p> Edge Font Family </p></td><td><p> - </p></td><td><p>X </p></td><td><p>o </p></td><td class="auto-generated"> </td></tr></tbody></table></div></div><br class="table-break"/><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Visual Styles Tutorials"/>Visual Styles Tutorials</h3></div></div></div><p>The following tutorials demonstrate some of the basic VizMapper features. Each tutorial is independent of the others. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Tutorial 1: Create a Basic Visual Style and Set Default Values"/>Tutorial 1: Create a Basic Visual Style and Set Default Values</h4></div></div></div><p>The goal of this tutorial is to learn how to create a new Visual Style and set some default values. </p><p><span class="strong"><strong>Step 1. Load a sample network.</strong></span> From the main menu, select File → Import → Network (Multiple file types), and select <code class="code">sampleData/galFiltered.sif</code>. </p><p><span class="strong"><strong>Step 2. Open the VizMapper.</strong></span> Select the View → Open VizMapper menu option, or select the VizMapper icon in the main button bar, or click on the VizMapper tab in the Control Panel at the left of the screen. You will now see a VizMapper Main Panel, as shown below. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/NewVizMapper.png" alt="attachment:NewVizMapper.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 3. Create a new visual style.</strong></span> Click the Options <span class="inlinemediaobject"><img src="images/VizMapOptionIcon.png" alt="attachment:VizMapOptionIcon.png"/></span> button, and select <span class="emphasis"><em>Create new visual style...</em></span> Then enter a name for your new visual style when prompted.  You will see an empty visual style in the VizMapper Main Panel, as shown below. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/EmptyVisualStyle.png" alt="attachment:EmptyVisualStyle.png"/></span> </p></li></ul></div><p>Since no mapping is set up yet, all visual attributes are listed in the Unused Properties category.  From this panel, you can create node/edge mappings for all visual properties. </p><p><span class="strong"><strong>Step 4. Edit default values.</strong></span> Open the Default Appearance Editor by clicking on the Defaults graphics window (shown below) in the VizMapper Main Panel.  </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/InitialDefaultEditor.png" alt="attachment:InitialDefaultEditor.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 5. Change the default node shape.</strong></span> To set the default node shape to triangles, click "Node Shape" in the Default Visual Properties list. A list of available node shapes will be shown. Click on the Triangle icon and then click the Apply button.  The Default Appearance Editor will be automatically updated. You can edit other default values by clicking on visual attribute names on the list. In the example shown below, the node shape is set to Triangle, while the node color is set to blue.  </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/TriangleDefaultEditor.png" alt="attachment:TriangleDefaultEditor.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 6. Apply your settings.</strong></span> When you finish editing, click the Apply button at the bottom of the editor.  Your new Visual Style will be applied to the current network, as shown below. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/Tut1GalFiltered.png" alt="attachment:Tut1GalFiltered.png"/></span> </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Tutorial 2: Creating a New Visual Style with a Discrete Mapper"/>Tutorial 2: Creating a New Visual Style with a Discrete Mapper</h4></div></div></div><p>The following tutorial demonstrates how to create a new visual style using a discrete mapper. The goal is to draw protein-DNA interactions as dashed blue lines, and protein-protein interactions as solid red lines. </p><p><span class="strong"><strong>Step 1. Load a sample network.</strong></span> From the main menu, select File → Import → Network (Multiple file types), and select <code class="code">sampleData/galFiltered.sif</code>. </p><p><span class="strong"><strong>Step 2. Open the VizMapper.</strong></span> Select the View → Open VizMapper menu option, or select the VizMapper icon in the main button bar, or click on the VizMapper tab in the Control Panel at the left of the screen. </p><p><span class="strong"><strong>Step 3. Create a new visual style.</strong></span> Click the Options <span class="inlinemediaobject"><img src="images/VizMapOptionIcon.png" alt="attachment:VizMapOptionIcon.png"/></span> button, and select <span class="emphasis"><em>Create new visual style...</em></span> Name your new style “Tutorial VS2”. </p><p><span class="strong"><strong>Step 4. Choose a visual attribute.</strong></span> Double click the Edge Color entry listed in Unused Properties.  Edge Color will now appear at the top of the list, under the Edge Visual Mapping category (as shown below). </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/EdgeMapping1.png" alt="attachment:EdgeMapping1.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 5. Choose a network attribute.</strong></span> Click on the cell to the right of the Edge Color entry and select "interaction" from the dropdown list that appears. </p><p><span class="strong"><strong>Step 6. Choose a mapping type.</strong></span> Set the Discrete Mapper option as the Mapping Type.  All available attribute values for "interaction" will be displayed, as shown below. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/EdgeMapping2.png" alt="attachment:EdgeMapping2.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 7. Set the mapping relationship.</strong></span> Click the empty cell next to "pd" (protein-DNA interactions).  On the right side of the cell, <span class="strong"><strong>...</strong></span> and <span class="strong"><strong>X</strong></span> buttons will appear. Click on the <span class="strong"><strong>...</strong></span> button. A popup window will appear; select blue, and the change will immediately appear on the network window. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/CellEditor1.png" alt="attachment:CellEditor1.png"/></span> </p></li></ul></div><p>Repeat step 7 for "pp" (protein-protein interactions), but select red as the edge color. Then repeat steps 4 through 7 for the Edge Line Style attribute. You can select the correct line style (dashed or solid) from the dropdown list. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/EdgeMapping3.png" alt="attachment:EdgeMapping3.png"/></span> </p></li></ul></div><p>Now your network should show "pd" interactions as dashed blue lines and "pp" interactions as solid red lines. A sample screenshot is provided below. </p><p><span class="inlinemediaobject"><img src="images/NewVizMapperInteractionsRedBlue.png" alt="attachment:NewVizMapperInteractionsRedBlue.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Tutorial 3: Visualizing Expression Data on a Network"/>Tutorial 3: Visualizing Expression Data on a Network</h4></div></div></div><p>The following tutorial demonstrates how to create a new visual style using a continuous mapper. The goal is to superimpose gene expression data onto a network and display gene expression values along a color gradient. </p><p><span class="strong"><strong>Step 1. Load a sample network.</strong></span> From the main menu, select File → Import → Network (Multiple file types), and select <code class="code">sampleData/galFiltered.sif</code>. </p><p><span class="strong"><strong> Step 2. Load sample expression data.</strong></span> From the main menu, select File → Import → Attribute/Expression Matrix, and select <code class="code">sampleData/galExpData/pvals</code>. </p><p><span class="strong"><strong>Step 3. Open the VizMapper.</strong></span> Select the View → Open VizMapper menu option, or select the VizMapper icon in the main button bar, or click on the VizMapper tab in the Control Panel at the left of the screen. </p><p><span class="strong"><strong>Step 4. Create a new visual style.</strong></span> Click the Options <span class="inlinemediaobject"><img src="images/VizMapOptionIcon.png" alt="attachment:VizMapOptionIcon.png"/></span> button, and select <span class="emphasis"><em>Create new visual style...</em></span> Name your new style “Tutorial VS3”. </p><p><span class="strong"><strong>Step 5. Choose a visual attribute.</strong></span> Double click the Node Color entry listed in Unused Properties.  Node Color will now appear at the top of the list, under the Node Visual Mapping category. </p><p><span class="strong"><strong>Step 6. Choose a network attribute.</strong></span> Click on the cell to the right of the Node Color entry and select "gal1RGexp" from the dropdown list that appears. </p><p><span class="strong"><strong>Step 7. Choose a mapping type.</strong></span> Set the Continuous Mapping option as the Mapping Type.  This automatically creates a default mapping </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/DefaultColorGradient.png" alt="attachment:DefaultColorGradient.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 8. Define the points where colors will change.</strong></span> Double-click on the black-and-white gradient rectangle next to Graphical View to open the Color Gradient Mapper. Click and drag one point to -1, or type the value in the Range Setting box. Set the second point to 2. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/DefaultColorGradientEditor.png" alt="attachment:DefaultColorGradientEditor.png"/></span> </p></li></ul></div><p><span class="strong"><strong>Step 9. Define the colors between points.</strong></span> Double-click on the leftmost triangle (facing left) and a color palette will appear. Choose a shade of yellow and click OK. Double-click on the triangle at -1 and set the color white.  For triangle at 2, set its color to red. Set the rightmost triangle to black. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/RedYellowColorGradient2.png" alt="attachment:RedYellowColorGradient2.png"/></span> </p></li></ul></div><p>The color gradients will immediately appear in the network window. All nodes with a gal1RGexp value less than –1 will be set to yellow, and all nodes with a gal1RGExp value greater than 2 will be black. Additionally, all values between –1 and 2 will be painted with a white/red color gradient. A sample screenshot is below. </p><p><span class="inlinemediaobject"><img src="images/VizMapperExpRedBlack.png" alt="attachment:VizMapperExpRedBlack.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Tutorial 4: How to Use Utilities for Discrete Mappers"/>Tutorial 4: How to Use Utilities for Discrete Mappers</h4></div></div></div><p>The following tutorial demonstrates new features in Cytoscape 2.5.  The new <a href="http://cytoscape.org/cgi-bin/moin.cgi/VizMapper" target="_top">VizMapper</a> user interface has some utilities to help users editing discrete mappings.  The goal of this section is learning how to set and adjust values for discrete mappings automatically.  </p><div class="orderedlist"><ol type="1"><li><p>Load a sample network: From the main menu, select <span class="underline">File → Import → Network</span>, and select sampleData/galFiltered.sif. </p></li><li><p>Apply layout to the network: From the main menu, select <span class="underline">Layout → Cytoscape Layouts → Degree Sorted Circle Layout</span>.  This layout algorithm sort nodes in a circle by degree of the nodes.  Degrees will be stored as node attribute names <span class="emphasis"><em>Degree</em></span> after you applied this algorithm. </p></li><li><p>Click the <a href="http://cytoscape.org/cgi-bin/moin.cgi/VizMap" target="_top">VizMap</a> <span class="inlinemediaobject"><img src="images/VizMapIcon.png" alt="attachment:VizMapIcon.png"/></span> button on the tool bar. </p></li><li><p>Click <span class="emphasis"><em>Defaults</em></span> panel on the <a href="http://cytoscape.org/cgi-bin/moin.cgi/VizMapper" target="_top">VizMapper</a> main panel.  Default Apearence Editor pops up (see below.) </p></li><li><p>Edit the following visual properties and press <span class="emphasis"><em><span class="strong"><strong>Apply</strong></span></em></span>.  Since you changed opacity of the node, you can see the nodes bihind the front node (see below.) </p><div class="itemizedlist"><ul type="disc"><li>Node Oppacity - 100 </li><li>Edge Color - White </li><li>Background Color - Black </li></ul></div><p><span class="inlinemediaobject"><img src="images/Opacity1.png" alt="attachment:Opacity1.png"/></span> </p></li><li><p>Cretate a <span class="strong"><strong>Discrete Node Color Mapping</strong></span>.  Select <span class="emphasis"><em>Degree</em></span> as controlling attribute. </p></li><li><p>Select <span class="strong"><strong>Node Color</strong></span>, then right click to show popup menu.  Select <span class="underline">Generate discrete values → Rainbow 1</span>.  It  generates different colors for different attribute values as shown below. </p><p><span class="inlinemediaobject"><img src="images/Rainbow1.png" alt="attachment:Rainbow1.png"/></span> </p></li><li><p>Cretate a <span class="strong"><strong>Discrete Node Size Mapping</strong></span>.  Select <span class="emphasis"><em>Degree</em></span> as controlling attribute. </p></li><li><p>Select <span class="strong"><strong>Node Size</strong></span> and right click to show popup menu.  Select <span class="underline">Generate Discrete Values → Series (Numbers Only)</span>.  Type <span class="emphasis"><em>30</em></span> for the first value and click OK.  Enter <span class="emphasis"><em>15</em></span> for increment. </p></li><li>Apply Force-Directed layout.  Final view of the window looks like the following. <p><span class="inlinemediaobject"><img src="images/tutorial4_last.png" alt="attachment:tutorial4_last.png"/></span> </p></li></ol></div><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Advanced Topics"/>Advanced Topics</h3></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Editing Discrete Mappings"/>Editing Discrete Mappings</h4></div></div></div><p>From version 2.5, several utility functions are available for Discrete Mappings.  You can use those functions by right clicking anywhere on the <span class="emphasis"><em>Visual Mapping Browser</em></span> (shown below.) </p><p><span class="inlinemediaobject"><img src="images/VizMapperPopupMenu.png" alt="attachment:VizMapperPopupMenu.png"/></span> </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Automatic Value Generators"/>Automatic Value Generators</h5></div></div></div><div class="itemizedlist"><ul type="disc"><li><p><span class="emphasis"><em><span class="strong"><strong>Generate Discrete Values</strong></span></em></span> - Functions in this menu category are value generator for discrete mappings.  Users can set values for discrete mappings automatically by these functions. </p><div class="itemizedlist"><ul type="circle"><li><p><span class="strong"><strong>Rainbow 1</strong></span> and <span class="strong"><strong>Rainbow 2</strong></span> - These functions try to assign as different colors as possible to each values.  see the example below: </p><div class="itemizedlist"><ul type="square"><li><p><span class="inlinemediaobject"><img src="images/RainbowExample1.png" alt="attachment:RainbowExample1.png"/></span> </p></li></ul></div></li><li><p><span class="strong"><strong>Randomize</strong></span> - Randomize colors and numbers.  If you use this function for numerical values (node size, opacity, etc.) you need to specify a range.  For example, if you want to set values from 1 to 100, you need to type <span class="emphasis"><em>1-100</em></span> in the dialog. </p></li><li><p><span class="strong"><strong>Series</strong></span> - Set series of numbers to the specified mapping. </p><div class="itemizedlist"><ul type="square"><li><p><span class="inlinemediaobject"><img src="images/SeriesExample1.png" alt="attachment:SeriesExample1.png"/></span> </p></li></ul></div></li><li><p><span class="strong"><strong>Fit node size to label</strong></span> - This function is only for node width and height.  When node size is <span class="emphasis"><em>unlocked</em></span> AND <span class="emphasis"><em>Node Width/Height</em></span> discrete mappings are available, you can fit the size of each node to label automatically by selecting this function.  See the example below: </p><div class="itemizedlist"><ul type="square"><li><p><span class="inlinemediaobject"><img src="images/NodeLabelFit.png" alt="attachment:NodeLabelFit.png"/></span> </p></li></ul></div></li></ul></div></li><li><p><span class="emphasis"><em><span class="strong"><strong>Modify Discrete Values</strong></span></em></span> - Currently, this is only for colors.  You can change overall brightness for discrete color mappings. </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Edit Selected Values at Once"/>Edit Selected Values at Once</h5></div></div></div><p>You can set multiple values at once.  First, you need to select rows you want to change values then select <span class="emphasis"><em>Edit selected values at once...</em></span>.  A dialog pops up and you can enter the new value for the selected rows. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Lock Node Width/Height"/>Lock Node Width/Height</h5></div></div></div><p>If this menu item is checked, <span class="emphasis"><em>Node Width</em></span> and <span class="emphasis"><em>Node Height</em></span> mappings are ignored and <span class="emphasis"><em>Node Size</em></span> overrides them.  If you want to use <span class="emphasis"><em>Fit node size to label</em></span> function, you need to unlock this. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Working with Continuous Mapping Editors"/>Working with Continuous Mapping Editors</h4></div></div></div><p>There are three kinds of Continuous Mapping Editors.  Each of them are associated with a specific visual attributes: </p><div class="table"><a id="d0e5406"/><p class="title"><b>Table 24. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/></colgroup><tbody><tr><td><p> <span class="emphasis"><em><span class="strong"><strong>Editor Type</strong></span></em></span> </p></td><td><p> <span class="emphasis"><em><span class="strong"><strong>Supported Data Type</strong></span></em></span> </p></td><td><p> <span class="emphasis"><em><span class="strong"><strong>Visual Attributes</strong></span></em></span> </p></td></tr><tr><td><p> <span class="strong"><strong>Color Gradient Editor</strong></span> </p></td><td><p> Color </p></td><td><p> node/edge/border/label colors </p></td></tr><tr><td><p> <span class="strong"><strong>Continuous-Continuous Editor</strong></span> </p></td><td><p> Numbers </p></td><td><p> size/width/opacity </p></td></tr><tr><td><p> <span class="strong"><strong>Continuous-Discrete Editor</strong></span> </p></td><td><p> All others </p></td><td><p> font/shape/text </p></td></tr></tbody></table></div></div><br class="table-break"/><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Range Setting Panel"/>Range Setting Panel</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/RangeSetting26.png" alt="attachment:RangeSetting26.png"/></span> </p><p>Each editor has a common section named <span class="emphasis"><em><span class="strong"><strong>Range Setting</strong></span></em></span>. </p><div class="orderedlist"><ol type="1"><li><p><span class="strong"><strong>Handle Value Box</strong></span> - This box displays current value for selected slider handle.  Also you can directly type value in this box to move the slider to exact location. </p></li><li><p><span class="strong"><strong>Min/Max Button</strong></span> - Set the overall range of this editor.  First time you open the editor, the <span class="emphasis"><em>Min</em></span> and <span class="emphasis"><em>Max</em></span> values are set by the range of attribute you selected, i.e., minimum and maximum value of the attribute will be set to the range of this editor.  You can change this range anytime you want by pressing this button. </p></li><li><p><span class="strong"><strong>Add Handle Button</strong></span> - Add a new handle to the editor. </p></li><li><p><span class="strong"><strong>Delete Handle Button</strong></span> - Delete selected handle from the slider widget.   </p></li></ol></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Gradient Editor"/>Gradient Editor</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/GradientEditorSample26.png" alt="attachment:GradientEditorSample26.png"/></span> </p><p>Gradient Editor is a editor to create continuous mapping for colors.  To change the color of each region, just double click the handles (small triangles on the top).  Color gradient will be created only when editor has two or more handles (see the example below). </p><div class="table"><a id="d0e5533"/><p class="title"><b>Table 25. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p> 1 Slider (No Graient) </p></td><td><p> 2 Sliders </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/OneSlider.png" alt="attachment:OneSlider.png"/></span> </p></td><td><p> <span class="inlinemediaobject"><img src="images/TwoSliders.png" alt="attachment:TwoSliders.png"/></span> </p></td></tr></tbody></table></div></div><br class="table-break"/><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Continuous-Continuous Editor"/>Continuous-Continuous Editor</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/C2CEditor26.png" alt="attachment:C2CEditor26.png"/></span> </p><p>Continuous-Continuous Editor is for creating mapping between numerical attributes and numerical visual properties (size/opacity).  To change the value assigned on Y-axis (visual property shown in the example above is node size), drag the red squares or double click on the squares to directly type exact value.  </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Continuous-Discrete Editor"/>Continuous-Discrete Editor</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/C2DEditor26.png" alt="attachment:C2DEditor26.png"/></span> </p><p>Continuous-Discrete Editor is used to create mapping from numerical attribute values to discrete visual properties, such as font, shape, or line style.  To edit a value for specific region, double click on the icon on the track.  </p><p/></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Managing Visual Styles"/>Managing Visual Styles</h3></div></div></div><p>All Cytoscape Visual Style settings are initially loaded from a a default file called <code class="code">vizmap.props</code> that cannot be altered by users. When users make changes to the visual properties, a <code class="code">vizmap.props</code> file is saved in the session file. This means that assuming you save your session, you will not lose your visual properties. No other <code class="code">vizmap.props</code> files are saved during normal operation. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Saving Visual Styles"/>Saving Visual Styles</h4></div></div></div><p>Visual styles are automatically saved with the session they were created in. Before Cytoscape exits, you will be prompted to make sure you save the session before quitting. It is also possible to save your visual styles in a file separate from the session file. To do this, navigate to the File → Export → Vizmap Property File menu option and save the properties as a file. This feature can be used to share visual styles with other users. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Importing Visual Styles"/>Importing Visual Styles</h4></div></div></div><p>To import existing visual styles, navigate to the File → Import → Vizmap Property File menu option and select a <code class="code">vizmap.props</code> file.  Imported properties will supplement existing properties or override existing properties if the properties have the same name. You can also specify a visual properties file using the -V command line option (<code class="code">cytoscape.sh -V myVizmap.props</code>).  Visual properties loaded from the command line will override any default properties. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Default Visual Styles"/>Default Visual Styles</h4></div></div></div><p>It is possible to change the default visual properties for all sessions of Cytoscape.  To do this, navigate to the Edit → Preferences → Properties... menu option, check the "Make Current Visual Styles Default" box in the Default Visual Styles section, and click the OK button.  This will save the current visual styles as a <code class="code">vizmap.props</code> file to your <code class="code">.cytoscape</code> directory (found in your home directory). These visual styles will then be loaded each time Cytoscape is started. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Bypassing Visual Styles"/>Bypassing Visual Styles</h3></div></div></div><p>Cytoscape has a feature that allows users to override visualizations created by the VizMapper for individual nodes and edges. This feature is available by right-clicking on a node or edge and then clicking on the Visual Mapping Bypass menu. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/VizmapBypass26.png" alt="attachment:VizmapBypass26.png"/></span> </p></li></ul></div><p>Each visual property of the node or edge is displayed.  When a property is overridden, a checkmark appears next to the property and a [Reset &lt;Property Name&gt;] menu option appears directly below it.  By clicking this Reset option, the bypass will be removed and the attribute will be displayed as defined by the VizMapper. At the bottom of the menu a Reset All option appears.  When clicked, this will remove all bypasses for the specified node or edge.  In the example above, you can see the the selected node size, color, and shape have been overridden.  This is apparent in the appearance of the node itself and by the check marks in the popup menu. </p><p>It is important to realize that the the Visual Mapping Bypass only works for individual nodes and edges and not for all nodes or edges of a specific type.  Using the bypass function is not particularly resource intensive, meaning you can use it as much as you like. However, if you ever find yourself repeating the same bypasses, then you should consider using the VizMapper instead. </p><p>Bypass is accomplished using special attributes with names like <code class="code">node.fillColor</code> and <code class="code">node.shape</code>.  These are normal Cytoscape attributes and can be seen and edited in the Attribute Browser.  The value of the attribute is a string representation of a property.  For example, color is represented by 3 integers representing the RGB (red, green, blue) value of the color.  Different types of properties have different string representations.  When in doubt, just use the right click menu to create valid attribute values. </p><p>Because bypass values are specified using normal attributes, these attributes will persist between sessions only as long as you save your session!  If you don't save your session, you will lose whatever bypass values you set. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Finding and Filtering Nodes and Edges"/>Finding and Filtering Nodes and Edges</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="QuickFind"/>QuickFind</h3></div></div></div><p>Cytoscape includes a Quick Find feature, which enables you to quickly find nodes and edges. </p><p><span class="inlinemediaobject"><img src="images/quick_find2.png" alt="attachment:quick_find2.png"/></span> </p><p>Using Quick Find is very simple. Here is how it works: </p><div class="itemizedlist"><ul type="disc"><li><p>Import a network. For example, load up <code class="code">sampleData/galFiltered.sif</code>. </p></li><li>Start typing in the text box. For example, enter "yd". The search box will automatically display a list of all matching nodes. </li></ul></div><p><span class="inlinemediaobject"><img src="images/quick_find3.png" alt="attachment:quick_find3.png"/></span> </p><div class="itemizedlist"><ul type="disc"><li>Select a matching node. Cytoscape will automatically zoom in on the selected node. </li></ul></div><p><span class="inlinemediaobject"><img src="images/quick_find4.png" alt="attachment:quick_find4.png"/></span> </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Configuring Quick Find"/>Configuring Quick Find</h4></div></div></div><p>Quick Find works by creating an internal index of all nodes within the network. By default, Cytoscape indexes all nodes by the node identifier. However, you can configure Quick Find to index nodes or edges, and you can set any attribute as the index. </p><p>For example, if you load up a BioPAX file (such as <code class="code">sampleData/Apoptosis_BioPAX.xml</code>), your network will be automatically annotated with numerous attributes. To index the network based on cellular location, click the Quick Find configuration button, and select "biopax.cellular_location" from the dropdown list. </p><p><span class="inlinemediaobject"><img src="images/quick_find5.png" alt="attachment:quick_find5.png"/></span> </p><p>You can then quickly find all proteins located in the "plasma membrane" by just typing "p". </p><p><span class="inlinemediaobject"><img src="images/quick_find6.png" alt="attachment:quick_find6.png"/></span> </p><p><span class="strong"><strong>Tip:</strong></span> If you don't know what to search for, just leave the search box empty, and click on the down arrow directly next to the search box. Cytoscape will provide you with an initial list of matches. In the case below, we will get a list of all distinct cellular locations in the network. </p><p><span class="inlinemediaobject"><img src="images/quick_find7.png" alt="attachment:quick_find7.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Dynamic Filtering via Quick Find"/>Dynamic Filtering via Quick Find</h4></div></div></div><p>If you choose to index on a numerical attribute, the Quick Find search box changes to a dynamic slider for quick filtering. </p><p>For example: </p><div class="orderedlist"><ol type="1"><li><p>Import a network such as <code class="code">sampleData/galFiltered.sif</code> using File → Import → Network (multiple file types). </p></li><li><p>Import an expression data file such as <code class="code">sampleData/galExpData.pvals</code> using File → Import → Attribute/Expression Matrix... . </p></li><li>Configure Quick Find to index the network using the gal1RGExp attribute. <div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/quick_find8.png" alt="attachment:quick_find8.png"/></span> </p></li></ul></div></li><li>Use the slider widget to automatically filter the entire network. </li></ol></div><p><span class="inlinemediaobject"><img src="images/quick_find9.png" alt="attachment:quick_find9.png"/></span> </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Filters"/>Filters</h3></div></div></div><p>Filters allow you to quickly select multiple nodes or edges of interest by comparing node and edge attributes to properties you specify. For example, you can select all the nodes whose name contains a specific pattern, or whose numeric attribute value falls within a certain range. You can also perform complex selection by defining filters and performing the Boolean operation between them. </p><p>Since Cytoscape 2.5, the filters are redesigned to provide more user-friendly interfaces. The newly designed filters preserve most of the functionality of the old filters while introducing more, its API can also be used by other plugins. The old filters are still preserved, because there are still a few features are not implemented in new filters, and many users already got used to the old ones. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Using New Filters"/>Using New Filters</h3></div></div></div><p>By default, you should see a filter icon on the toolbar: <span class="inlinemediaobject"><img src="images/filter_icon.png" alt="attachment:filter_icon.png"/></span> . If you click on it, the filters tab on the Control Panel will be selected. You can also access the filters by clicking directly on the tab or by using the Select menu pull-down and choose “Use Filters” menu item. The filters panel initially looks like this: </p><p><span class="inlinemediaobject"><img src="images/filterInit.png" alt="attachment:filterInit.png"/></span> </p><p><span class="strong"><strong>1. Filter Definition</strong></span> </p><p>To create a new filter, click the option button and select “Create new filter” from the list provided. Enter a name for the new filter. </p><p><span class="inlinemediaobject"><img src="images/FilterPopupMenu.png" alt="attachment:FilterPopupMenu.png"/></span> </p><p><span class="inlinemediaobject"><img src="images/Filter_create_filter_dialog.png" alt="attachment:Filter_create_filter_dialog.png"/></span> </p><p><span class="emphasis"><em>Definition of simple filter</em></span> </p><p>When a new filter is created, it is empty initially. It can be defined by choosing attributes (one at a time) in the Attribute/Filter comboBox and clicking the Add button. Note that in the comboBox each attribute has a prefix, either node. or edge., denoting what type of attribute it is. After add button is clicked, a widget, depended on the item selected, will be added to the definition panel. If the attribute type is text-based (or String), the widget will be a indexed-text-box widget, if numeric attribute, the widget will be a range slider; both are shown below. </p><p><span class="inlinemediaobject"><img src="images/FilterDefinition.png" alt="attachment:FilterDefinition.png"/></span> </p><p>For each widget, the name of the attribute it represents is on the left. A NOT checkbox gives the option to get the negation of selection for the widget. There is a trash-can icon on the right. Clicking on the trash-can icon will delete the widget. In this way, the filter definition can be modified after it is defined. </p><p>Note that if more than one widget is added on the filter definition panel, the relationship between them is “AND” by default. This relationship can be changed to “OR” by selecting the OR relation in Advanced Panel. </p><p>The Advanced panel can be opened by clicking on the plus (+) sign. </p><p><span class="inlinemediaobject"><img src="images/FilterAdvancedPanel.png" alt="attachment:FilterAdvancedPanel.png"/></span> </p><p>There are three rows in the advanced panel: </p><p>1. The first row, labeled Save. The two checkboxes (Global and Session) will determine where the filter is saved. By default, filters are saved in individual sessions. If the Global checkbox is checked, the filter will be saved in the global properties file. Note also that the prefix of the filter name in the Current Filter dropdown list reflects where it is saved. </p><p>2. The second row, Relation, will determine what Boolean operation (AND or OR) will be applied to each individual widget. </p><p>3. Negation checkbox. If this checkbox is checked, the result of the filter will be negated. </p><p><span class="emphasis"><em>Definition of complex filter</em></span> </p><p>In the pull-down list of Attribute/Filter comboBox, there are two sections, the top one is the list of attributes, the bottom one is the list of previously defined filter. Those previously dfined filters can also be used as components of other filters. By combinational use of AND, OR, NOT and pre-defined filters, filters with arbitary complexity can be defined. </p><p><span class="strong"><strong>3. Apply the filter</strong></span> </p><p>If a network is small (number of nodes or edges less than 1000), the filter will be applied dynamically when a widget is added or any value is adjusted. If the network is large, then Apply button should be clicked to apply the filter. </p><p>The threshold (1000 by default) can be changed by defining a Cytoscape property value <span class="emphasis"><em>dynamicFilterThreshold</em></span>. </p><p><span class="strong"><strong>4. Other filters </strong></span> </p><p>In the option menu pulldown, there are menu items “Create new topology filter”, “Create new <a href="http://cytoscape.org/cgi-bin/moin.cgi/NodeInteraction" target="_top">NodeInteraction</a> filter” and “Create new <a href="http://cytoscape.org/cgi-bin/moin.cgi/EdgeInteraction" target="_top">EdgeInteraction</a> filter”. </p><p><span class="emphasis"><em>Topology Filter</em></span> </p><p>Topology filter will select nodes based on the properties of its near-by nodes (neighbors). To create a topology filter, choose the menu item “Create new topology filter” from the option menu. See below, </p><p><span class="inlinemediaobject"><img src="images/FilterTopo.png" alt="attachment:FilterTopo.png"/></span> </p><p><span class="emphasis"><em>Interaction filter</em></span> </p><p>Interaction filters are used to select nodes/edges based on the properties of their neighboring edges/nodes. See below for a node interaction filter. </p><p><span class="inlinemediaobject"><img src="images/FilterNodeInteraction.png" alt="attachment:FilterNodeInteraction.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Using Old Filters"/>Using Old Filters</h3></div></div></div><p>The old filters is the one existed in previous versions of Cytoscape. Several types of filters are available. </p><p>Basic filters allow the selection of multiple nodes or edges according to singe attribute data: </p><div class="itemizedlist"><ul type="disc"><li>String filters allow selection of nodes or edges with attributes matching specified text patterns. These patterns may include the wildcards * and ?. </li><li><p>Numerical filters allow selection of nodes or edges according to numerical attributes and the mathematical operators &gt;, =, and &lt;. </p></li></ul></div><p>Compound filters allow selection based on the application of pre-existing filters: </p><div class="itemizedlist"><ul type="disc"><li>Topology filters allow selection of nodes with neighbors that match some pre-existing filter. </li><li>Boolean filters allow the combination of multiple filters using the AND, OR and XOR operators. </li></ul></div><p>Example filters are shipped with the plugin to get started. </p><p>To use the old filters, go to Select → Use Old Filters. You will see a filters window which initially looks like the following: </p><p><span class="inlinemediaobject"><img src="images/init_filter_dialog.png" alt="attachment:init_filter_dialog.png"/></span> </p><p>If the first filter is selected, then the window looks as shown: </p><p><span class="inlinemediaobject"><img src="images/first_filter_dialog.png" alt="attachment:first_filter_dialog.png"/></span> </p><p>There are three panels in the Filters window: </p><div class="itemizedlist"><ul type="disc"><li>The right-hand panel: An existing or newly created filter can be edited in this area. Each filter type has its own user interface for editing. </li><li>The lower left panel: All available filters are shown in this list. Initially, this list will contain only sample filters, but as you create more, they will be added here. </li><li>The upper left panel: Pressing the Create new filter button adds a filter to the “Available Filters” box, and the Remove selected filter button deletes the currently selected filter. </li></ul></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Creating Filters"/>Creating Filters</h4></div></div></div><p>Clicking the Create new filter button allows you to select a type of filter to create, as shown below. </p><p><span class="inlinemediaobject"><img src="images/create_new_filter_dialog.png" alt="attachment:create_new_filter_dialog.png"/></span> </p><p>The important thing to realize when creating a filter is that the filter does not do anything by itself. Once created, the filter must be applied to the network. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="String Filter"/>String Filter</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/string_filter.png" alt="attachment:string_filter.png"/></span> </p><p>The String Filter allows you to filter nodes or edges by a given text attribute of your current loaded network. Search terms are entered in the text box at the bottom. For example, to match any Node whose canonicalName starts with “YDL”, you would select “Node”, “canonicalName” and then type in “YDL*”. The * is important, as it matches any number of characters after YDL. If you want to be more specific and only select nodes whose canonicalName starts with YDL00 followed by any other two characters, you would type “YDL00??”. The “?” denotes any single character, while the “*” represents zero or more characters. Full regular expression searching is supported, although is not covered here. Once the filter is defined, it will be assigned a default descriptive name. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Numerical Filter"/>Numerical Filter</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/numerical_filter.png" alt="attachment:numerical_filter.png"/></span> </p><p>The Numerical Filter also allows you to filter nodes or edges, and presents you with a list of available attributes. This filter matches all values in the network of your selected attribute that are greater than, less than, or equal to a number you type in the search box. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Node Topology Filter"/>Node Topology Filter</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/node_topology_filter.png" alt="attachment:node_topology_filter.png"/></span> </p><p>The node topology filter allows you to select all nodes selected by a preexisting filter and having <span class="emphasis"><em>n</em></span> neighbours found at a distance of <span class="emphasis"><em>m</em></span> or fewer edges away. For instance, to select all the nodes adjacent to a node with the canonical name matching ‘YD*’, you would “select nodes with 1 neighbors”, “within distance 1”, “that pass the filter Node: canonicalName ~ YD*”. </p><p>The node topology filter depends on the existence of other filters. By default, [No Filter] is selected; it doesn't filter out anything, but rather selects all nodes. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Boolean Filter"/>Boolean Filter</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/boolean_filter.png" alt="attachment:boolean_filter.png"/></span> </p><p>The Boolean Meta-Filter allows you to define a new filter that is a logical combination of existing filters. Available filters are displayed. By selecting one or more filters, you can then choose whether Nodes or Edges pass “ALL” (AND), “AT LEAST ONE” (OR), or “ONLY ONE” (XOR) of the selected filters. Once created, Boolean filters can then themselves be combined using the Boolean filter to create arbitrarily complex logical combinations of filters. Note that unlike the String and Numerical Filters, Boolean filters will need to be assigned a name manually. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Saving Filters"/>Saving Filters</h4></div></div></div><p>Filters are currently saved automatically in the <code class="code">filters.props</code> file found in the <code class="code">.cytoscape</code> directory, found in each user's home directory. Once created, filters are saved for future sessions, as long as you exit Cytoscape normally via File → Quit (i.e. not via ctrl-c on Linux). </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Running filters"/>Running filters</h4></div></div></div><p>Any available filter can be run by pressing the ‘Apply selected filter’ button. When a filter is applied to your network and multiple nodes or edges are selected, all of the normal selection-related operations may be performed, such as Delete Selected Node/Edges, Copy To New Network, and Invert Selection. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="The Select Menu"/>The Select Menu</h3></div></div></div><p>The Select → Nodes and Select → Edges menus provide several mechanisms for selecting nodes and edges. Most options are fairly straightforward; however, some need extra explanation. </p><p>Select → Nodes → By Name... selects nodes by the node identifier (ID). This is the value seen in the leftmost column of the attribute browser. This does not change, even if the displayed node label changes! </p><p>Select → Nodes → From File... selects nodes based on node identifiers found in a specified file. The file format is simply one node id per line: </p><p/><pre class="screen">Node1
Node2
Node3
...
</pre></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Editing Networks"/>Editing Networks</h2></div></div></div><p>Using Cytoscape's Editor, you can build and modify networks interactively by dragging and dropping nodes and edges from a palette onto the main network view window. The palette contains a set of shapes (for nodes) and arrows (for edges). The shapes on the palette are defined by the current Visual Style, with Node Shape and Node Color mapping into the shape and color of a node, and Edge Target Arrow mapping into the target arrow of an edge. An example of an editor, with the palette contained in CytoPanel 1, is shown below. </p><p><span class="inlinemediaobject"><img src="images/editor1_25.png" alt="attachment:editor1_25.png"/></span> </p><p>To edit an existing network, just select the Editor tab in CytoPanel 1.  To start editing a new network, create a new network by going to File → New → Network → Empty Network. </p><p>To add a node to a network, drag and drop a node shape from the palette onto the canvas.  To connect two nodes with an edge, drag and drop an arrow shape onto a node on the canvas. This node becomes the source node of the edge. Move the cursor and a rubber-banded line will follow the cursor. As the cursor passes over another node, that node is highlighted and the rubber-banded line will snap to a connection point on that second node. Click the mouse while over this node and the connection will be established. </p><p><span class="inlinemediaobject"><img src="images/editor2_25.png" alt="attachment:editor2_25.png"/></span> </p><p>You can abort the drawing of the edge by clicking on an empty spot on the palette. </p><p>Note that if you change the Visual Style, the palette used by the current network view will also change to be consistent with the mappings in the new Visual Style. </p><p>There is also an Edit → Connect Selected Nodes option that, when chosen, creates a clique amongst the selected nodes. </p><p>The editor provides accelerators for adding nodes and edges. Control-clicking at a position on the canvas creates a node at that position. The NODE_TYPE attribute of the node will be the same as the NODE_TYPE of the node most recently added, defaulting back to the original visual style. In this manner, you can use control-clicking as a kind of "stamp" to add multiple nodes of the same type to the network. Control-clicking on a node on the canvas starts an edge with its source at that node. Move the cursor and a rubber-banded line will follow the cursor. As the cursor passes over another node, that node is highlighted and the rubber-banded line will snap to a connection point on that second node. Control-click the mouse again and the connection will be established. The EDGE_TYPE attribute of the edge will be the same as the EDGE_TYPE of the edge most recently added, defaulting back to the original visual style. You can abort the drawing of the edge by control-clicking on an empty spot on the palette. </p><p>You can delete nodes and edges by selecting a number of nodes and edges, then selecting Edit → Delete Selected Nodes and Edges. You can recover any nodes and edges deleted from a network by going to Edit → Undo. However, note that this will restore <span class="strong"><strong>all</strong></span> nodes and edges that were previously deleted from the network, not just those deleted by the most recent delete operation. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Add Interactions  (SIF Style)"/>Add Interactions  (SIF Style)</h3></div></div></div><p>You can also add interactions by typing SIF (Simple Interaction Format) style entries. </p><div class="orderedlist"><ol type="1"><li>Select a network you want to add new interactions. </li><li><p>Open <span class="strong"><strong>Editor</strong></span>. </p></li><li><p>Double-click space on the network window.  You will be asked to type interaction in SIF style.  Try <span class="emphasis"><em>A pp B</em></span>. </p></li><li>New interaction appears on the network. </li></ol></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Plugins and the Plugin Manager"/>Plugins and the Plugin Manager</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="The Plugin Manager"/>The Plugin Manager</h3></div></div></div><p>The Plugin Manager, first introduced in Cytoscape 2.5, allows users to quickly and conveniently add extra features to Cytoscape directly from within Cytoscape, eliminating the need for manual searches through different webpages to install and update plugins. </p><p>For plugin developer, to enable automatic download of your plugin to Cytoscape users, your plugin should be in compliance with Cytoscape 2.5, and the plugin jar/zip files should be uploaded to the Cytoscape plugin web site at <a href="http://cytoscape.org/plugins25/index.php" target="_top">http://cytoscape.org/plugins25/index.php</a>. </p><p><span class="strong"><strong>Note:</strong></span> If you do not have Internet access enabled, you will not see the list of available plugins or be able to automatically update existing ones; however, you will still be able to view and delete previously installed plugins. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Get New Plugins"/>Get New Plugins</h3></div></div></div><p>To install new features, go to the Plugin Manager at Plugins → Manage Plugins. On the left side of the window that pops up, you will see plugin folders labeled Currently Installed and Available for Install. Double-clicking on these will show sub-folders, and then the plugins themselves. To find out more about a specific plugin, click on its name to display some basic information on the right-hand side of the window. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/small_manager.png" alt="attachment:small_manager.png"/></span> </p></li></ul></div><p>The Currently Installed folder contains a number of default plugins that are fully integrated in every copy of Cytoscape, as well as any additional installed plugins. In contrast, the Available for Install folder displays plugins that may be installed. To install and use these plugins, click on the file name and then click on the Install button at the bottom of the window. A license agreement may appear, in which case you must accept it in order to download the plugin. You will then see a progress bar as the plugin is automatically downloaded and installed into your current version of Cytoscape. The progress bar will disappear when the download is complete. Other plugins can then be downloaded, or the manager can be closed by clicking the Close button. </p><p>Once a new plugin is added, it can immediately be used. Closing and restarting Cytoscape is not required. For example, installing the MCODE plugin will automatically result in new menu options such as Plugins → MCODE → Start MCODE. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Themes"/>Themes</h4></div></div></div><p>Some groups of plugins have been released in a single bundle called a "Theme".  Installing it will get you the functionality provided by all of the plugins listed within the theme.  To see available themes just look under the "Theme" heading under Available for Install.  Note that once a theme is installed updates to the included plugins will only be installed when the entire theme has been released with a newer version.  If a theme has been installed none of the plugins may be installed or deleted separately. </p><p><span class="strong"><strong>Troubleshooting</strong></span> </p><div class="itemizedlist"><ul type="disc"><li><p>If an installation error appears, automatic installation of the plugin may not be supported. To manually install the plugin, go to the Cytoscape plugins page (<a href="http://cytoscape.org/plugins25/index.php" target="_top">http://cytoscape.org/plugins25/index.php</a>), scroll down to find the plugin, click on the appropriate link to download the file, and then save it in the <code class="code">Cytoscape/plugins</code> folder on your hard drive. </p></li><li>Cytoscape will require a restart in order to load the manually installed plugin. </li><li>If the plugin does not appear in the Currently Installed folder of the Plugin Manager, then Cytoscape was unable to load the plugin. Your command line will display the error message generated. </li></ul></div><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Delete Existing Plugins"/>Delete Existing Plugins</h3></div></div></div><p>To delete an existing plugin or theme, go to the Plugin Manager at Plugins → Manage Plugins, click on the appropriate plugin once to select it, and then click the Delete button at the bottom of the manager. It is best to avoid deleting core plugins, as this may cause Cytoscape or other plugins to malfunction. A plugin will be completely deleted after you have restarted Cytoscape. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Update Existing Plugins"/>Update Existing Plugins</h3></div></div></div><p>Going to Plugins → Update Plugins will result in Cytoscape automatically accessing the Internet to look for updated versions of the plugins you have installed. If newer versions are available, Cytoscape will then display the available updates in the Plugin Manager; otherwise, it will inform you that there are no updatable plugins. When multiple updates are detected, you will have the option to download some or all of the updates. Some updates may require re-accepting the license before the download will begin. All updates require Cytoscape to be restarted in order for the update to be fully installed. After clicking the Update All or Update Selected buttons, a progress bar will appear showing the status of the download; it will disappear once the update is complete, and the Plugin Manager can then be closed by clicking the Close button. </p><p><span class="strong"><strong>Note:</strong></span> The Update All button will update all listed plugins to the newest available version. If you prefer a different version, select the name and click Update Selected. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Download Plugins from a Custom Site"/>Download Plugins from a Custom Site</h3></div></div></div><p>The Plugin Manager automatically connects to the Cytoscape website and the sites of registered Cytoscape plugins to check for and download any available plugins and updates. To change this setting, open the Plugin Manager (by going to Plugins → Manage Plugins) and click the Change Download Site button. You will be asked to choose a site from a preset list of bookmarks. Select one and click Ok. Alternatively, click the Add Site button to open the Bookmark Manager and modify the list of bookmarks. In the Bookmark Manager, choose the plugins category from the dropdown list, and then add/edit/delete URLs using the appropriate buttons on the right-hand side of the window. After the correct site is selected, set it as the download site by clicking OK. The Plugin Manager will refresh using that site. </p><p><span class="strong"><strong>Note:</strong></span> Custom download sites added from the Plugin Manager are not currently saved when you exit Cytoscape. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="CytoPanels"/>CytoPanels</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="What are CytoPanels?"/>What are CytoPanels?</h3></div></div></div><p><span class="emphasis"><em><span class="strong"><strong><a href="http://cytoscape.org/cgi-bin/moin.cgi/CytoPanels" target="_top">CytoPanels</a></strong></span></em></span> are floatable/dockable panels designed to cut down on the number of pop-up windows within Cytoscape and to create a more unified user experience.  These panels used to be called <a href="http://cytoscape.org/cgi-bin/moin.cgi/CytoPanel" target="_top">CytoPanel</a> 1, 2, and 3.  From 2.5, they are named based on their functions.  The following screenshot shows the file <code class="code">yeastHighQuality.sif</code> and GO annotations loaded into Cytoscape, performed Force-Directed layout, enable Align and Distribute tools, and then run MCODE plugin for the data sets.  In <span class="emphasis"><em>Control Panel</em></span> (at the left-hand side of the screen), the Network Manager, Network Overview, <a href="http://cytoscape.org/cgi-bin/moin.cgi/VizMapper" target="_top">VizMapper</a>, Filters, and Cytoscape Editor have been loaded.  On the bottom of the panel, there is another <a href="http://cytoscape.org/cgi-bin/moin.cgi/CytoPanel" target="_top">CytoPanel</a> called <span class="emphasis"><em>Tool Panel</em></span>.  In the <span class="emphasis"><em>Data Panel</em></span>, the Attribute Browser has been loaded.  In addition, Result of the analysis by MCODE plugin is shown in <span class="emphasis"><em>Result Panel</em></span> (at the right-hand side).  </p><p><span class="inlinemediaobject"><img src="images/cytopanel-default_25.png" alt="attachment:cytopanel-default_25.png"/></span> </p><p>The user can then choose to resize, hide or float <a href="http://cytoscape.org/cgi-bin/moin.cgi/CytoPanels" target="_top">CytoPanels</a>. For example, in the screenshot below, the user has chosen to float all panels and toolbar: </p><p><span class="inlinemediaobject"><img src="images/cytopanel-float_25.png" alt="attachment:cytopanel-float_25.png"/></span> </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Basic Usage"/>Basic Usage</h3></div></div></div><p>Cytoscape includes four CytoPanels: <span class="emphasis"><em><span class="strong"><strong>Control Panel</strong></span></em></span> on the left, <span class="emphasis"><em><span class="strong"><strong>Tool Panel</strong></span></em></span> on the bottom of <span class="emphasis"><em>Control Panel</em></span>, <span class="emphasis"><em><span class="strong"><strong>Data Panel</strong></span></em></span> on the bottom, and <span class="emphasis"><em><span class="strong"><strong>Result Panel</strong></span></em></span> on the right. By default, <span class="emphasis"><em>Control Panel</em></span> and <span class="emphasis"><em>Data Panel</em></span> will appear.  <span class="emphasis"><em>Result Panel</em></span> may appear, depending on the mix of Cytoscape plugins that you currently have installed.  <span class="emphasis"><em>Tool Panel</em></span> will appear when you select the following commands under the <span class="emphasis"><em>Layout</em></span> menu: <span class="emphasis"><em>Rotate</em></span>, <span class="emphasis"><em>Scale</em></span>, and <span class="emphasis"><em>Align and Distribute</em></span>. </p><p>All panels can be shown or hidden using the View → Show/Hide functions. </p><p><span class="inlinemediaobject"><img src="images/cytopanel-menu-items_25.png" alt="attachment:cytopanel-menu-items_25.png"/></span> </p><p>In addition, CytoPanels can be floated or docked by selecting the icon at the top right corner of each CytoPanel. The icon and tooltip will change based on the CytoPanel state. If the CytoPanel is docked, clicking on the icon will float the CytoPanel, as indicated by the “Float Window” tooltip. Alternatively, if the CytoPanel is floating, clicking on the icon will dock the CytoPanel, as indicated by the “Dock Window” tooltip. </p><p><span class="inlinemediaobject"><img src="images/cytopanel-float-dock-icon.png" alt="attachment:cytopanel-float-dock-icon.png"/></span> </p></div></div><div class="itemizedlist"><ul type="disc"><li><a id="renderer"/></li></ul></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Rendering Engine"/>Rendering Engine</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="What is Level of Detail (LOD)?"/>What is Level of Detail (LOD)?</h3></div></div></div><p>In Cytoscape 2.3, a new network rendering engine is being introduced.  The goal of the rendering engine is to be able to display large networks (&gt;10,000 nodes), yet retain interactive speed.  To accomplish this goal, a technique involving <span class="emphasis"><em><span class="strong"><strong>level of detail (LOD)</strong></span></em></span> is being used.  Based on the number of objects (nodes and edges) being rendered, an appropriate <span class="emphasis"><em>level of detail</em></span> is chosen.  For example, by default, node labels (if present) are only rendered when less than 100 nodes are visible because drawing text is a relatively expensive operation.  This can create some unusual behavior. If the screen currently contains 98 nodes, node labels will be displayed.  If you pan across the network, such that now 101 nodes are displayed, the node labels will disappear.  As another example, if the sum of rendered edges and rendered nodes is greater than or equal to a default value of 2000, a very coarse level of detail is chosen, where edges are always straight lines, nodes are always rectangles, and no antialiasing is done.  The default values used to determine these thresholds can be changed by setting properties under Edit → Preferences → Properties... . </p><p><span class="strong"><strong>Low LOD vs High LOD</strong></span>  </p><div class="table"><a id="d0e6408"/><p class="title"><b>Table 26. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p> Large Network with <span class="strong"><strong>Low</strong></span> LOD </p></td><td><p> Large Network with <span class="strong"><strong>High</strong></span> LOD </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/LowLOD.png" alt="attachment:LowLOD.png"/></span> </p></td><td><p> <span class="inlinemediaobject"><img src="images/HighLOD.png" alt="attachment:HighLOD.png"/></span> </p></td></tr><tr><td><p> <span class="inlinemediaobject"><img src="images/LowLODZoom.png" alt="attachment:LowLODZoom.png"/></span> </p></td><td><p> <span class="inlinemediaobject"><img src="images/HighLODZoom.png" alt="attachment:HighLODZoom.png"/></span> </p></td></tr></tbody></table></div></div><br class="table-break"/><p>With low LOD values, all nodes are displayed as square and anti-alias is turned off.  With high LOD values, anti-alias is turned on and nodes are displayed as actual shape user specified in the Visual Style.  </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Parameters for Controlling LOD"/>Parameters for Controlling LOD</h4></div></div></div><p><span class="strong"><strong>NOTE:</strong></span> <span class="underline">The greater these thresholds become, the slower performance will become.</span>  If you work with small networks (a few hundred nodes), this shouldn't be a problem, but for large networks it will produce noticeable slowing.  The various thresholds are described below. </p><div class="table"><a id="d0e6482"/><p class="title"><b>Table 27. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/><col/></colgroup><tbody><tr><td><p> render.coarseDetailThreshold </p></td><td colspan="2"><p> If the sum of <span class="emphasis"><em>rendered</em></span> nodes and <span class="emphasis"><em>rendered</em></span> edges equals to or exceeds this number, a very coarse level of detail will be chosen and all other detail parameters will be ignored.  If the total number of nodes and edges is below this threshold, anti-alias will be turned on; this value defaults to 2000. </p></td></tr><tr><td><p> render.nodeBorderThreshold </p></td><td colspan="2"><p> If the number of <span class="emphasis"><em>rendered</em></span> nodes equals to or exceeds this number, node borders will not be rendered; this value defaults to 200. </p></td></tr><tr><td><p> render.nodeLabelThreshold </p></td><td colspan="2"><p> If the number of <span class="emphasis"><em>rendered</em></span> nodes equals to or exceeds this number, node labels will not be rendered; this value defaults to 100. </p></td></tr><tr><td><p> render.edgeArrowThreshold </p></td><td colspan="2"><p> If the number of <span class="emphasis"><em>rendered</em></span> edges equals to or exceeds this number, edge arrows will not be rendered; this value defaults to 300. </p></td></tr><tr><td><p> render.edgeLabelThreshold </p></td><td colspan="2"><p> If the number of <span class="emphasis"><em>rendered</em></span> edges equals to or exceeds this number, edge labels will not be rendered; this value defaults to 150. </p></td></tr></tbody></table></div></div><br class="table-break"/><p>When printing networks or exporting to formats such as PostScript, the highest level of detail is always chosen, regardless of what is currently being displayed on the screen. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Force to Display Detail"/>Force to Display Detail</h4></div></div></div><p>If you want to display every detail of the network regardless of LOD values, you can use <span class="strong"><strong>View → Show Graphics Details</strong></span>.  This option force to display every graphics details.  If the network is large, this option slows down rendering speed. </p></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Annotation"/>Annotation</h2></div></div></div><p>Annotations in Cytoscape are stored as a set of ontologies (e.g. the Gene Ontology, or GO). An ontology consists of a set of controlled vocabulary terms that annotate the objects. For example, using the Gene Ontology, the Saccharomyces Cerevisiae <span class="strong"><strong>CDC55</strong></span> gene has a biological process described as “protein biosynthesis”, to which GO has assigned the number 6412 (a GO ID).  </p><p/><pre class="screen">GO 8150 biological_process
 GO 7582 physiological processes
   GO 8152 metabolism
    GO 44238 primary metabolism
      GO 19538 protein metabolism
        GO 6412 protein biosynthesis
</pre><p><span class="strong"><strong>Graphical View of GO Term 6412: protein biosynthesis</strong></span> </p><p><span class="inlinemediaobject"><img src="images/ontology_dag1.png" alt="attachment:ontology_dag1.png"/></span> </p><p>Cytoscape can use this ontology DAG (Directed Acyclic Graph) to annotate objects in networks.  The Ontology Server (originally called "BioDataServer") is a Cytoscape feature which allows you to load, navigate, and assign annotation terms to nodes and edges in a network.  Cytoscape 2.4 now has an enhanced GUI for loading ontology and associated annotation, enabling you to load both local and remote files. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Ontology and Annotation File Format"/>Ontology and Annotation File Format</h3></div></div></div><p>The standard file formats used in Cytoscape Ontology Server are OBO and Gene Association. The GO website details these file formats: </p><div class="itemizedlist"><ul type="disc"><li><p>Ontologies and Definitions:  <a href="http://www.geneontology.org/GO.downloads.shtml#ont" target="_top">http://www.geneontology.org/GO.downloads.shtml#ont</a> </p></li><li><p>Current Annotations: <a href="http://www.geneontology.org/GO.current.annotations.shtml" target="_top">http://www.geneontology.org/GO.current.annotations.shtml</a> </p></li></ul></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="OBO File"/>OBO File</h4></div></div></div><p>An OBO file is the ontology DAG itself.  This file defines the relationships between ontology terms.  Cytoscape 2.4 and onwards can load all ontology files written in OBO format.  The full listing of ontology files are available from the Open Biomedical Ontologies (OBO) website:  </p><div class="itemizedlist"><ul type="disc"><li><p>OBO Ontology Browser: <a href="http://obo.sourceforge.net/browse.html" target="_top">http://obo.sourceforge.net/browse.html</a> </p></li></ul></div><p>Sample OBO File - gene_ontology.obo: <a href="http://www.geneontology.org/ontology/gene_ontology_edit.obo" target="_top">http://www.geneontology.org/ontology/gene_ontology_edit.obo</a> </p><pre class="screen">format-version: 1.2
date: 27:11:2006 17:12
saved-by: midori
auto-generated-by: OBO-Edit 1.002
subsetdef: goslim_generic "Generic GO slim"
subsetdef: goslim_goa "GOA and proteome slim"
subsetdef: goslim_plant "Plant GO slim"
subsetdef: goslim_yeast "Yeast GO slim"
subsetdef: gosubset_prok "Prokaryotic GO subset"
default-namespace: gene_ontology
remark: cvs version: $Revision: 5.49 $

[Term]
id: GO:0000001
name: mitochondrion inheritance
namespace: biological_process
def: "The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton." [GOC:mcc, PMID:10873824, PMID:11389764]
synonym: "mitochondrial inheritance" EXACT []
is_a: GO:0048308 ! organelle inheritance
is_a: GO:0048311 ! mitochondrion distribution

[Term]
id: GO:0000002
name: mitochondrial genome maintenance
namespace: biological_process
def: "The maintenance of the structure and integrity of the mitochondrial genome." [GOC:ai]
is_a: GO:0007005 ! mitochondrion organization and biogenesis
</pre><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Default List of Ontologies"/>Default List of Ontologies</h4></div></div></div><p>Cytoscape provides a list of ontologies available in OBO format.  If an Internet connection is available, Cytoscape will import ontology and annotation files directly from the remote source.  The table below lists the included ontologies. </p><div class="table"><a id="d0e6623"/><p class="title"><b>Table 28. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p> <span class="strong"><strong>Ontology Name</strong></span> </p></td><td><p> <span class="strong"><strong>Description</strong></span> </p></td></tr><tr><td><p>Gene Ontology Full</p></td><td><p>This data source contains a full-size GO DAG, which contains all GO terms.  This OBO file is written in version 1.2 format.</p></td></tr><tr><td><p>Generic GO slim</p></td><td><p> A subset of general GO Terms, including higer-level terms only.</p></td></tr><tr><td><p>Yeast GO slim</p></td><td><p>A subset of GO Terms for annotating Yeast data sets maintained by SGD.</p></td></tr><tr><td><p>Molecule role (INOH Protein name/family name ontology)</p></td><td><p>A structured controlled vocabulary of concrete and abstract (generic) protein names. This ontology is a INOH pathway annotation ontology, one of a set of ontologies intended to be used in pathway data annotation to ease data integration. This ontology is used to annotate protein names, protein family names, and generic/concrete protein names in the INOH pathway data.  INOH is part of the BioPAX working group.</p></td></tr><tr><td><p>Event (INOH pathway ontology)</p></td><td><p>A structured controlled vocabulary of pathway-centric biological processes. This ontology is a INOH pathway annotation ontology, one of a set of ontologies intended to be used in pathway data annotation to ease data integration. This ontology is used to annotate biological processes, pathways, and sub-pathways in the INOH pathway data. INOH is part of the BioPAX working group.</p></td></tr><tr><td><p>Protein-protein interaction</p></td><td><p>A structured controlled vocabulary for the annotation of experiments concerned with protein-protein interactions.</p></td></tr><tr><td><p>Pathway Ontology</p></td><td><p>The Pathway Ontology is a controlled vocabulary for pathways that provides standard terms for the annotation of gene products.</p></td></tr><tr><td><p>PATO</p></td><td><p>PATO is an ontology of phenotypic qualities, intended for use in a number of applications, primarily phenotype annotation.  For more information, please visit the PATO wiki (<a href="http://www.bioontology.org/wiki/index.php/PATO:Main_Page" target="_top">http://www.bioontology.org/wiki/index.php/PATO:Main_Page</a>). </p></td></tr><tr><td><p>Mouse pathology</p></td><td><p>The Mouse Pathology Ontology (MPATH) is an ontology for mutant mouse pathology.  This is Version 1.</p></td></tr><tr><td><p>Human disease</p></td><td><p>This ontology is a comprehensive hierarchical controlled vocabulary for human disease representation.  For more information, please visit the Disease Ontology website (<a href="http://diseaseontology.sourceforge.net/" target="_top">http://diseaseontology.sourceforge.net/</a>).</p></td></tr></tbody></table></div></div><br class="table-break"/><p>Although Cytoscape can import all kinds of ontologies in OBO format, annotation files are associated with specific ontologies. Therefore, you need to provide the correct ontology-specific annotation file to annotate nodes/edges/networks in Cytoscape.  For example, while you can annotate human network data using the GO Full ontology with human Gene Association files, you cannot use a combination of the human Disease Ontology file and human Gene Association files, because the Gene Association file is only compatible with GO. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Visualize and Browse Ontology DAG (for Advanced Users)"/>Visualize and Browse Ontology DAG (for Advanced Users)</h4></div></div></div><p>Relationships between ontology terms are usually represented as Directed Acyclic Graphs (DAGs).  This is a special case of a network (or graph), where nodes are ontology terms and edges are relationships between terms.  Originally, Cytoscape used a special data structure called BioDataServer to store ontology DAGs.  Starting in version 2.4, ontology data will now be stored in the same data structure as normal networks.  This enables users and plugin writers to visualize, browse and manipulate ontology DAGs just like other networks.  The following is an example of visualization of an ontology DAG (Generic GO Slim): </p><p><span class="inlinemediaobject"><img src="images/ontology_dag2.png" alt="attachment:ontology_dag2.png"/></span> </p><p>Every ontology term and relationship can have attributes.  In the OBO files, ontology terms have optional fields such as definition, synonyms, comments, or cross-references.  These fields will be imported as node attributes.  To browse those attributes, please use the attribute browser (see the example below): </p><p><span class="inlinemediaobject"><img src="images/ontology_attrs.png" alt="attachment:ontology_attrs.png"/></span> </p><div class="itemizedlist"><ul type="disc"><li>Note 1: Some ontologies have a lot of terms.  For example, the full Gene Ontology contains more than 20,000 terms.  If you need to save memory, you can remove this ontology DAG from Network Panel (right-click on the ontology name at the left-hand side of the screen and select Destroy Network). </li><li>Note 2: All ontology DAGs will be saved in the session file.  To minimize the session file size, you can delete the Ontology DAG before saving session. </li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Gene Association File"/>Gene Association File</h4></div></div></div><p>The Gene Association (GA) file provides annotation only for the Gene Ontology.  It is a species-specific annotation file for GO terms.  <span class="strong"><strong>Gene Association files will only work with Gene Ontologies and NOT others!</strong></span> </p><p>Sample Gene Association File (gene_association.sgd - annotation file for yeast): </p><p/><pre class="screen">SGD     S000003916      AAD10           GO:0006081      SGD_REF:S000042151|PMID:10572264        ISS             P       aryl-alcohol dehydrogenase (putative)        YJR155W gene    taxon:4932      20020902        SGD
SGD     S000005275      AAD14           GO:0008372      SGD_REF:S000069584      ND              C       aryl-alcohol dehydrogenase (putative)        YNL331C gene    taxon:4932      20010119        SGD
</pre><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Node Name Mapping"/>Node Name Mapping</h3></div></div></div><p>If you have a network file and an attribute file, they should have a common key to map attributes onto network data.  If those two do not have a common key, you need to do an extra step to add a shared key.  The following is a quick tutorial to learn how to use Gene Name Mapping files. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Import New ID Sets from BioMart"/>Import New ID Sets from BioMart</h4></div></div></div><p>From Cytoscape 2.6.0, you can import various kinds of ID sets from BioMart (<a href="http://www.biomart.org/index.html" target="_top">http://www.biomart.org/index.html</a>).  BioMart web service client is available as a set of plugins.  You can install <span class="emphasis"><em>BioMartClient</em></span> and <span class="emphasis"><em>BioMartUserInterface</em></span> plugins from Plugin Manager window. </p><p><span class="inlinemediaobject"><img src="images/id_mapping1.png" alt="attachment:id_mapping1.png"/></span> </p><div class="orderedlist"><ol type="1"><li><p>Select: <span class="strong"><strong>File → Import → Import attributes from Biomart...</strong></span> </p></li><li><p>Select a data source.  For ID mapping, select one of the <span class="emphasis"><em>Ensemble Genes</em></span> data set.  You need to choose correct species for your network. </p></li><li><p>Select <span class="strong"><strong>Attribute</strong></span>.  If you want to import new ID sets matching current node IDs, select <span class="emphasis"><em>ID</em></span>. </p></li><li><p>Select <span class="strong"><strong>Data Type</strong></span>.  This should be the type of ID set selected in <span class="strong"><strong>Attribute</strong></span> list.  For example, if you select <span class="emphasis"><em>ID</em></span> for <span class="strong"><strong>Attribute</strong></span> and your network uses <span class="emphasis"><em>Entrez Gene ID</em></span> for its node ID, you need to select <span class="emphasis"><em>EntrezGene ID(s)</em></span> for <span class="strong"><strong>Data Type</strong></span>. </p></li><li><p>Select new ID sets from the list.  Because BioMart server does not accept query to import lots of annotations at once, you can select only 3-5 attributes for each import. </p></li><li><p>Press <span class="strong"><strong>Import</strong></span>. </p></li></ol></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Import Network and Name Mapping Files"/>Import Network and Name Mapping Files</h4></div></div></div><div class="orderedlist"><ol type="1"><li><p>Download name mapping files.  Mapping files are available at: <a href="http://chianti.ucsd.edu/kono/genenamemapping.html" target="_top">http://chianti.ucsd.edu/kono/genenamemapping.html</a>.  In this tutorial, we are going to use <span class="emphasis"><em>dictionary_no_prefix.zip</em></span>, which is  a file set <span class="strong"><strong>without</strong></span> prefixes for each gene names.  Unzip the archive. </p></li><li><p>Load sample network file.  Open network import dialog from <span class="emphasis"><em><span class="strong"><strong>File--&gt;Import--&gt;Network (multiple file types)...</strong></span></em></span>  Then click URL radio button and import <span class="emphasis"><em><span class="strong"><strong>Human Protein-Protein: Rual et al. (Subnetwork for tutorial)</strong></span></em></span>. </p></li><li><p>Open attribute table import dialog from <span class="emphasis"><em><span class="strong"><strong>File--&gt;Import--&gt;Attribute from Table</strong></span></em></span>. </p></li><li><p>Select <span class="emphasis"><em><span class="strong"><strong>human.dic_cyto.txt</strong></span></em></span> as the input file. </p></li><li><p>Check "Show Text File Import Options<span class="emphasis"><em> and click </em></span>Transfer first line as attribute names<span class="emphasis"><em> checkbox. </em></span></p><span class="ERROR">&lt;listitem&gt;<p>Uncheck "Show Text File Import Options</p> and check <span class="emphasis"><em>Mapping Options</em></span>. &lt;/listitem&gt;</span></li><li><p>Select <span class="strong"><strong>EntrezGene</strong></span> as <span class="strong"><strong>Primary Key</strong></span>. </p></li><li><p>Right-click on <span class="strong"><strong>EntrezGene</strong></span> column name and set the type to <span class="emphasis"><em>String</em></span>. </p></li><li><p>Do the same for <span class="strong"><strong>HGNC</strong></span>. </p></li><li><p>Right-click on <span class="strong"><strong>Other Aliases</strong></span> and select <span class="emphasis"><em>List</em></span> as the data type. </p></li><li><p>Check <span class="strong"><strong>Other Aliases</strong></span> as Alias (under "Alias?" checkboxes). </p></li><li>Now the Table Import dialog looks like the following screenshot: <div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/importdialog1.png" alt="attachment:importdialog1.png"/></span> </p></li></ul></div></li><li><p>Press <span class="emphasis"><em>Import</em></span>.  The network has new names in the text file as attributes. </p><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/nameMapping1.png" alt="attachment:nameMapping1.png"/></span> </p></li></ul></div></li></ol></div><p>At this point, nodes have multiple names including HGNC, <a href="http://cytoscape.org/cgi-bin/moin.cgi/UniProt" target="_top">UniProt</a>, and <a href="http://cytoscape.org/cgi-bin/moin.cgi/EntrezGene" target="_top">EntrezGene</a> ID.  You can import other attribute files using these keys.  These imported names (IDs) are useful when you import GO Annotation.  </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Import Ontology and Annotation"/>Import Ontology and Annotation</h3></div></div></div><p><span class="inlinemediaobject"><img src="images/ontology_and_annotation_Import_main.png" alt="attachment:ontology_and_annotation_Import_main.png"/></span> </p><p>Cytoscape 2.4 provides a graphical user interface to import both ontology and annotation files at the same time. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Import Gene Ontology and Gene Association Files"/>Import Gene Ontology and Gene Association Files</h4></div></div></div><p>For convenience, Cytoscape has a list of URLs for commonly used ontology data and a complete set of Gene Association files.  To import Gene Ontology and Gene Association files for the loaded networks, please follow these steps: </p><p><span class="emphasis"><em><span class="strong"><strong>Important: All data sources in the preset list are remote URLs, meaning a network connection is required!</strong></span></em></span> </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Step 1. Select an Annotation File"/>Step 1. Select an Annotation File</h5></div></div></div><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/ontology_import_annotation.png" alt="attachment:ontology_import_annotation.png"/></span> </p><p>Select File → Import → Ontology and Annotation... to open the "Import Ontology and Annotation" window.  From the Annotation dropdown list, select a gene association file for your network.  For example, if you want to annotate the yeast network, select "Gene Association file for Saccharomyces cerevisiae". </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Step 2. Select an Ontology File"/>Step 2. Select an Ontology File</h5></div></div></div><div class="itemizedlist"><ul type="disc"><li><p><span class="inlinemediaobject"><img src="images/ontology_import_obo.png" alt="attachment:ontology_import_obo.png"/></span> </p></li></ul></div><div class="itemizedlist"><ul type="disc"><li>Select an Ontology data (OBO file) from the Ontology dropdown list.  If the file is not loaded yet, it will be shown in red.  The first three files are Gene Ontology files.  You can load other ontologies, but you need your own annotation file to annotate networks.  </li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Step 3. Import the files"/>Step 3. Import the files</h5></div></div></div><div class="itemizedlist"><ul type="disc"><li>Once you click the Import button, Cytoscape will start loading OBO and Gene Association files from the remote sources.  If you choose GO Full it may take a while since it is a large data file. </li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Step 4."/>Step 4.</h5></div></div></div><div class="itemizedlist"><ul type="disc"><li><p>When Cytoscape finishes importing files, the import window will be automatically closed.  All attributes mapped by this function have the prefix "annotation" and look like this: <code class="code">annotation.[attribute_name]</code>.  All ontologies will be added to the end of the Ontology DAGs branch in the Network Manager.  </p><div class="itemizedlist"><ul type="circle"><li><p><span class="inlinemediaobject"><img src="images/ontology_tree.png" alt="attachment:ontology_tree.png"/></span> </p></li></ul></div></li></ul></div><p>Ontology DAGs have some attributes associated with the terms.  All attributes associated with ontology terms will have the prefix <span class="emphasis"><em>ontology</em></span>.  They have at least one attribute: <code class="code">ontology.name</code>.  For more detailed information about attributes for ontology DAGs, please read the official <a href="http://www.geneontology.org/GO.format.obo-1_2.shtml" target="_top">OBO specification document</a>. </p><div class="itemizedlist"><ul type="disc"><li>Note: Cytoscape supports both OBO formats: version 1.0 and 1.2. </li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a id="Note: Switching Primary Key for Go Annotation Import"/>Note: Switching Primary Key for Go Annotation Import</h5></div></div></div><p>If node IDs in a network file are NOT <span class="emphasis"><em>DB_Object_Symbol</em></span> (3rd column in Gene Association file), you need to select a primary key column.  Click <span class="emphasis"><em>Show Mapping Options</em></span> to change the key.  Usually, <span class="emphasis"><em>DB_Object_ID</em></span> can be an alternative primary key.    </p><p/></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Custom Annotation Files for Ontologies Other than GO (for Advanced Users)"/>Custom Annotation Files for Ontologies Other than GO (for Advanced Users)</h3></div></div></div><div class="itemizedlist"><ul type="disc"><li>The "Import Ontology and Annotation" function is designed to import general ontology and annotation files.  Internally, mapping ontology terms onto existing networks is the same as joining three data tables in Cytoscape.  An Ontology DAG, an annotation file, and network data are used in this process (see the example below). </li></ul></div><p><span class="underline">Network Data</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_net_table.png" alt="attachment:ontology_net_table.png"/></span> </p><p><span class="underline">Ontology Data</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_obo_table.png" alt="attachment:ontology_obo_table.png"/></span> </p><p><span class="underline">Annotation Data</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_ga_table.png" alt="attachment:ontology_ga_table.png"/></span> </p><p><span class="underline">Mapping Result</span> </p><p><span class="inlinemediaobject"><img src="images/ontology_mapping_result.png" alt="attachment:ontology_mapping_result.png"/></span> </p><p>If you want to map ontology terms onto network objects, you need to create a custom annotation file.  The annotation file should contain at least 2 columns: a primary key and an ontology term ID.  The primary key is the value used for mapping between the annotation file and network.  Usually, the node/edge ID is used as the primary key, but you may choose any of the available attributes.  The Ontology term ID is the key used for mapping between the annotation file and the ontology DAG.  Using these data sources, you can annotate network objects in Cytoscape. </p><p>Suppose you have a small network: </p><pre class="screen">node_1 pp node_2
node_3 pp node_1
node_2 pp node_3
</pre><p>and you want to annotate this network with <span class="emphasis"><em>Ontology A</em></span>, which is an ontology DAG available in OBO format.  In this case, you need an annotation table file that looks like this: </p><pre class="screen">node_1  OA_0000232
node_2  OA_0000441
node_3  OA_0000702
</pre><p>where <span class="emphasis"><em>OA_***</em></span> represents an ontology term ID.  This example is a file with the minimum necessary number of columns; however, you can include additional columns that will appear as additional node attributes. </p><p>Some ontologies will be used to annotate edges or networks.  For example, the <a href="http://obo.sourceforge.net/cgi-bin/detail.cgi?psi-mi" target="_top">Protein-protein interaction ontology</a> is a controlled set of terms for annotating interactions between proteins, so ontology terms should be mapped onto edges (see example below). </p><pre class="screen">node_1 (pp) node_2  MI:0445
node_3 (pp) node_1  MI:0046
node_2 (pp) node_3  MI:0346
</pre><p><span class="inlinemediaobject"><img src="images/ontology_import_custom1.png" alt="attachment:ontology_import_custom1.png"/></span> </p><p>The basic operation of the Ontology and Annotation Import function is the same as that of the Attribute Table Import.  The main difference is that you need to specify an additional key for mapping: </p><p><span class="inlinemediaobject"><img src="images/ontology_import_custom2.png" alt="attachment:ontology_import_custom2.png"/></span> </p><p>By selecting a column from the "Key Column in Annotation File" dropdown list,  you can specify the key for mapping between ontology terms and the annotation file. </p><div class="itemizedlist"><ul type="disc"><li>Note: When you load Gene Association files, Cytoscape uses a special loader program designed only for Gene Association files.  Because of this program, all attributes will be named automatically.  Also, ontology IDs will be converted into term names and NCBI taxonomy ID will be converted into actual species name.  However, for custom annotation files, those processes will not be applied.  All ontology terms will be mapped as term IDs. </li></ul></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Linkout"/>Linkout</h2></div></div></div><p>LinkOut provides a mechanism to link nodes and edges to external web resources within Cytoscape. Right-clicking on a node or edge in Cytoscape view opens a popup menu with a list of web links. </p><p>The external links are specified in a <code class="code">linkout.props</code> file which is included in the cytoscape.jar file. The defaults include a number of links such as Entrez, SGD, iHOP, and Google, as well as a number of species-specific links. In addition to the default links, users can customize the LinkOut menu by adding (or removing) links by editing the linkout properties (found under Edit → Preferences → Properties...). </p><p>External links are listed as <span class="emphasis"><em>‘key’-‘value’</em></span> pairs in the <code class="code">linkout.props</code> file where <span class="emphasis"><em>key</em></span> specifies the name of the link and <span class="emphasis"><em>value</em></span> is the search URL. The LinkOut menus are organized in a hierarchical structure that is specified in the key. Linkout key terms specific for nodes start with the keyword <code class="code">nodelinkouturl</code>, for edges this is <code class="code">edgelinkouturl</code>. </p><p>For example, the following entry: </p><p/><pre class="screen">nodelinkouturl.Model Organism DB.SGD (yeast)=http://db.yeastgenome.org/cgi-bin/locus.pl?locus=%ID%
</pre><p>places the SGD link under the yeast submenu. This link will appear in Cytoscape as: </p><p><span class="inlinemediaobject"><img src="images/Figure1_linkout26.png" alt="attachment:Figure1_linkout26.png"/></span> </p><p>In a similar fashion one can added new submenus. </p><p>The <span class="strong"><strong><code class="code">%ID%</code></strong></span> string in the URL is a place-holder for the node label. When the popup menu is generated this marker is substituted with the node label. In the above example, the generated SGD link for the YIM1 protein is: </p><p/><pre class="screen">http://db.yeastgenome.org/cgi-bin/locus.pl?locus\=YIM1
</pre><p>If you want to query based on a different attribute you currently need to specify a different Node Label using the VizMapper. </p><p>For edges the mechanism is much the same; however here the placeholders <span class="strong"><strong><code class="code">%ID1%</code></strong></span> and <span class="strong"><strong><code class="code">%ID2%</code></strong></span> reflect the source and target node label respectively. </p><p>Currently there is no mechanism to check whether the constructed URL query is correct and if the node label is meaningful. Similarly, there is no ID mapping between various identifiers. For example, a link to NCBI Entrez from a network that uses ensembl gene identifiers as node labels will produce a link to Entrez using ensembl ID, which results in an incorrect link. It is the user's responsibility to ensure that the node label that is used as the search term in the URL link will result in a meaningful link. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Adding or Removing Links"/>Adding or Removing Links</h3></div></div></div><p>The default links are defined in a <code class="code">linkout.props</code> file contained in <code class="code">linkout.jar</code>.  These links are normal java properties and can be edited by going to Edit → Preferences → Properties... .  New links can be defined this way as well.  New links can be defined at startup in a separate file and loaded from the command line, either by specifying a file containing the links </p><p/><pre class="screen">cytoscape.sh -P new_linout.props
</pre><p> </p><p>or as individual properties  </p><p/><pre class="screen">cytoscape.sh -P nodelinkouturl.yeast.SGD=http://db.yeastgenome.org/cgi-bin/locus.pl?locus\=%ID%
</pre><p>Any links defined on the command line will supersede the default links. </p><p>To remove a link from the menu simply delete the property, using Edit → Preferences → Properties... . </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Use LinkOut from Attribute Browser"/>Use LinkOut from Attribute Browser</h3></div></div></div><p>From Cytoscape 2.6.0, you can use <a href="http://cytoscape.org/cgi-bin/moin.cgi/LinkOut" target="_top">LinkOut</a> from Attribute Browser.  Basic functionality is the same, and the only difference is the parameter passed to the <a href="http://cytoscape.org/cgi-bin/moin.cgi/LinkOut" target="_top">LinkOut</a> is value in the selected cell. </p><p><span class="inlinemediaobject"><img src="images/Figure2_linkout26.png" alt="attachment:Figure2_linkout26.png"/></span> </p></div></div><div class="itemizedlist"><ul type="disc"><li><a id="acknowledgements"/></li></ul></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Acknowledgements"/>Acknowledgements</h2></div></div></div><p>Cytoscape is built with a number of open source third-party Java libraries. The Cytoscape team gratefully acknowledges the following libraries: </p><div class="itemizedlist"><ul type="disc"><li><p>The Colt Distribution: Open Source Libraries for High Performance Scientific and Technical Computing in Java. Information is available at: <a href="http://hoschek.home.cern.ch/hoschek/colt/" target="_top">http://hoschek.home.cern.ch/hoschek/colt/</a>. </p></li><li><p>Graph INterface librarY a.k.a. GINY. Information is available at: <a href="http://csbi.sourceforge.net/" target="_top">http://csbi.sourceforge.net/</a>. </p></li><li><p>JDOM. Information is available at: <a href="http://www.jdom.org" target="_top">http://www.jdom.org</a>. </p></li><li><p>JUnit. Information is available at: <a href="http://junit.org" target="_top">http://junit.org</a>. </p></li><li><p>JGoodies Looks. Information is available at: <a href="http://www.jgoodies.com/freeware/looks/index.html" target="_top">http://www.jgoodies.com/freeware/looks/index.html</a>. </p></li><li><p>Piccolo. Information is available at: <a href="http://www.cs.umd.edu/hcil/jazz/" target="_top">http://www.cs.umd.edu/hcil/jazz/</a>. </p></li><li><p>Type-Specific Collections Library, from Sosnoski Software Solutions, Inc. Information is available at: <a href="http://www.sosnoski.com/opensrc/tclib/" target="_top">http://www.sosnoski.com/opensrc/tclib/</a>. </p></li><li><p>Xerces Java XML parser. Information is available at: <a href="http://xml.apache.org/xerces-j/" target="_top">http://xml.apache.org/xerces-j/</a>. </p></li><li><p>CLI command line parser. Information is available at: <a href="http://jakarta.apache.org/commons/cli/" target="_top">http://jakarta.apache.org/commons/cli/</a>. </p></li><li><p>FreeHEP library. Information is available at: <a href="http://java.freehep.org" target="_top">http://java.freehep.org</a>. </p></li></ul></div><p>This product includes software developed by the Apache Software Foundation (<a href="http://www.apache.org/" target="_top">http://www.apache.org/</a>). </p><p>This product includes software developed by the JDOM Project (<a href="http://www.jdom.org/" target="_top">http://www.jdom.org/</a>). </p><p>One-step installation of the Cytoscape software is accomplished using the <span class="emphasis"><em>InstallAnywhere</em></span> product from ZeroG Software, Inc. (<a href="http://zerog.com" target="_top">http://zerog.com</a>) </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Appendix A: Old Annotation Server Format"/>Appendix A: Old Annotation Server Format</h2></div></div></div><p>Handlers for the following format still exist in Cytoscape as legacy code, however we <span class="emphasis"><em><span class="strong"><strong>strongly</strong></span></em></span> recommend using the new formats (OBO + Gene Association) described in the previous section, since they are easier to download directly from the Gene Ontology project and use directly.  Currently, users have no access to an import interface for this old format.  </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Building your own annotation files"/>Building your own annotation files</h3></div></div></div><p>The annotation server requires that the gene annotations and associated ontology of controlled vocabulary terms follow a simple format. This simple format was chosen because it is efficient to parse and easy to use. </p><p>The flat file formats are explained below: </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="The Ontology Format"/>The Ontology Format</h4></div></div></div><p>By example (the Gene Ontology - GO): </p><p/><pre class="screen">(curator=GO) (type=all)
0003673 = Gene_Ontology
0003674 = molecular_function [partof: 0003673 ]
0008435 = anticoagulant [isa: 0003674 ]
0016172 = antifreeze [isa: 0003674 ]
0016173 = ice nucleation inhibitor [isa: 0016172 ]
0016209 = antioxidant [isa: 0003674 ]
0045174 = glutathione dehydrogenase (ascorbate) [isa: 0009491 0015038 0016209 0016672 ]
0004362 = glutathione reductase (NADPH) [isa: 0015038 0015933 0016209 0016654 ]
0017019 = myosin phosphatase catalyst [partof: 0017018 ]
...
</pre><p>A second example (KEGG pathway ontology): </p><p/><pre class="screen">(curator=KEGG) (type=Metabolic Pathways)
90001 = Metabolism
80001 = Carbohydrate Metabolism [isa: 90001 ]
80003 = Lipid Metabolism [isa: 90001 ]
80002 = Energy Metabolism [isa: 90001 ]
80004 = Nucleotide Metabolism [isa: 90001 ]
80005 = Amino Acid Metabolism [isa: 90001 ]
80006 = Metabolism of Other Amino Acids [isa: 90001 ]
80007 = Metabolism of Complex Carbohydrates [isa: 90001 ]
...
</pre><p>The format has these required features: </p><div class="itemizedlist"><ul type="disc"><li><p>The first line contains two parenthesized assignments for <code class="code">curator</code> and <code class="code">type</code>. In the GO example above, the ontology file (which is created from the XML that GO provides) nests all three specific ontologies (molecular function, biological process, cellular component) below the 'root' ontology, named 'Gene_Ontology'. <code class="code">(type=all)</code> tells you that all three ontologies are included in that file. </p></li><li>Following the mandatory title line, there are one or more category lines, each with the form: <div class="itemizedlist"><ul type="circle"><li><p><code class="code">number0 = name [isa:|partof: number1 number2 ...]</code> </p></li></ul></div><p>where <code class="code">isa</code> and <code class="code">partof</code> are terms used in GO; they describe the relation between parent and child terms in the ontology hierarchy. </p></li><li><p>The trailing blank before each left square bracket is <span class="emphasis"><em>not</em></span> required; it is an artifact of the python script that creates these files. </p></li></ul></div><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="The Annotation Format"/>The Annotation Format</h4></div></div></div><p>By example (from the GO biological process annotation file): </p><p/><pre class="screen">(species=Saccharomyces cerevisiae) (type=Biological Process) (curator=GO)
YMR056C = 0006854
YBR085W = 0006854
YJR155W = 0006081
...
</pre><p>and from KEGG: </p><p/><pre class="screen">(species=Mycobacterium tuberculosis) (type=Metabolic Pathways) (curator=KEGG)
RV0761C = 10
RV0761C = 71
RV0761C = 120
RV0761C = 350
RV0761C = 561
RV1862 = 10
...
</pre><p>The format has these required features: </p><div class="itemizedlist"><ul type="disc"><li><p>The first line contains three parenthesized assignments, for <code class="code">species</code>, <code class="code">type</code> and <code class="code">curator</code>. In the example just above, the annotation file (created for budding yeast from the flat text file maintained by SGD for the Gene Ontology project and available both at their web site and at GO's) shows three yeast ORFs annotated for biological process with respect to GO, as described above. </p></li><li>Following the mandatory title line, there are one or more annotation lines, each with the form: <div class="itemizedlist"><ul type="circle"><li><p><code class="code">canonicalName = ontologyTermID</code> </p></li></ul></div></li><li><p>Once loaded, this annotation (along with the accompanying ontology) can be assigned to nodes in a Cytoscape network. <span class="strong"><strong>For this to work, the species type of the node must exactly match the species named on the first line of the annotation file. The canonicalName of your node must exactly match the canonicalName present in the annotation file.</strong></span> If you don’t see the expected results when using this feature of Cytoscape, check this again, as getting either of these wrong is a common mistake. </p></li></ul></div><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Load Data into Cytoscape"/>Load Data into Cytoscape</h3></div></div></div><p>The easiest way to make annotations available to Cytoscape is by loading annotations into the Cytoscape annotation server. This is the default behavior for the official release of Cytoscape. </p><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="The Annotation Manifest"/>The Annotation Manifest</h4></div></div></div><p>You must first create a text file to specify the files you want Cytoscape to load. Here is an example, from a file which (for convenience) we usually call <code class="code">manifest</code>: </p><p/><pre class="screen">ontology=goOntology.txt
annotation=yeastBiologicalProcess.txt
annotation=yeastMolecularFunction.txt
annotation=yeastCellularComponent.txt
</pre><p>Use the Cytoscape <code class="code">-b</code> command line argument to specify the annotation manifest file to read (e.g. <code class="code">-b manifest</code>). Please note that the <code class="code">-s</code> switch, which sets the default species for your data, is required to exactly match the species named in any annotation file you wish to use. </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Getting and Reformatting GO Data"/>Getting and Reformatting GO Data</h3></div></div></div><p>The Gene Ontology (GO) project is a valuable source of annotation for the genes of many organisms. In this section we will explain how to: </p><div class="orderedlist"><ol type="1"><li>Obtain the GO ontology file </li><li>Reformat it into the simpler flat file Cytoscape uses </li><li>Obtain an annotation file (we illustrate with yeast and human annotation) </li><li>Reformat the annotation files into the simple Cytoscape format </li></ol></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Obtain the GO ontology file"/>Obtain the GO ontology file</h4></div></div></div><p>Go to the GO XML FTP (<a href="ftp://ftp.geneontology.org/pub/go/xml/" target="_top">ftp://ftp.geneontology.org/pub/go/xml/</a>) page. Download the latest <code class="code">go-YYYYMM-termdb.xml.gz</code> file. </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Reformat GO XML ontology file into a flat file"/>Reformat GO XML ontology file into a flat file</h4></div></div></div><p/><pre class="screen"> gunzip go-YYYYMM-termdb.xml.gz
 python parseGoTermsToFlatFile.py go-YYYYMM-termdb.xml &gt; goOntology.txt
</pre><p>(see below for Python script listing) </p><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a id="Obtain the association file for your organism"/>Obtain the association file for your organism</h4></div></div></div><p>GO maintains a list of association files for many organisms; these files associate genes with GO terms. The next step is to get the file for the organism(s) you are interested in, and parse it into the form Cytoscape needs. A list of files may be seen at <a href="http://www.geneontology.org/GO.current.annotations.shtml" target="_top">http://www.geneontology.org/GO.current.annotations.shtml</a>. The rightmost column contains links to tab-delimited files of gene associations, by species. Choose the species you are interested in, and click 'Download'. </p><p>Let's use "GO Annotations @ EBI: Human" as an example. After you have downloaded and saved the file, look at the first few lines: </p><p/><pre class="screen">SPTR    O00115  DRN2_HUMAN              GO:0003677      PUBMED:9714827  TAS             F       Deoxyribonuclease II precursor  IPI00010348     protein taxon:9606              SPTR
SPTR    O00115  DRN2_HUMAN              GO:0004519      GOA:spkw        IEA             F       Deoxyribonuclease II precursor  IPI00010348     protein taxon:9606        20020425      SPTR
SPTR    O00115  DRN2_HUMAN              GO:0004531      PUBMED:9714827  TAS             F       Deoxyribonuclease II precursor  IPI00010348     protein taxon:9606              SPTR
...
</pre><p>Note that line wrapping has occurred here, so each line of the actual file is wrapped to two lines. The goal is to create from these lines the following lines: </p><p/><pre class="screen">(species=Homo sapiens) (type=Molecular Function) (curator=GO)
IPI00010348 = 0003677
IPI00010348 = 0004519
IPI00010348 = 0004531
...
</pre><p>or </p><p/><pre class="screen">(species=Homo sapiens) (type=Biological Process) (curator=GO)
NP_001366 = 0006259
NP_001366 = 0006915
NP_005289 = 0007186
NP_647593 = 0006899
...
</pre><p>The first sample contains molecular function annotations for proteins, and each protein is identified by its IPI number. IPI is the International Protein Index, which maintains cross references to the main databases for human, mouse and rat proteomes.  The second sample contains biological process annotation, and each protein is identified by its NP (RefSeq) number.  These two naming systems, IPI and RefSeq, are two of many that you can use to define canonical names when you run Cytoscape. For budding yeast, it is much easier: the yeast community always uses standard ORF names, and so Cytoscape uses these as canonical names. For human proteins and genes, there is no single standard. </p><p>The solution (for those working with human genes or proteins) is, once you have downloaded the annotations file, to: </p><div class="orderedlist"><ol type="1"><li>Decide which naming system you want to use. </li><li><p>Download <a href="ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/HUMAN/xrefs.goa" target="_top">ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/HUMAN/xrefs.goa</a>. This cross-reference file, when used strategically, allows you to create Cytoscape-compatible annotation files in which the canonical name is the one most meaningful to you. </p></li><li><p>Examine <code class="code">xrefs.goa</code> to figure out which column contains the names you wish to use. </p></li><li>Make a very slight modification to the python script described below, and then </li><li><p>Run that script, supplying both <code class="code">xrefs.goa</code> and that annotation file as arguments. </p></li></ol></div><p>Here are a few sample lines from <code class="code">xrefs.goa</code>: </p><p/><pre class="screen">SP      O00115  IPI00010348             ENSP00000222219;        NP_001366;              BAA28623;AAC77366;AAC35751;AAC39852;BAB55598;AAB51172;AAH10419; 2960,DNASE2     1777,DNASE2
SP      O00116  IPI00010349             ENSP00000324567;ENSP00000264167;        NP_003650;              CAA70591;       327,AGPS        8540,AGPS
SP      O00124  IPI00010353             ENSP00000265616;ENSP00000322580;        NP_005662;              BAA18958;BAA18959;AAH20694;             7993,D8S2298E
...
</pre><p>Note that line wrapping has occurred here – each line in this example starts with the letters SP. See the README file for more information (<a href="ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/HUMAN/README" target="_top">ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/HUMAN/README</a>). </p><p>Finally, run the script to create your three annotation files for human proteins: </p><div class="itemizedlist"><ul type="disc"><li><p><code class="code">bioproc.anno</code> (GO biological process annotation) </p></li><li><p><code class="code">molfunc.anno</code> (GO molecular function annotation) </p></li><li><p><code class="code">cellcomp.anno</code> (GO cellular component annotation) </p></li></ul></div><p>using the supplied python script. It may be necessary to modify this script slightly if RefSeq identifiers are not used as canonical names or if you are using a more recent version of Python. </p><p/><pre class="screen">python parseAssignmentsToFlatFileFromGoaProject.py gene_association.goa_human xrefs.goa
</pre><p>(See below for Python script listing) </p><p/></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Python script examples"/>Python script examples</h3></div></div></div><p>These scripts, as described above, require Python version 2.2 or later. </p><p>Script 1 - parseGoTermsToFlatFile.py </p><p/><pre class="screen">
# parseGoTermToFlatFile.py:  translate a GO XML ontology file into a simpler
#  Cytoscape flat file
#-----------------------------------------------------------------------------------
# RCS: $Revision: 1.3 $   $Date: 2003/05/18 00:38:43 $
#-----------------------------------------------------------------------------------
import re, pre, sys
#-----------------------------------------------------------------------------------
def flatFilePrint (id, name, isaIDs, partofIDs):
  isa = ''
  if (len (isaIDs) &gt; 0):
    isa = '[isa: '
    for isaID in isaIDs:
      isa += isaID
      isa += ' '
    isa += ']'
  partof = ''
  if (len (partofIDs) &gt; 0):
    partof = '[partof: '
    for partofID in partofIDs:
      partof += partofID
      partof += ' '
    partof += ']'
  result = '~np~%~/np~s = ~np~%~/np~s ~np~%~/np~s ~np~%~/np~s' ~np~%~/np~ (id, name, isa, partof)
  result = result.strip ()
  if (result == 'isa = isa' or result == 'partof = partof'):
    print &gt;&gt; sys.stderr, 'meaningless term: ~np~%~/np~s' ~np~%~/np~ result
  else:
    print result
#-----------------------------------------------------------------------------------
if (len (sys.argv) != 2):
  print 'usage:  ~np~%~/np~s &lt;someFile.xml&gt;' ~np~%~/np~ sys.argv [0]
  sys.exit ();
inputFilename = sys.argv [1];
print &gt;&gt; sys.stderr,  'reading ~np~%~/np~s...' ~np~%~/np~ inputFilename
text = open (inputFilename).read ()
print &gt;&gt; sys.stderr,  'read ~np~%~/np~d characters' ~np~%~/np~ len (text)
regex = '&lt;go:term .*?&gt;(.*?)&lt;/go:term&gt;';
cregex = pre.compile (regex, re.DOTALL)   # . matches newlines
m = pre.findall (cregex, text)
print &gt;&gt; sys.stderr, 'number of go terms: ~np~%~/np~d' ~np~%~/np~ len (m)
regex2 = '&lt;go:accession&gt;GO:(.*?)&lt;/go:accession&gt;.*?&lt;go:name&gt;(.*?)&lt;/go:name&gt;'
cregex2 = re.compile (regex2, re.DOTALL)
regex3 = '&lt;go:isa\s*rdf:resource="http://www.geneontology.org/go#GO:(.*?)"\s*/&gt;'
cregex3 = re.compile (regex3, re.DOTALL)
regex4 = '&lt;go:part-of\s*rdf:resource="http://www.geneontology.org/go#GO:(.*?)"\s*/&gt;'
cregex4 = re.compile (regex4, re.DOTALL)
goodElements = 0
badElements = 0
print '(curator=GO) (type=all)'
for term in m:
  m2 = re.search (cregex2, term)
  if (m2):
    goodElements += 1;
    id = m2.group (1)
    name = m2.group (2)
    isaIDs = []
    m3 = re.findall (cregex3, term);
    for ref in m3:
      isaIDs.append (ref)
    m4 = re.findall (cregex4, term);
    partofIDs = []
    for ref in m4:
      partofIDs.append (ref)
    flatFilePrint (id, name, isaIDs, partofIDs)
  else:
    badElements += 1;
    print &gt;&gt; sys.stderr, 'no match to m2...'
    print &gt;&gt; sys.stderr, "---------------\n~np~%~/np~s\n------------------" ~np~%~/np~ term
print &gt;&gt; sys.stderr,  'goodElements ~np~%~/np~d' ~np~%~/np~ goodElements
print &gt;&gt; sys.stderr, 'badElements ~np~%~/np~d' ~np~%~/np~ badElements
#--------------------------------------
</pre><p>Script 2 - parseAssignmentsToFlatFileFromGoaProject.py </p><p/><pre class="screen">
import sys
#-----------------------------------------------------------------------------------
def fixCanonicalName (rawName):
# for instance, trim 'YBR085W|ANC3' to 'YBR085W'
  bar = rawName.find ('|')
  if (bar &lt; 0):
    return rawName
  return rawName [:bar]
#-----------------------------------------------------------------------------------
def fixGoID (rawID):
  bar = rawID.find (':') + 1
  return rawID [bar:]
#-----------------------------------------------------------------------------------
def readGoaXrefFile (filename):
  lines = open (filename).read().split ('\n')
  result = {}
  for line in lines:
    if (len (line) &lt; 10):
      continue
    tokens = line.split ('\t')
    ipi = tokens [2]
    np = tokens [5]
    semicolon = np.find (';')
    if (semicolon &gt;= 0):
      np = np [:semicolon]
    if (len (ipi) &gt; 0 and len (np) &gt; 0):
      result [ipi] = np
  return result
#-----------------------------------------------------------------------------------
if (len (sys.argv) != 3):
  print 'error!  parse   &lt;gene_associations file from GO&gt; &lt;goa xrefs file&gt; '
  sys.exit ()
associationFilename = sys.argv [1];
xrefsFilename = sys.argv [2]
species = 'Homo sapiens'
ipiToNPHash = readGoaXrefFile (xrefsFilename)
tester = 'IPI00099416'
print 'hash size: ~np~%~/np~d' ~np~%~/np~ len (ipiToNPHash)
print 'test map: ~np~%~/np~s -&gt; NP_054861: ~np~%~/np~s ' ~np~%~/np~ (tester, ipiToNPHash [tester])
bioproc = open ('bioproc.txt', 'w')
molfunc = open ('molfunc.txt', 'w')
cellcomp = open ('cellcomp.txt', 'w')
bioproc.write ('(species=~np~%~/np~s) (type=Biological Process) (curator=GO)\n' ~np~%~/np~ species)
molfunc.write ('(species=~np~%~/np~s) (type=Molecular Function) (curator=GO)\n' ~np~%~/np~ species);
cellcomp.write ('(species=~np~%~/np~s) (type=Cellular Component) (curator=GO)\n' ~np~%~/np~ species);
lines=open(associationFilename).read().split('\n')
sys.stderr.write ('found ~np~%~/np~d lines\n' ~np~%~/np~ len (lines))

for line in lines:
  if (line.find ('!') == 0 or len (line) &lt; 2):
    continue
  tokens = line.split ('\t')
  goOntology = tokens [8]
  goIDraw = tokens [4]
  goID = goIDraw.split (':')[1]
  ipiName = fixCanonicalName (tokens [10])
  if (len (ipiName) &lt; 1):
    continue


  if (not ipiToNPHash.has_key (ipiName)):
    continue
  refseqName = ipiToNPHash [ipiName]
  printName = refseqName
  #printName = ipiName
  if (ipiName == tester):
    print '~np~%~/np~s (~np~%~/np~s) has go term ~np~%~/np~s' ~np~%~/np~ (tester, printName, goID)
  if (goOntology == 'C'):
    cellcomp.write ('~np~%~/np~s = ~np~%~/np~s\n' ~np~%~/np~ (printName, goID))
  elif (goOntology == 'P'):
    bioproc.write ('~np~%~/np~s = ~np~%~/np~s\n' ~np~%~/np~ (printName, goID))
  elif (goOntology == 'F'):
    molfunc.write ('~np~%~/np~s = ~np~%~/np~s\n' ~np~%~/np~ (printName, goID))
#-----------------------------------------------------------------------------------
</pre></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Appendix B: GNU Lesser General Public License"/>Appendix B: GNU Lesser General Public License</h2></div></div></div><p>GNU LESSER GENERAL PUBLIC LICENSE </p><div class="itemizedlist"><ul type="disc"><li>Version 2.1, February 1999 Copyright (C) 1991, 1999 Free Software Foundation, Inc. 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA Everyone is permitted to copy and distribute verbatim copies of this license document, but changing it is not allowed. [This is the first released version of the Lesser GPL.  It also counts as the successor of the GNU Library Public License, version 2, hence the version number 2.1.] </li></ul></div><p>Preamble </p><p>The licenses for most software are designed to take away your freedom to share and change it.  By contrast, the GNU General Public Licenses are intended to guarantee your freedom to share and change free software--to make sure the software is free for all its users. </p><p>This license, the Lesser General Public License, applies to some specially designated software packages--typically libraries--of the Free Software Foundation and other authors who decide to use it.  You can use it too, but we suggest you first think carefully about whether this license or the ordinary General Public License is the better strategy to use in any particular case, based on the explanations below. </p><p>When we speak of free software, we are referring to freedom of use, not price.  Our General Public Licenses are designed to make sure that you have the freedom to distribute copies of free software (and charge for this service if you wish); that you receive source code or can get it if you want it; that you can change the software and use pieces of it in new free programs; and that you are informed that you can do these things. </p><p>To protect your rights, we need to make restrictions that forbid distributors to deny you these rights or to ask you to surrender these rights.  These restrictions translate to certain responsibilities for you if you distribute copies of the library or if you modify it. </p><p>For example, if you distribute copies of the library, whether gratis or for a fee, you must give the recipients all the rights that we gave you.  You must make sure that they, too, receive or can get the source code.  If you link other code with the library, you must provide complete object files to the recipients, so that they can relink them with the library after making changes to the library and recompiling it.  And you must show them these terms so they know their rights. </p><p>We protect your rights with a two-step method: (1) we copyright the library, and (2) we offer you this license, which gives you legal permission to copy, distribute and/or modify the library. </p><p>To protect each distributor, we want to make it very clear that there is no warranty for the free library.  Also, if the library is modified by someone else and passed on, the recipients should know that what they have is not the original version, so that the original author's reputation will not be affected by problems that might be introduced by others. </p><p>Finally, software patents pose a constant threat to the existence of any free program.  We wish to make sure that a company cannot effectively restrict the users of a free program by obtaining a restrictive license from a patent holder.  Therefore, we insist that any patent license obtained for a version of the library must be consistent with the full freedom of use specified in this license. </p><p>Most GNU software, including some libraries, is covered by the ordinary GNU General Public License.  This license, the GNU Lesser General Public License, applies to certain designated libraries, and is quite different from the ordinary General Public License.  We use this license for certain libraries in order to permit linking those libraries into non-free programs. </p><p>When a program is linked with a library, whether statically or using a shared library, the combination of the two is legally speaking a combined work, a derivative of the original library.  The ordinary General Public License therefore permits such linking only if the entire combination fits its criteria of freedom.  The Lesser General  Public License permits more lax criteria for linking other code with  the library.    We call this license the "Lesser" General Public License  because it does Less to protect the user's freedom than the ordinary  General Public License.  It also provides other free software  developers Less of an advantage over competing non-free programs.   These disadvantages are the reason we use the ordinary General Public  License for many libraries.  However, the Lesser license provides  advantages in certain special circumstances.    For example, on rare  occasions, there may be a special need to encourage the widest possible  use of a certain library, so that it becomes a de-facto standard.  To  achieve this, non-free programs must be allowed to use the library.   A more frequent case is that a free library does the same job as widely  used non-free libraries.  In this case, there is little to gain by limiting the  free library to free software only, so we use the Lesser General Public  License.    In other cases, permission to use a particular library in  non-free programs enables a greater number of people to use a large body  of free software.  For example, permission to use the GNU C Library in  non-free programs enables many more people to use the whole GNU  operating system, as well as its variant, the GNU/Linux operating system.     Although the Lesser General Public License is Less protective of the users'  freedom, it does ensure that the user of a program that is linked with the  Library has the freedom and the wherewithal to run that program using a  modified version of the Library.    The precise terms and conditions for  copying, distribution and modification follow.  Pay close attention to the  difference between a "work based on the library" and a "work that uses  the library".  The former contains code derived from the library, whereas  the latter must be combined with the library in order to run.      </p><p>GNU LESSER GENERAL PUBLIC LICENSE    TERMS AND  CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION     </p><p>0. This License Agreement applies to any software library or other program  which contains a notice placed by the copyright holder or other authorized  party saying it may be distributed under the terms of this Lesser General  Public License (also called "this License"). Each licensee is addressed  as "you".    A "library" means a collection of software functions and/or  data prepared so as to be conveniently linked with application programs  (which use some of those functions and data) to form executables.     The "Library", below, refers to any such software library or work  which has been distributed under these terms.  A "work based on the  Library" means either the Library or any derivative work under copyright  law: that is to say, a work containing the Library or a portion of it, either  verbatim or with modifications and/or translated straightforwardly into  another language.  (Hereinafter, translation is included without limitation  in the term "modification".)    "Source code" for a work means the preferred  form of the work for making modifications to it.  For a library, complete  source code means all the source code for all modules it contains, plus any  associated interface definition files, plus the scripts used to control  compilation and installation of the library.    Activities other than copying,  distribution and modification are not covered by this License; they are  outside its scope.  The act of running a program using the Library is not  restricted, and output from such a program is covered only if its contents  constitute a work based on the Library (independent of the use of the  Library in a tool for writing it).  Whether that is true depends on what  the Library does and what the program that uses the Library does.       </p><p>1. You may copy and distribute verbatim copies of the Library's complete  source code as you receive it, in any medium, provided that you  conspicuously and appropriately publish on each copy an appropriate  copyright notice and disclaimer of warranty; keep intact all the notices  that refer to this License and to the absence of any warranty; and distribute a copy of this License along with the Library. </p><p>You may charge a fee for the physical act of transferring a copy, and you may at your option offer warranty protection in exchange for a fee. </p><p>2. You may modify your copy or copies of the Library or any portion of it, thus forming a work based on the Library, and copy and distribute such modifications or work under the terms of Section 1 above, provided that you also meet all of these conditions: </p><p>a) The modified work must itself be a software library. </p><p>b) You must cause the files modified to carry prominent notices stating that you changed the files and the date of any change. </p><p>c) You must cause the whole of the work to be licensed at no charge to all third parties under the terms of this License. </p><p>d) If a facility in the modified Library refers to a function or a table of data to be supplied by an application program that uses the facility, other than as an argument passed when the facility is invoked, then you must make a good faith effort to ensure that, in the event an application does not supply such function or table, the facility still operates, and performs whatever part of its purpose remains meaningful. (For example, a function in a library to compute square roots has a purpose that is entirely well-defined independent of the application.  Therefore, Subsection 2d requires that any application-supplied function or table used by this function must be optional: if the application does not supply it, the square root function must still compute square roots.) </p><p>These requirements apply to the modified work as a whole.  If identifiable sections of that work are not derived from the Library, and can be reasonably considered independent and separate works in themselves, then this License, and its terms, do not apply to those sections when you distribute them as separate works.  But when you distribute the same sections as part of a whole which is a work based on the Library, the distribution of the whole must be on the terms of this License, whose permissions for other licensees extend to the entire whole, and thus to each and every part regardless of who wrote it. </p><p>Thus, it is not the intent of this section to claim rights or contest your rights to work written entirely by you; rather, the intent is to exercise the right to control the distribution of derivative or collective works based on the Library. In addition, mere aggregation of another work not based on the Library with the Library (or with a work based on the Library) on a volume of a storage or distribution medium does not bring the other work under the scope of this License. </p><p>3. You may opt to apply the terms of the ordinary GNU General Public License instead of this License to a given copy of the Library.  To do this, you must alter all the notices that refer to this License, so that they refer to the ordinary GNU General Public License, version 2, instead of to this License.  (If a newer version than version 2 of the ordinary GNU General Public License has appeared, then you can specify that version instead if you wish.)  Do not make any other change in these notices. </p><p>Once this change is made in a given copy, it is irreversible for that copy, so the ordinary GNU General Public License applies to all subsequent copies and derivative works made from that copy. </p><p>This option is useful when you wish to copy part of the code of the Library into a program that is not a library. </p><p>4. You may copy and distribute the Library (or a portion or derivative of it, under Section 2) in object code or executable form under the terms of Sections 1 and 2 above provided that you accompany it with the complete corresponding machine-readable source code, which must be distributed under the terms of Sections 1 and 2 above on a medium customarily used for software interchange. </p><p>If distribution of object code is made by offering access to copy from a designated place, then offering equivalent access to copy the source code from the same place satisfies the requirement to distribute the source code, even though third parties are not compelled to copy the source along with the object code. </p><p>5. A program that contains no derivative of any portion of the Library, but is designed to work with the Library by being compiled or linked with it, is called a "work that uses the Library".  Such a work, in isolation, is not a derivative work of the Library, and therefore falls outside the scope of this License. </p><p>However, linking a "work that uses the Library" with the Library creates an executable that is a derivative of the Library (because it contains portions of the Library), rather than a "work that uses the library".  The executable is therefore covered by this License. Section 6 states terms for distribution of such executables. </p><p>When a "work that uses the Library" uses material from a header file that is part of the Library, the object code for the work may be a derivative work of the Library even though the source code is not. Whether this is true is especially significant if the work can be linked without the Library, or if the work is itself a library.  The threshold for this to be true is not precisely defined by law. </p><p>If such an object file uses only numerical parameters, data structure layouts and accessors, and small macros and small inline functions (ten lines or less in length), then the use of the object file is unrestricted, regardless of whether it is legally a derivative work.  (Executables containing this object code plus portions of the Library will still fall under Section 6.) </p><p>Otherwise, if the work is a derivative of the Library, you may distribute the object code for the work under the terms of Section 6. Any executables containing that work also fall under Section 6, whether or not they are linked directly with the Library itself. </p><p>6. As an exception to the Sections above, you may also combine or link a "work that uses the Library" with the Library to produce a work containing portions of the Library, and distribute that work under terms of your choice, provided that the terms permit modification of the work for the customer's own use and reverse engineering for debugging such modifications. </p><p>You must give prominent notice with each copy of the work that the Library is used in it and that the Library and its use are covered by this License.  You must supply a copy of this License.  If the work during execution displays copyright notices, you must include the copyright notice for the Library among them, as well as a reference directing the user to the copy of this License.  Also, you must do one of these things: </p><p>a) Accompany the work with the complete corresponding machine-readable source code for the Library including whatever changes were used in the work (which must be distributed under Sections 1 and 2 above); and, if the work is an executable linked with the Library, with the complete machine-readable "work that uses the Library", as object code and/or source code, so that the user can modify the Library and then relink to produce a modified executable containing the modified Library.  (It is understood that the user who changes the contents of definitions files in the Library will not necessarily be able to recompile the application to use the modified definitions.) </p><p>b) Use a suitable shared library mechanism for linking with the Library.  A suitable mechanism is one that (1) uses at run time a copy of the library already present on the user's computer system, rather than copying library functions into the executable, and (2) will operate properly with a modified version of the library, if the user installs one, as long as the modified version is interface-compatible with the version that the work was made with. </p><p>c) Accompany the work with a written offer, valid for at least three years, to give the same user the materials specified in Subsection 6a, above, for a charge no more than the cost of performing this distribution. </p><p>d) If distribution of the work is made by offering access to copy from a designated place, offer equivalent access to copy the above specified materials from the same place. </p><p>e) Verify that the user has already received a copy of these materials or that you have already sent this user a copy. </p><p>For an executable, the required form of the "work that uses the Library" must include any data and utility programs needed for reproducing the executable from it.  However, as a special exception, the materials to be distributed need not include anything that is normally distributed (in either source or binary form) with the major components (compiler, kernel, and so on) of the operating system on which the executable runs, unless that component itself accompanies the executable. </p><p>It may happen that this requirement contradicts the license restrictions of other proprietary libraries that do not normally accompany the operating system.  Such a contradiction means you cannot use both them and the Library together in an executable that you distribute. </p><p>7. You may place library facilities that are a work based on the Library side-by-side in a single library together with other library facilities not covered by this License, and distribute such a combined library, provided that the separate distribution of the work based on the Library and of the other library facilities is otherwise permitted, and provided that you do these two things: </p><p>a) Accompany the combined library with a copy of the same work based on the Library, uncombined with any other library facilities.  This must be distributed under the terms of the Sections above. </p><p>b) Give prominent notice with the combined library of the fact that part of it is a work based on the Library, and explaining where to find the accompanying uncombined form of the same work. </p><p>8. You may not copy, modify, sublicense, link with, or distribute the Library except as expressly provided under this License.  Any attempt otherwise to copy, modify, sublicense, link with, or distribute the Library is void, and will automatically terminate your rights under this License.  However, parties who have received copies, or rights, from you under this License will not have their licenses terminated so long as such parties remain in full compliance. </p><p>9. You are not required to accept this License, since you have not signed it.  However, nothing else grants you permission to modify or distribute the Library or its derivative works.  These actions are prohibited by law if you do not accept this License.  Therefore, by modifying or distributing the Library (or any work based on the Library), you indicate your acceptance of this License to do so, and all its terms and conditions for copying, distributing or modifying the Library or works based on it. </p><p>10. Each time you redistribute the Library (or any work based on the Library), the recipient automatically receives a license from the original licensor to copy, distribute, link with or modify the Library subject to these terms and conditions.  You may not impose any further restrictions on the recipients' exercise of the rights granted herein. You are not responsible for enforcing compliance by third parties with this License. </p><p>11. If, as a consequence of a court judgment or allegation of patent infringement or for any other reason (not limited to patent issues), conditions are imposed on you (whether by court order, agreement or otherwise) that contradict the conditions of this License, they do not excuse you from the conditions of this License.  If you cannot distribute so as to satisfy simultaneously your obligations under this License and any other pertinent obligations, then as a consequence you may not distribute the Library at all.  For example, if a patent license would not permit royalty-free redistribution of the Library by all those who receive copies directly or indirectly through you, then the only way you could satisfy both it and this License would be to refrain entirely from distribution of the Library. If any portion of this section is held invalid or unenforceable under any particular circumstance, the balance of the section is intended to apply, and the section as a whole is intended to apply in other circumstances. It is not the purpose of this section to induce you to infringe any patents or other property right claims or to contest validity of any such claims; this section has the sole purpose of protecting the integrity of the free software distribution system which is implemented by public license practices.  Many people have made generous contributions to the wide range of software distributed through that system in reliance on consistent application of that system; it is up to the author/donor to decide if he or she is willing to distribute software through any other system and a licensee cannot impose that choice. </p><p>This section is intended to make thoroughly clear what is believed to be a consequence of the rest of this License. </p><p>12. If the distribution and/or use of the Library is restricted in certain countries either by patents or by copyrighted interfaces, the original copyright holder who places the Library under this License may add an explicit geographical distribution limitation excluding those countries, so that distribution is permitted only in or among countries not thus excluded.  In such case, this License incorporates the limitation as if written in the body of this License. </p><p>13. The Free Software Foundation may publish revised and/or new versions of the Lesser General Public License from time to time. Such new versions will be similar in spirit to the present version, but may differ in detail to address new problems or concerns. Each version is given a distinguishing version number.  If the Library specifies a version number of this License which applies to it and "any later version", you have the option of following the terms and conditions either of that version or of any later version published by the Free Software Foundation.  If the Library does not specify a license version number, you may choose any version ever published by the Free Software Foundation. </p><p>14. If you wish to incorporate parts of the Library into other free programs whose distribution conditions are incompatible with these, write to the author to ask for permission.  For software which is copyrighted by the Free Software Foundation, write to the Free Software Foundation; we sometimes make exceptions for this.  Our decision will be guided by the two goals of preserving the free status of all derivatives of our free software and of promoting the sharing and reuse of software generally. </p><p>NO WARRANTY </p><p>15. BECAUSE THE LIBRARY IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE LIBRARY, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE LIBRARY "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE LIBRARY IS WITH YOU.  SHOULD THE LIBRARY PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION. </p><p>16. IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE LIBRARY AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE LIBRARY (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A FAILURE OF THE LIBRARY TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES. </p><p>END OF TERMS AND CONDITIONS </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="Appendix C: Increasing memory for Cytoscape"/>Appendix C: Increasing memory for Cytoscape</h2></div></div></div><p/><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="How to increase memory for Cytoscape"/>How to increase memory for Cytoscape</h3></div></div></div><p><span class="emphasis"><em>from Notes on memory consumption, Cytoscape User Manual</em></span> </p><p>For users interested in loading large networks, the amount of memory needed by Cytoscape will increase. Memory usage depends on both number of network objects (nodes+edges) and the number of attributes. Here are some rough suggestions for memory allocation: </p><p>Suggested Memory Size Without View </p><div class="table"><a id="d0e7817"/><p class="title"><b>Table 29. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p>Number of Objects (nodes + edges) </p></td><td><p> Suggested Memory Size </p></td></tr><tr><td><p> 0 - 70,000 </p></td><td><p> 512M (default) </p></td></tr><tr><td><p> 70,000 - 150,000 </p></td><td><p> 800M </p></td></tr></tbody></table></div></div><br class="table-break"/><p>Suggested Memory Size With View </p><div class="table"><a id="d0e7846"/><p class="title"><b>Table 30. </b></p><div class="table-contents"><div class="caption"/><table border="1"><colgroup><col/><col/></colgroup><tbody><tr><td><p> Number of Objects (nodes + edges) </p></td><td><p> Suggested Memory Size </p></td></tr><tr><td><p> 0 - 20,000 </p></td><td><p> 512M (default) </p></td></tr><tr><td><p> 20,000 - 70,000 </p></td><td><p> 800M </p></td></tr><tr><td><p> 70,000 - 150,000 </p></td><td><p> 1G </p></td></tr></tbody></table></div></div><br class="table-break"/><p/></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a id="Changing memory allocations on Windows, Mac, and Linux machines"/>Changing memory allocations on Windows, Mac, and Linux machines</h3></div></div></div><p>There are a number of ways to change Cytoscape's memory allocation, depending on your preferred method of opening the application. All of them will change Cytoscape's default memory parameters except starting from the command line. </p><p><span class="strong"><strong>Option A: Command line startup</strong></span> (note: this does not permanently change Cytoscape's default 512M setting) </p><p>If you are opening Cytoscape from the command line using the command </p><div class="itemizedlist"><ul type="disc"><li><p><code class="code">java –Xmx512M –jar cytoscape.jar –p plugins</code> </p></li></ul></div><p>then you can increase the value of –Xmx to the desired amount of memory. For example: </p><div class="itemizedlist"><ul type="disc"><li><p><code class="code">java –Xmx800M –jar cytoscape.jar –p plugins</code> </p></li></ul></div><p><span class="strong"><strong>Option B: Using cytoscape.bat (Windows systems)</strong></span> </p><div class="orderedlist"><ol type="1"><li><p>Open the file <code class="code">cytoscape.bat</code> in a text editor (eg. right-click and select Open With Notepad). </p></li><li>Increase the value of the –Xmx tag (found in the last line of the file), as per Option A. Do not modify other parts of the file. </li><li>Save and close the file. </li><li>Open Cytoscape by double-clicking on cytoscape.bat. </li></ol></div><p><span class="strong"><strong>Option C: Using cytoscape.sh (UNIX, Linux, and Mac OS X systems)</strong></span> </p><div class="orderedlist"><ol type="1"><li><p>Open the file <code class="code">cytoscape.sh</code> in a text editor (eg. right-click and select Open With TextEdit). </p></li><li>Increase the value of the –Xmx tag (found in the last line of the file), as per Option A. Do not modify other parts of the file. </li><li>Save and close the file. </li><li>Open Cytoscape  by running cytoscape.sh from the command line. </li></ol></div><p><span class="strong"><strong>Option D: Using the Cytoscape icon (Mac OS X systems)</strong></span> </p><div class="orderedlist"><ol type="1"><li>In the Finder, right-click on the Cytoscape icon and select Show Package Contents. </li><li><p>Go to the Contents folder and open the file <code class="code">info.plist</code>. </p></li><li>In the Property List Editor, expand the Root directory, then Java, and modify the VMOptions value (originally set as -Xmx512M) as per Option A. Do not modify other parts of the file. </li><li>Save and close the file. </li><li>Open Cytoscape by double-clicking on the icon. </li></ol></div><p><span class="strong"><strong>Option E: Using the Cytoscape icon (Windows systems)</strong></span>  </p><div class="orderedlist"><ol type="1"><li><p>Open the file <code class="code">cytoscape.lax</code> in a text editor. </p></li><li>Increase the numerical value (bytes of memory) of the heap size and stack size lines, shown below at 800M: <div class="itemizedlist"><ul type="disc"><li><pre class="screen">lax.nl.java.option.java.heap.size.max=838860800 
lax.nl.java.option.native.stack.size.max=838860800</pre>Do not modify other parts of the file, and be careful not to add any trailing spaces to these lines. </li></ul></div></li><li>Save and close the file. </li><li>Open Cytoscape by double-clicking on the icon. </li></ol></div></div></div></div></body></html>