<html><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"><title>Building and Storing Interaction Networks</title><meta name="generator" content="DocBook XSL Stylesheets V1.65.1"><link rel="home" href="index.html" title="Cytoscape 2.3 User Manual"><link rel="up" href="index.html" title="Cytoscape 2.3 User Manual"><link rel="previous" href="CommandLineArgumentsandProperties.html" title="Command Line Arguments and Properties"><link rel="next" href="LoadingGeneExpressionAttributeMatrixData.html" title="Loading Gene Expression (Attribute Matrix) Data"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="BuildingandStoringInteractionNetworks"></a>Building and Storing Interaction Networks</h2></div></div><div></div></div><p>Cytoscape reads an interaction network in three ways:  a simple interaction file (SIF or .sif format), a standard format known as Graph Markup Language (GML or .gml format), and an XML standard called XGMML (extensible graph markup and modelling language). SIF specifies nodes and interactions only, while GML and XGMML store additional information about network layout and allows network data exchange with a variety of other network display programs. Typically, SIF is used to import interactions when building a network for the first time, since it is easy to create in a text editor or spreadsheet. Once the interactions have been loaded and layout has been performed, the network may be saved to and subsequently reloaded from GML or XGMML format in future Cytoscape sessions.  SIF, GML, and XGMML are text files and you can edit and view them in a regular text editor. Additionally, GML and XGMML are supported by some other network visualization tools. </p><p>
   <span class="strong">SIF FORMAT: </span>
  </p><p>The simple interaction format is convenient for building a graph from a list of interactions. It also makes it easy to combine different interaction sets into a larger network, or add new interactions to an existing data set. The main disadvantage is that this format does not include any layout information, forcing Cytoscape to re-compute a new layout of the network each time it is loaded. </p><p>Lines in the SIF file specify a source node, a relationship type (or edge type), and one or more target nodes: </p><p></p><pre class="screen">
nodeA &lt;relationship type&gt; nodeB
nodeC &lt;relationship type&gt; nodeA
nodeD &lt;relationship type&gt; nodeE nodeF nodeB
nodeG
...
nodeY &lt;relationship type&gt; nodeZ

  </pre><p>A more specific example is: </p><p></p><pre class="screen">
node1 typeA node2
node2 typeB node3 node4 node5
node0

  </pre><p>The first line identifies two nodes, called node1 and node2, and a single relationship between node1 and node2 of type typeA. The second line specifies three new nodes, node3, node4, and node5; here "node2" refers to the same node as in the first line. The second line also specifies three relationships, all of type typeB and with node2 as the source, with node3, node4, and node5 as the targets, respectively. This second form is simply shorthand for specifying multiple relationships of the same type with the same source node. The third line indicates how to specify a node that has no relationships with other nodes. This form is not needed for nodes that do have relationships, since the specification of the relationship implicitly identifies the nodes as well. Duplicate entries are allowed and indicate multiple edges between the same nodes. For example, the following specifies three edges between the same pair of nodes, two of type pp and one of type pd: </p><p></p><pre class="screen">
node1 pp node2
node1 pp node2
node1 pd node2

  </pre><p>Edges connecting a node to itself (self-edges) are also allowed: </p><p></p><pre class="screen">
node1 pp node1

  </pre><p>Every node and edge in Cytoscape has an identifying name, most commonly used with the node and edge data attribute structures. Node names must be unique as identically named nodes will be treated as identical nodes. The name of each node will be the name in this file by default (unless another string is mapped to display on the node using the visual mapper. This is discussed in the section on visual styles. The name of each edge will be formed from the name of the source and target nodes plus the interaction type: for example, sourceName edgeType targetName. </p><p>The tag &lt;interaction type&gt; should be one of: </p><pre class="screen">
  pp .................. protein &#8211; protein interaction
  pd .................. protein -&gt; DNA
  (e.g. transcription factor binding upstream of a regulating gene.)

  </pre><p>Any text string will work, but the above are the conventions that have been followed thus far. </p><p>Additional interaction types are also possible, but not widely used, e.g.: </p><pre class="screen">
  pr .................. protein -&gt; reaction
  rc .................. reaction -&gt; compound   
  cr .................. compound -&gt; reaction
  gl .................. genetic lethal relationship
  pm .................. protein-metabolite interaction 
  mp .................. metabolite-protein interaction 

  </pre><p>Even whole words or concatenated words may be used to define other types of relationships e.g. geneFusion, cogInference, pullsDown, activates, degrades, inactivates, inhibits, phosphorylates, upRegulates </p><p>
   <span class="strong">Delimiters</span>. Whitespace (space or tab) is used to delimit the names in the simple interaction file format. However, in some cases spaces are desired in a node name or edge type. The standard is that, if the file contains any tab characters, then tabs are used to delimit the fields and spaces are considered part of the name. If the file contains no tabs, then any spaces are delimiters that separate names (and names cannot contain spaces). </p><p>If your network unexpectedly contains no edges and node names that look like edge names, it probably means your file contains a stray tab that's fooling the parser. On the other hand, if your network has nodes whose names are half of a full name, then you probably meant to use tabs to separate node names with spaces. </p><p>Networks in simple interactions format are often stored in files with a ".sif" extension, and Cytoscape recognizes this extension when browsing a directory for files of this type. </p><p>
   <span class="strong">GML FORMAT:</span>
  </p><p>In contrast to SIF, GML is a rich graph format language supported by many other network visualization packages. The GML file format specification is available at: </p><p>
   <a href="http://www.infosun.fmi.uni-passau.de/Graphlet/GML/" target="_top">http://www.infosun.fmi.uni-passau.de/Graphlet/GML/</a>
  </p><p>It is generally not necessary to modify the content of a GML file directly. Once a network is built in SIF format and then laid out, the layout is preserved by saving to and loading from GML. Visual attributes specified in a GML file will result in a new visual style named &#8220;<span class="emphasis"><em>Filename</em></span>.style&#8221; when that GML file is loaded. </p><p>
   <span class="strong">XGMML FORMAT:</span>
  </p><p>XGMML is the XML evolution of GML and is based on the GML definition. The XGMML file format specification is available at: </p><p>
   <a href="http://www.cs.rpi.edu/~puninj/XGMML/" target="_top">http://www.cs.rpi.edu/~puninj/XGMML/</a>
  </p><p>XGMML is now preferred to GML because it offers the flexibility associated with all XML document types.  If you're unsure about which to use, choose XGMML. </p><p>
   <span class="strong">COMMANDS:</span>
  </p><p>Load and save network files using the File menu of Cytoscape. Network files may also be loaded directly from the command line using the &#8211;N option (works for SIF, GML, and XGMML). </p><p>
   <span class="strong">FOR EXAMPLE:</span>
  </p><p>To load a sample molecular interaction network in SIF format, use the menu <span class="underline">File &amp;#8594; Import &amp;#8594; Network</span>.<span class="strong"></span> In the resulting file dialog box, select the file &#8220;sampleData/galFiltered.sif&#8221;. After a few seconds, a small network of 331 nodes should appear in the main window. To load the same interaction network as a GML, use the menu: <span class="underline">File &amp;#8594; Import &amp;#8594; Network</span> again. In the resulting file dialog box, select the file &#8220;sampleData/galFiltered.gml&#8221;. Node and edge attribute files as well as expression data and extra annotation can be loaded as well. </p><p>
   <span class="strong">NODE NAMING ISSUES IN CYTOSCAPE:</span>
  </p><p>Typically, genes are represented by nodes, and interactions (or other biological relationships) are represented by edges between nodes. For compactness, a gene also represents its corresponding protein. Nodes may also be used to represent compounds and reactions (or anything else) instead of genes. </p><p>If a network of genes or proteins is to be integrated with Gene Ontology (GO) annotation or gene expression data, the gene names must exactly match the names specified in the other data files. We strongly encourage naming genes and proteins by their systematic ORF name or standard accession number; common names may be displayed on the screen for ease of interpretation, so long as these are available to the program in the annotation directory or in a node attribute file. Cytoscape ships with all yeast ORF-to-common name mappings in a synonym table within the annotation/ directory. Other organisms will be supported in the future. </p><p>Why do we recommend using standard gene names? All of the external data formats recognized by Cytoscape provide data associated with particular names of particular objects. For example, a network of protein-protein interactions would list the names of the proteins, and the attribute and expression data would likewise be indexed by the name of the object. </p><p>The problem is in connecting data from different data sources that don't necessarily use the same name for the same object. For example, genes are commonly referred to by different names, including a formal "location on the chromosome" identifier and one or more common names that are used by ordinary researchers when talking about that gene. Additionally, database identifiers from every database where the gene is stored may be used to refer to a gene (e.g. protein accession numbers from Swiss-Prot). If one data source uses the formal name while a different data source used a common name or identifier, then Cytoscape must figure out that these two different names really refer to the same biological entity. </p><p>Cytoscape has two strategies for dealing with this naming issue, one simple and one more complex. The simple strategy is to assume that every data source uses the same set of names for every object. If this is the case, then Cytoscape can easily connect all of the different data sources. </p><p>To handle data sources with different sets of names, as is usually the case when manually integrating gene information from different sources, Cytoscape needs a data server that provides synonym information (See<span class="emphasis"><em>12. Annotation.</em></span>). A synonym table gives a canonical name for each object in a given organism and one or more recognized synonyms for that object. Note that the synonym table itself defines what set of names are the "canonical" names. For example, in budding yeast the ORF names are commonly used as the canonical names. </p><p>If a synonym server is available, then by default Cytoscape will convert every name that appears in a data file to the associated canonical name. Unrecognized names will not be changed. This conversion of names to a common set allows Cytoscape to connect the genes present in different data sources, even if they have different names &#8211; as long as those names are recognized by the synonym server. </p><p>For this to work, Cytoscape must also be provided with the species to which the objects belong, since the data server requires the species in order to uniquely identify the object referred to by a particular name. This is usually done in Cytoscape by specifying the species name on the command line with the &#8211;P option (<tt class="code">cytoscape.sh -P "defaultSpeciesName=Saccharomyces cerevisiae"</tt>) or by editing the properties in the Preferences Dialog (<span class="underline">Edit &amp;#8594; Preference...</span>). </p><p>The automatic canonicalization of names can be turned off using the -P option (<tt class="code">cytoscape.sh -P canonicalizeName=false"</tt>) or by editing the properties in the Preferences Dialog (<span class="underline">Edit &amp;#8594; Preferences...</span>).  This canonicalization of names currently does not apply to expression data. Expression data should use the same names as the other data sources or use the canonical names as defined by the synonym table. </p><p></p></div></body></html>
