<html><head>
      <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
   <title>Annotation</title><meta name="generator" content="DocBook XSL Stylesheets V1.70.1"><link rel="start" href="index.html" title="Cytoscape User Manual"><link rel="up" href="index.html" title="Cytoscape User Manual"><link rel="prev" href="ar01s14.html" title="Rendering Engine"><link rel="next" href="ar01s16.html" title="Filters"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="Annotation"></a>Annotation</h2></div></div></div><p>Annotations in Cytoscape are stored as a set of ontologies (e.g. the Gene Ontology - GO). An ontology consists of a set of controlled vocabulary terms that annotate the genes for a given organism and a synonym table for gene names. For example, using the Gene Ontology, the Saccharomyces Cerevisiae <span class="strong"><strong>GAL4</strong></span> gene&#8217;s biological process is described as &#8220;regulation of transcription&#8221;, to which GO has assigned the number 45449 (a GO ID). You can see below that &#8220;regulation of transcription&#8221; is a subcategory of &#8220;transcription&#8221;, which is a subcategory of &#8220;nucleobase, nucleoside, nucleotide and nucleic acid metabolism&#8221;, etc. </p><p></p><pre class="screen">GO 8150 biological_process
 GO 7582 physiological processes
  GO 8152 metabolism
   GO 6139 nucleobase, nucleoside, nucleotide and nucleic acid metabolism
    GO 6350 transcription
     GO 45449 regulation of transcription
</pre><p>Cytoscape can use this &#8216;hierarchical&#8217; ontology to annotate recognized genes if the user chooses a level of the hierarchy to use for a given set of annotations. The ontology provided to Cytoscape does not have to be hierarchical, but if it is not, there is no real advantage to storing annotations this way compared to just storing the annotation terms in a node attribute file. </p><p>The Gene Ontology Server (originally called the biodata server) is a Cytoscape feature which allows you to load, navigate, and assign annotation terms to nodes in a network. </p><p>In version 2.3, Cytoscape has enhanced the <span class="emphasis"><em>Gene Ontology Server Wizard</em></span>, which enables users to select a local ontology and annotation files from the user interface. Although 2.3 remains backwards compatible with the old file formats (the manifest file, which includes a list of <span class="emphasis"><em>anno (annotation)</em></span> and <span class="emphasis"><em>onto (ontology)</em></span> files), we strongly suggest using the new file formats. See Appendix B for a description of how the old file formats can be created and used. </p><p><span class="underline"><span class="strong"><strong>New Ontology and Annotation Files</strong></span></span> </p><p>The standard file formats used in this release are OBO and gene association. The GO website details these file formats: </p><div class="itemizedlist"><ul type="disc"><li><p>Ontologies and Definitions:  <a href="http://www.geneontology.org/GO.downloads.shtml#ont" target="_top">http://www.geneontology.org/GO.downloads.shtml#ont</a> </p></li><li><p>Current Annotations: <a href="http://www.geneontology.org/GO.current.annotations.shtml" target="_top">http://www.geneontology.org/GO.current.annotations.shtml</a> </p></li></ul></div><p>Cytoscape provides a copy of the <span class="emphasis"><em>Gene_ontology.obo</em></span> text file in the <span class="emphasis"><em>annotation</em></span> directory for use with the software. An up-to-date <span class="emphasis"><em>Gene_ontology.obo</em></span> file is maintained by the GO project (updated every 30 minutes). Please download the latest version of this file (from <a href="http://www.geneontology.org" target="_top">http://www.geneontology.org</a>) if you intend to use it during analysis. In addition to the gene ontology file, we also provide gene association files for Yeast and Human (<span class="emphasis"><em>gene_association.sgd</em></span> and <span class="emphasis"><em>gene_association.goa_human</em></span>, both in the <span class="emphasis"><em>annotations</em></span> directory). Other gene association files can be found at the gene ontology web site. </p><p><span class="underline">Sample Files</span> gene_ontology.obo: </p><p></p><pre class="screen">format-version: 1.0
date: 11:08:2005 16:57
saved-by: midori
auto-generated-by: DAG-Edit 1.419 rev 3
default-namespace: gene_ontology
remark: cvs version: $Revision: 1.1 $
subsetdef: goslim_yeast "Yeast GO slim"
subsetdef: goslim_goa "GOA and proteome slim"
subsetdef: goslim_plant "Plant GO slim"
subsetdef: goslim_generic "Generic GO slim"
subsetdef: gosubset_prok "Prokaryotic GO subset"
[Term]
id: GO:0000001
name: mitochondrion inheritance
namespace: biological_process
def: "The distribution of mitochondria\, including the mitochondrial genome\, into daughter cells after mitosis or meiosis\, mediated by interactions between mitochondria and the cytoskeleton." [PMID:10873824, PMID:11389764, SGD:mcc]
is_a: GO:0048308 ! organelle inheritance
is_a: GO:0048311 ! mitochondrion distribution
[Term]
id: GO:0000002
name: mitochondrial genome maintenance
namespace: biological_process
def: "The maintenance of the structure and integrity of the mitochondrial genome." [GO:ai]
is_a: GO:0007005 ! mitochondrion organization and biogenesis
</pre><p>gene_association.sgd (annotation file for yeast): </p><p></p><pre class="screen">SGD     S000003916      AAD10           GO:0006081      SGD_REF:S000042151|PMID:10572264        ISS             P       aryl-alcohol dehydrogenase (putative)        YJR155W gene    taxon:4932      20020902        SGD
SGD     S000005275      AAD14           GO:0008372      SGD_REF:S000069584      ND              C       aryl-alcohol dehydrogenase (putative)        YNL331C gene    taxon:4932      20010119        SGD
</pre><p></p><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Gene Ontology Server Wizard"></a>Gene Ontology Server Wizard</h3></div></div></div><p>The Gene Ontology Server Wizard is a graphical user interface used to select ontology and annotation files. By following the wizard, you do not have to write manifest files, which were used in older versions of Cytoscape to specify the file location of the ontology and gene annotation files. </p><p><span class="underline">Step 1.</span> Select Data Format </p><div class="itemizedlist"><ul type="disc"><li><p>[attachment:ontology_wizard_begin.png] </p></li></ul></div><p>In this panel you can select two types of data formats: Cytoscape <a href="http://cytoscape.org/cgi-bin/moin.cgi/BioDataServer" target="_top">BioDataServer</a> or Gene Ontology. Select <span class="emphasis"><em>Gene Ontology</em></span> to use the new formats (obo and gene_association files). If you want to use the old file formats, select <span class="emphasis"><em>Cytoscape <a href="http://cytoscape.org/cgi-bin/moin.cgi/BioDataServer" target="_top">BioDataServer</a></em></span>. A description of how the old formats can be created and used can be found in Appendix B. </p><p><span class="underline">Step 2.</span> Select OBO File </p><p>[attachment:ontology_wizard_obo.png] If you choose <span class="emphasis"><em>Gene Ontology </em></span>in the first step, this panel appears to help you first select an OBO (ontology) file. The wizard displays the header information of the OBO file loaded.  Once an OBO file is loaded, the <span class="emphasis"><em>Next</em></span> button becomes enabled.  Click the <span class="emphasis"><em>Next</em></span> button to proceed. </p><p><span class="underline">Step 3.</span> Select Gene Association File </p><p>[attachment:ontology_wizard_goa.png] </p><p>You can add a gene_association file to the list by pressing <span class="emphasis"><em>Add</em></span> button. Once you select a file, the file name will be shown in the list in the left. If you click on the file name, the header information will be displayed in the center window. You can select multiple gene_association files. </p><p>The "Transfer GO Species Name to Nodes" can be used to apply the species name (taxon) found in the selected Gene Association file to the nodes of the currently selected network.  This may be an important feature if the nodes of your network do not have a species attribute set.  You can can see the current network's nodes and species attribute for the nodes in the left hand window.  You can see the GO symbols, terms, aspects, synonyms, and taxons (species), in the bottom window.  For nodes in your network to map to GO terms, either a symbol or synonym of a term must match the node ID. </p><p><span class="underline">Step 4.</span> </p><p>Once you have finished specifying the options, click on finish. The OBO file and gene_association files specified in this panel will be automatically loaded. This will then open up the Ontology Mapper dialog. </p><p>[attachment:ontology_mapper.png] </p><p>The Ontology Mapper is used to reduce the number of node attributes created from the gene ontology.  Selecting a term in the left hand column and clicking the <span class="strong"><strong>&gt;&gt;</strong></span> button will select that term to create an attribute for.  You can unselect a term using the <span class="strong"><strong>&lt;&lt;</strong></span> button.  The <span class="emphasis"><em><span class="strong"><strong>Apply All</strong></span></em></span> button will select all terms for attribute creation. In most cases, <span class="emphasis"><em><span class="strong"><strong>Apply All</strong></span></em></span> should be used to import all GO Terms for the network.  Otherwise, only subset of GO Terms in the tree is imported.  Once the desired terms have been selected, click the <span class="strong"><strong>OK</strong></span> button.  This will create node attributes for all selected GO terms.  If you select some GO Terms in the right-side  window, nodes with the GO Terms will be shown in the Node <a href="http://cytoscape.org/cgi-bin/moin.cgi/AttributeBrowser" target="_top">AttributeBrowser</a>.  By right-clicking on a GO Term in the <span class="emphasis"><em><span class="strong"><strong>GO Data as Attributes</strong></span></em></span> window, you can search the GO Term in AmiGO (this function requires internet connection and a web browser).  The GO terms will now be visibile in the Node Attribute Browser by clicking on the Select button ( [attachment:attributes_select_icon.png] ). </p><p>If you would like to choose deifferent GO terms from the same gene association file, you do not need to re-import the files. Instead, you can click the GO button ( [attachment:attributes_go_icon.png] ) in the Attribute Browser panel or on the <span class="underline">File &#8594; Import &#8594; Ontology &#8594; Ontology Mapper</span> menu and the Ontology Mapper dialog will open. </p><p><span class="strong"><strong>Note 1:</strong></span> The <span class="underline">OBO file and gene_association files MUST be in the same directory</span>, and you must have write permission for the directory. </p><p><span class="strong"><strong>Note 2:</strong></span> Please make sure that your network file uses exact same ID&#8217;s used in the gene association file (Symbol or Synonym column). Otherwise, the annotations will not appear in the Annotation Panel. </p></div></div></body></html>