<html><head>
      <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
   <title>Supported Network File Formats</title><meta name="generator" content="DocBook XSL Stylesheets V1.70.1"><link rel="start" href="index.html" title="Cytoscape User Manual"><link rel="up" href="index.html" title="Cytoscape User Manual"><link rel="prev" href="ar01s07.html" title="Creating Networks"><link rel="next" href="ar01s09.html" title="Node and Edge Attributes"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="Supported Network File Formats"></a><a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/Cytoscape_User_Manual/Network_Formats" target="_top">Supported Network File Formats</a></h2></div></div></div><p>Cytoscape can read network/pathway files written in the following formats: </p><div class="itemizedlist"><ul type="disc"><li><p>Simple interaction file (SIF or .sif format) </p></li><li><p>Nested network format (NNF or .nnf format) </p></li><li><p>Graph Markup Language (GML or .gml format) </p></li><li><p>XGMML (extensible graph markup and modelling language). </p></li><li><p>SBML </p></li><li><p>BioPAX </p></li><li><p>PSI-MI Level 1 and 2.5 </p></li><li><p>Delimited text </p></li><li><p>Excel Workbook (.xls) </p></li></ul></div><p>The SIF format specifies nodes and interactions only, while other formats store additional information about network layout and allow network data exchange with a variety of other network programs and data sources. Typically, SIF files are used to import interactions when building a network for the first time, since they are easy to create in a text editor or spreadsheet. Once the interactions have been loaded and network layout has been performed, the network may be saved to GML or XGMML format for interaction with other systems. All file types listed (except Excel) are text files and you can edit and view them in a regular text editor. </p><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="SIF Format"></a>SIF Format</h3></div></div></div><p>The simple interaction format is convenient for building a graph from a list of interactions. It also makes it easy to combine different interaction sets into a larger network, or add new interactions to an existing data set. The main disadvantage is that this format does not include any layout information, forcing Cytoscape to re-compute a new layout of the network each time it is loaded. </p><p>Lines in the SIF file specify a source node, a relationship type (or edge type), and one or more target nodes: </p><pre class="screen">nodeA &lt;relationship type&gt; nodeB
nodeC &lt;relationship type&gt; nodeA
nodeD &lt;relationship type&gt; nodeE nodeF nodeB
nodeG
...
nodeY &lt;relationship type&gt; nodeZ</pre><p>A more specific example is: </p><pre class="screen">node1 typeA node2
node2 typeB node3 node4 node5
node0</pre><p>The first line identifies two nodes, called node1 and node2, and a single relationship between node1 and node2 of type typeA. The second line specifies three new nodes, node3, node4, and node5; here "node2" refers to the same node as in the first line. The second line also specifies three relationships, all of type typeB and with node2 as the source, with node3, node4, and node5 as the targets. This second form is simply shorthand for specifying multiple relationships of the same type with the same source node. The third line indicates how to specify a node that has no relationships with other nodes. This form is not needed for nodes that do have relationships, since the specification of the relationship implicitly identifies the nodes as well. </p><p>Duplicate entries are ignored. Multiple edges between the same nodes must have different edge types.  For example, the following specifies two edges between the same pair of nodes, one of type xx and one of type yy: </p><pre class="screen">node1 xx node2
node1 xx node2
node1 yy node2</pre><p>Edges connecting a node to itself (self-edges) are also allowed: </p><pre class="screen">node1 xx node1</pre><p>Every node and edge in Cytoscape has an identifying name, most commonly used with the node and edge data attribute structures. Node names must be unique, as identically named nodes will be treated as identical nodes. The name of each node will be the name in this file by default (unless another string is mapped to display on the node using the visual mapper). This is discussed in the section on <a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/Cytoscape_User_Manual/Visual_Styles#" target="_top">visual styles</a>. The name of each edge will be formed from the name of the source and target nodes plus the interaction type: for example, <code class="code">sourceName (edgeType) targetName</code>. </p><p>The tag &lt;relationship type&gt; can be any string.  Whole words or concatenated words may be used to define types of relationships, e.g. geneFusion, cogInference, pullsDown, activates, degrades, inactivates, inhibits, phosphorylates, upRegulates, etc. </p><p>Some common interaction types used in the Systems Biology community are as follows: </p><pre class="screen">  pp .................. protein &#8211; protein interaction
  pd .................. protein -&gt; DNA
  (e.g. transcription factor binding upstream of a regulating gene.)</pre><p>Some less common interaction types used are: </p><pre class="screen">  pr .................. protein -&gt; reaction
  rc .................. reaction -&gt; compound
  cr .................. compound -&gt; reaction
  gl .................. genetic lethal relationship
  pm .................. protein-metabolite interaction
  mp .................. metabolite-protein interaction</pre><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Delimiters"></a>Delimiters</h4></div></div></div><p>Whitespace (space or tab) is used to delimit the names in the simple interaction file format. However, in some cases spaces are desired in a node name or edge type. The standard is that, if the file contains any tab characters, then tabs are used to delimit the fields and spaces are considered part of the name. If the file contains no tabs, then any spaces are delimiters that separate names (and names cannot contain spaces). </p><p>If your network unexpectedly contains no edges and node names that look like edge names, it probably means your file contains a stray tab that's fooling the parser. On the other hand, if your network has nodes whose names are half of a full name, then you probably meant to use tabs to separate node names with spaces. </p><p>Networks in simple interactions format are often stored in files with a <code class="code">.sif</code> extension, and Cytoscape recognizes this extension when browsing a directory for files of this type. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="NNF"></a>NNF</h3></div></div></div><p>The NNF format is a very simple format that unlike SIF allows the optional assignment of single nested network per node.  No other node attributes can be specified.  There are only 2 possible line formats: </p><div class="itemizedlist"><ul type="disc"><li><p>A node "node" contained in a "network:" </p></li></ul></div><p><code class="code">network node</code> </p><div class="itemizedlist"><ul type="disc"><li><p>2 nodes linked together contained in a network: </p></li></ul></div><p><code class="code">network node1 interaction node2</code> </p><p>If a network name (first entry on a line) appeared previously as a node name (in columns 2 or 4), the network will be nested in the node with the same name.  Also, if a name that has been previously defined as a network (by being listed in the first column), later appears as a node name (in columns 2 or 4), the previously defined network will be nested in the node with the same name.  In summary: any time a name is used as both, a network name , and a node name, this implies that the network will be nested in the node of the same name.  Additionally comments may be included on all lines.  Comments start with a hash mark '#' and continue to the end of a line.  Trailing comments (after data lines) and entirely blank lines anywhere are also permissible.  Please <span class="strong"><strong>note</strong></span> that if you load multiple NNF files in Cytoscape they will be treated like a single, long concatenated NNF file! </p><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Examples"></a>Examples</h4></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a name="Example 1"></a>Example 1</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/NNFExample1.png" alt="NNFExample1.png"></span> </p><pre class="screen">Example_1      C
Example_1      network1
network1       A        pp        B
network1       B        pp        A
Example_1      C        pp        B</pre></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a name="Example 2"></a>Example 2</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/NNFExample2.png" alt="NNFExample2.png"></span> </p><pre class="screen">Example_2      M1
Example_2      M2
M1             A
M2             B        pp        C
Example_2      A        pp        B
Example_2      M1       im        M2</pre></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a name="Example 3"></a>Example 3</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/NNFExample3.png" alt="NNFExample3.png"></span> </p><pre class="screen">Example_3      M1       im        M2
Example_3      M3       im        M1
Example_3      M2       im        M3
Example_3      C        pp        M3
Example_3      M2       pp        C
M1             A
M2             A        pp        B
M3             B        pp        C</pre></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a name="Example 4"></a>Example 4</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/NNFExample4.png" alt="NNFExample4.png"></span> </p><pre class="screen">Example_4      M1
root           M3
M1             A        pp        B
M1             B        pp        A
Example_4      C        pp        B
M3             M2
M2             D
M3             E        pp        F
M3             D        pp        F
M3             D        pp        E
Example_4      D        pp        C
Example_4      A        pp        M2
Example_4      B        pp        M3
Example_4      M2       pp        B</pre></div><div class="section" lang="en"><div class="titlepage"><div><div><h5 class="title"><a name="Example 5"></a>Example 5</h5></div></div></div><p><span class="inlinemediaobject"><img src="images/NNFExample5.png" alt="NNFExample5.png"></span> </p><pre class="screen">Example_5      M4
M4             D
M4             M3
M3             M2        pp        C
M2             M1        pp        B
M1             A
M4             C         pp        D</pre></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="GML Format"></a>GML Format</h3></div></div></div><p>In contrast to SIF, GML is a rich graph format language supported by many other network visualization packages. The GML file format specification is available at: </p><p><a href="http://www.infosun.fmi.uni-passau.de/Graphlet/GML/" target="_top">http://www.infosun.fmi.uni-passau.de/Graphlet/GML/</a> </p><p>It is generally not necessary to modify the content of a GML file directly. Once a network is built in SIF format and then laid out, the layout is preserved by saving to and loading from GML. Visual attributes specified in a GML file will result in a new visual style named <code class="code">Filename.style</code> when that GML file is loaded. </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="XGMML Format"></a>XGMML Format</h3></div></div></div><p>XGMML is the XML evolution of GML and is based on the GML definition.  In addition to network data, XGMML contains node/edge/network attributes.  The XGMML file format specification is available at: </p><p><a href="http://www.cs.rpi.edu/~puninj/XGMML/" target="_top">http://www.cs.rpi.edu/~puninj/XGMML/</a> </p><p>XGMML is now preferred to GML because it offers the flexibility associated with all XML document types. If you're unsure about which to use, choose XGMML. </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="SBML (Systems Biology Markup Language) Format"></a>SBML (Systems Biology Markup Language) Format</h3></div></div></div><p>The Systems Biology Markup Language (SBML) is an XML format to describe biochemical networks.  SBML file format specification is available at: </p><p><a href="http://sbml.org/documents/" target="_top">http://sbml.org/documents/</a> </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="BioPAX (Biological PAthways eXchange) Format"></a>BioPAX (Biological PAthways eXchange) Format</h3></div></div></div><p>BioPAX is an OWL (Web Ontology Language) document designed to exchange biological pathways data.  The complete set of documents for this format is available at: </p><p><a href="http://www.biopax.org/index.html" target="_top">http://www.biopax.org/index.html</a> </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="PSI-MI Format"></a>PSI-MI Format</h3></div></div></div><p>The PSI-MI format is a data exchange format for protein-protein interactions.  It is an XML format used to describe PPI and associated data.  PSI-MI XML format specification is available at: </p><p><a href="http://psidev.sourceforge.net/mi/xml/doc/user/" target="_top">http://psidev.sourceforge.net/mi/xml/doc/user/</a> </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Delimited Text Table and Excel Workbook"></a>Delimited Text Table and Excel Workbook</h3></div></div></div><p>Cytoscape has native support for Microsoft Excel files (.xls) and delimited text files.  The tables in these files can have network data and edge attributes.  Users can specify columns containg source nodes, target nodes, interaction types, and edge attributes during file import.  Some of the other network analysis tools, such as igraph (<a href="http://cneurocvs.rmki.kfki.hu/igraph/" target="_top">http://cneurocvs.rmki.kfki.hu/igraph/</a>), has feature to export graph as simple text files.  Cytoscape can read these text files and build networks from them.  For more detail, please read the Import Free-Format Tables section section of the <a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/Cytoscape_User_Manual/Creating_Networks#" target="_top">Creating Networks</a> chapter. </p><p><span class="inlinemediaobject"><img src="images/huge_network_igraph.png" alt="huge_network_igraph.png"></span> </p><p><span class="strong"><strong>Network generated by igraph's Watts-Strogatz small-world model (50k nodes and 250k esges) visualized by Cytoscape:</strong></span>  You can import networks created by other applications using this Table Import feature. </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Node Naming Issues in Cytoscape"></a>Node Naming Issues in Cytoscape</h3></div></div></div><p>Typically, genes are represented by nodes, and interactions (or other biological relationships) are represented by edges between nodes. For compactness, a gene also represents its corresponding protein. Nodes may also be used to represent compounds and reactions (or anything else) instead of genes. </p><p>If a network of genes or proteins is to be integrated with Gene Ontology (GO) annotation or gene expression data, the gene names must exactly match the names specified in the other data files. We strongly encourage naming genes and proteins by their systematic ORF name or standard accession number; common names may be displayed on the screen for ease of interpretation, so long as these are available to the program in the annotation directory or in a node attribute file. Cytoscape ships with all yeast ORF-to-common name mappings in a synonym table within the <code class="code">annotation/</code> directory. Other organisms will be supported in the future. </p><p>Why do we recommend using standard gene names? All of the external data formats recognized by Cytoscape provide data associated with particular names of particular objects. For example, a network of protein-protein interactions would list the names of the proteins, and the attribute and expression data would likewise be indexed by the name of the object. </p><p>The problem is in connecting data from different data sources that don't necessarily use the same name for the same object. For example, genes are commonly referred to by different names, including a formal "location on the chromosome" identifier and one or more common names that are used by ordinary researchers when talking about that gene. Additionally, database identifiers from every database where the gene is stored may be used to refer to a gene (e.g. protein accession numbers from Swiss-Prot). If one data source uses the formal name while a different data source used a common name or identifier, then Cytoscape must figure out that these two different names really refer to the same biological entity. </p><p>Cytoscape has two strategies for dealing with this naming issue, one simple and one more complex. The simple strategy is to assume that every data source uses the same set of names for every object. If this is the case, then Cytoscape can easily connect all of the different data sources. </p><p>To handle data sources with different sets of names, as is usually the case when manually integrating gene information from different sources, Cytoscape needs a data server that provides synonym information (see the chapter on <a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/Cytoscape_User_Manual/Annotation#" target="_top">Annotation</a>). A synonym table gives a canonical name for each object in a given organism and one or more recognized synonyms for that object. Note that the synonym table itself defines which set of names are the "canonical" names. For example, in budding yeast, the ORF names are commonly used as the canonical names. </p><p>If a synonym server is available, then by default Cytoscape will convert every name that appears in a data file to the associated canonical name. Unrecognized names will not be changed. This conversion of names to a common set allows Cytoscape to connect the genes present in different data sources, even if they have different names &#8211; as long as those names are recognized by the synonym server. </p><p>For this to work, Cytoscape must also be provided with the species to which the objects belong, since the data server requires the species in order to uniquely identify the object referred to by a particular name. This is usually done in Cytoscape by specifying the species name on the command line with the &#8211;P option (<code class="code">cytoscape.sh -P "defaultSpeciesName=Saccharomyces cerevisiae"</code>) or by editing the properties (under Edit &#8594; Preferences &#8594; Properties...). </p><p>The automatic canonicalization of names can be turned off using the -P option (<code class="code">cytoscape.sh -P canonicalizeName=false"</code>) or by editing the properties (under Edit &#8594; Preferences &#8594; Properties...). This canonicalization of names currently does not apply to expression data. Expression data should use the same names as the other data sources or use the canonical names as defined by the synonym table. </p></div></div></body></html>