<html><head>
      <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
   <title>Importing Networks and Attributes from External Databases</title><meta name="generator" content="DocBook XSL Stylesheets V1.70.1"><link rel="start" href="index.html" title="Cytoscape User Manual"><link rel="up" href="index.html" title="Cytoscape User Manual"><link rel="prev" href="ar01s10.html" title="Loading Gene Expression (Attribute Matrix) Data"><link rel="next" href="ar01s12.html" title="Navigation and Layout"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="Importing Networks and Attributes from External Databases"></a><a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/Cytoscape_User_Manual/ImportingNetworksFromWebServices" target="_top">Importing Networks and Attributes from External Databases</a></h2></div></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Web Service Client Manager"></a>Web Service Client Manager</h3></div></div></div><p>Cytoscape 2.6.0 has a new feature called <span class="strong"><strong>Web Service Client Manager</strong></span>.  This is a framework to manage various kinds of web service clients in Cytoscape.  By using web service clients, users can access remote data sources easily. </p><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="What is a Web Service?"></a>What is a Web Service?</h4></div></div></div><p>A web service is a standardized, platform-independent mechanism for machines to interact over the network.  These days, many major biological databases publish their data with web service API: </p><div class="itemizedlist"><ul type="disc"><li><p>List of Biological Web Services:  <a href="http://taverna.sourceforge.net/services" target="_top">http://taverna.sourceforge.net/services</a> </p></li><li><p>Web Services at the EBI:  <a href="http://www.ebi.ac.uk/Tools/webservices/" target="_top">http://www.ebi.ac.uk/Tools/webservices/</a> </p></li></ul></div><p>This enables developers to write a program to access these services.  Cytoscape core developer team have developed several sample web service clients using this framework.  Currently, Cytoscape supports the following web services: </p><div class="itemizedlist"><ul type="disc"><li><p><a href="http://www.ebi.ac.uk/intact/site/index.jsf" target="_top">IntAct</a>:  an open source database of protein interaction data, hosted at EMBL-EBI. </p></li><li><p><a href="http://www.pathwaycommons.org" target="_top">Pathway Commons</a>: an open source portal, providing access to multiple integrated data sets, including:  Reactome, IntAct, HPRD, HumanCyc, MINT, the MSKCC Cancer Cell Map, and the NCI/Nature Pathway Interaction database. </p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=gene" target="_top">NCBI Entrez Gene</a>: a public database of genes, including annotation, sequence and interactions. </p></li><li><p><a href="http://www.biomart.org/" target="_top">Biomart</a>: an open source biological database engine.  Useful for ID/Name mapping. </p></li></ul></div><p>All of these clients are available as Plugins and users can install them through Plugin Manager. </p><p>In the following sections, users learn how to import network from extrenal databases. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Getting Started"></a>Getting Started</h3></div></div></div><p>To get started, select: File &#8594; Import &#8594; Network from web services... </p><p><span class="inlinemediaobject"><img src="images/file_import.png" alt="file_import.png"></span> </p><div class="informaltable"><table border="1"><colgroup><col></colgroup><tbody><tr><td><p> <span class="strong"><strong>Tip:</strong></span>  View the <a href="http://cbio.mskcc.org/~cerami/cytoscape/CytoWebServices.mov" target="_top">animation demo</a> for importing networks from web services. </p></td></tr></tbody></table></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Example #1:  Retrieving Protein-Protein Interaction Networks from IntAct"></a>Example #1:  Retrieving Protein-Protein Interaction Networks from IntAct</h3></div></div></div><div class="itemizedlist"><ul type="disc"><li><p>Select: File &#8594; Import &#8594; Network from web services... </p></li><li><p>From the pull-down menu, select the IntAct Web Service Client. </p></li><li><p>Enter one or more search terms, such as BRCA1 </p></li><li><p>Click the Search button. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/intact_import.png" alt="intact_import.png"></span> </p><p>After confirming the download of interaction data, the network of BRCA1 will be imported and visualized. </p><p><span class="inlinemediaobject"><img src="images/node_context2.png" alt="node_context2.png"></span> </p><p><span class="strong"><strong> Tip:  Expanding the Network:</strong></span>  Several of the Cytoscape web services provide additional options in the node context menu.  To access these options, right-click on a node and select "Use Web Services."  For example, in the screenshot to the right, we have loaded the BRCA1 network from IntAct, and have chosen to merge this node's neighbors into the existing network. </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Example #2:  Retrieving Protein-Protein Interaction Networks from NCBI Entrez Gene"></a>Example #2:  Retrieving Protein-Protein Interaction Networks from NCBI Entrez Gene</h3></div></div></div><p>An entry of NCBI Entrez Gene has a section called <span class="emphasis"><em>Interactions</em></span>.  NCBI web service client uses this section to build networks. </p><div class="itemizedlist"><ul type="disc"><li><p>Select: <span class="strong"><strong>File &#8594; Import &#8594; Network from web services...</strong></span> </p></li><li><p>From the pull-down menu, select the <span class="strong"><strong>NCBI Web Service Client</strong></span>. </p></li><li><p>Enter free-keywords.  For example, type <span class="emphasis"><em>human muscular dystrophy</em></span>. </p></li><li><p>Click the Search button. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/entrez_import.png" alt="entrez_import.png"></span> </p><p><span class="strong"><strong> Network generated from Entrez Gene data:</strong></span> The network above is generated from interaction data matching the keyword <span class="emphasis"><em>human muscular dystrophy</em></span>.  Edge color represents data source type (BIND, BioGRID, or HPRD). </p><p><span class="strong"><strong>Note: since NCBI client extracts interaction data from a huge dataset, it takes a long time (30 seconds - 5 minutes, depends on machine specifications and network connection) to import large set of interactions. </strong></span> </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Example #3:  Retrieving Pathways and Networks from Pathway Commons"></a>Example #3:  Retrieving Pathways and Networks from Pathway Commons</h3></div></div></div><div class="itemizedlist"><ul type="disc"><li><p>Select: File &#8594; Import &#8594; Network from web services... </p></li><li><p>From the pull-down menu, select the Pathway Commons Web Service Client. </p></li></ul></div><p>Then, follow the three-step process outlined below: </p><p><span class="inlinemediaobject"><img src="images/3_steps.png" alt="3_steps.png"></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Step 1:  Enter your search term;  for example:  BRCA1 </p></li><li><p>Step 2:  Select the protein or small molecule of interest.  Full details regarding each molecule is shown in the bottom left panel. </p></li><li><p>Step 3:  Download a specific pathway or interaction network. </p></li></ul></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Downloading Pathways and Interaction Networks"></a>Downloading Pathways and Interaction Networks</h4></div></div></div><p>In Step 3, you can simply double-click on a pathway of interest, or click on the Interaction Networks tab.  The Interaction Networks tab enables you to filter interactions by data source and/or interaction type.  For example, you can choose to restrict your network to direct physical interactions from HPRD and MINT only: </p><p><span class="inlinemediaobject"><img src="images/intxn_filter.png" alt="intxn_filter.png"></span> </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Pathway Commons Options"></a>Pathway Commons Options</h4></div></div></div><p>You can configure access options from the Options tab.  There are two retrieval options: </p><div class="itemizedlist"><ul type="disc"><li><p>Simplified Binary Model:  Retrieve a simplified binary network, as inferred from the original BioPAX representation.  In this representation, nodes within a network refer to physical entities only, and edges refer to inferred interactions.  </p></li><li><p>Full Model:  Retrieve the full model, as stored in the original BioPAX representation.  In this representation, nodes within a network can refer to physical entities and interactions. </p></li></ul></div><p>By default, the simplified binary model is selected. </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Future Directions"></a>Future Directions</h3></div></div></div><p>As additional web service clients become available, they will be made available via the Cytoscape Plugin Manager.  Once installed, these web service clients will be centrally accessible via the same steps defined above: </p><div class="itemizedlist"><ul type="disc"><li><p>File &#8594; Import &#8594; Network from web services... </p></li></ul></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Import Attributes from External Database"></a>Import Attributes from External Database</h3></div></div></div><p>Some of the web service clients can import attributes from external databases.  <a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/BioMart#" target="_top">BioMart</a> client is an example.  You can install it from Plugin Manager. </p><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Example 1: Import Additional ID Sets and Annotations from BioMart"></a>Example 1: Import Additional ID Sets and Annotations from BioMart</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/biomart1.png" alt="biomart1.png"></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Load a network.  In this example, we use <span class="emphasis"><em>galFiltered.sif</em></span> in <span class="emphasis"><em>sampleData</em></span> directory. </p></li><li><p>File &#8594; Import &#8594; Import Attributes from BioMart... </p></li><li><p>Select <span class="strong"><strong>Data Source</strong></span>.  Since <span class="emphasis"><em>galFiltered.sif</em></span> is a yeast network, select yeast dataset. </p></li><li><p>For <span class="strong"><strong>Key Attribute</strong></span> section, select <span class="emphasis"><em>ID</em></span> for <span class="strong"><strong>Attribute</strong></span> and <span class="strong"><strong>Data Type</strong></span> should be <span class="emphasis"><em>Ensembl Gene ID</em></span>.  <span class="strong"><strong>Attribute</strong></span> is the list of available attributes in current Cytoscape session and <span class="strong"><strong>Data Type</strong></span> is the type of ID set of the attribute.  In this case, Cytoscape uses <span class="emphasis"><em>ID</em></span> as the key for mapping.  Because the sample network <span class="emphasis"><em>galFiletred.sif</em></span> uses <span class="emphasis"><em>Ensemble Gene ID</em></span> for its node ID, like <span class="emphasis"><em>YOR072W</em></span>, you need to select <span class="emphasis"><em>Ensembl Gene ID</em></span> for <span class="strong"><strong>Data Type</strong></span>.  So you need to know the type of ID set (<span class="emphasis"><em>Entrez Gene ID</em></span>, <span class="emphasis"><em><a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/UniProt#" target="_top">UniProt</a> Unified Acc. Number</em></span>, <span class="emphasis"><em>Ensemble Gene ID</em></span>, etc.) of the attribute selected in the <span class="strong"><strong>Attribute</strong></span> box. </p></li><li><p>Select attributes you want to import.  (Note: You cannot select too many attributes at once because BioMart server has maximum number of selectable annotations.) </p></li><li><p>Press <span class="strong"><strong>Import</strong></span>. </p></li><li><p>Now you can see the newly imported attributes on the Attribute Browser.  You may see some attribute names ends with <span class="emphasis"><em>-TOP</em></span> if there are multiple attribute values for a node.  This is an attribute taken from the first entry of the original list attribute. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/biomart2.png" alt="biomart2.png"></span> </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Example 2: Import Annotations from NCBI Entrez Gene Database"></a>Example 2: Import Annotations from NCBI Entrez Gene Database</h4></div></div></div><p><span class="inlinemediaobject"><img src="images/ncbi1.png" alt="ncbi1.png"></span> </p><p>NCBI Entrez Gene database () can be used as network data source and annotation repository.  You can use NCBI web service client to import gene annotations from Entrez Gene. </p><div class="itemizedlist"><ul type="disc"><li><p>File &#8594; Import &#8594; Import attributes from NCBI Entrez Gene... </p></li><li><p><span class="strong"><strong>Data Source</strong></span> is fixed to <span class="emphasis"><em>NCBI Entrez Gene</em></span>. </p></li><li><p><span class="strong"><strong>Data Type</strong></span> is also fixed to <span class="emphasis"><em>Entrez Gene ID</em></span>.  This means Cytoscape attribute selected in <span class="strong"><strong>Attribute</strong></span> list should be Entrez Gene ID. </p></li><li><p>Select <span class="strong"><strong>Annotation Category</strong></span>.  If you select a category, all of the annotation under the category will be imported (i.e., multiple Cytoscape attributes will be created for each category). </p></li><li><p>Press <span class="strong"><strong>Import</strong></span>.  The following is a sample human network (<span class="emphasis"><em>RUAL.subset.sif</em></span>) annotated by NCBI client. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/ncbi2.png" alt="ncbi2.png"></span>  </p></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="Use Multiple Services in a Workflow"></a>Use Multiple Services in a Workflow</h3></div></div></div><p>Web services are useful when you combine the result from multiple data sources. </p><div class="section" lang="en"><div class="titlepage"><div><div><h4 class="title"><a name="Example: Import and Annotate Networks"></a>Example: Import and Annotate Networks</h4></div></div></div><div class="itemizedlist"><ul type="disc"><li><p>Import network from <a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/IntAct#" target="_top">IntAct</a> using keyword.  In this example, type <span class="emphasis"><em>p53 AND species:mouse</em></span>. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/workflow1.png" alt="workflow1.png"></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Import human orthologs from BioMart. </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/workflow2.png" alt="workflow2.png"></span> </p><div class="itemizedlist"><ul type="disc"><li><p>Show the othologs as the list of Ensembl Gene ID on the Data Panel.  Copy them and use them as the query for <a href="http://cytoscape.wodaklab.org/wiki/Cytoscape_User_Manual/wiki/IntAct#" target="_top">IntAct</a>. </p></li><li><p>Import <span class="emphasis"><em>Entrez Gene ID</em></span> from BioMart.  Use <span class="emphasis"><em>ensembl</em></span> attribute for the mapping key. </p></li><li><p>Import annotations from NCBI.  The resulting networks looks like the following: </p></li></ul></div><p><span class="inlinemediaobject"><img src="images/workflow_final.png" alt="workflow_final.png"></span> </p></div></div></div></body></html>